// @dynamic-labs/sdk-api-core/src/models/AuthenticatorTransportProtocol.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.AuthenticatorTransportProtocol = void 0;
(function (AuthenticatorTransportProtocol) {
    AuthenticatorTransportProtocol["Internal"] = "AUTHENTICATOR_TRANSPORT_INTERNAL";
    AuthenticatorTransportProtocol["Usb"] = "AUTHENTICATOR_TRANSPORT_USB";
    AuthenticatorTransportProtocol["Nfc"] = "AUTHENTICATOR_TRANSPORT_NFC";
    AuthenticatorTransportProtocol["Ble"] = "AUTHENTICATOR_TRANSPORT_BLE";
    AuthenticatorTransportProtocol["Hybrid"] = "AUTHENTICATOR_TRANSPORT_HYBRID";
})(exports.AuthenticatorTransportProtocol || (exports.AuthenticatorTransportProtocol = {}));
function AuthenticatorTransportProtocolFromJSON(json) {
    return AuthenticatorTransportProtocolFromJSONTyped(json);
}
function AuthenticatorTransportProtocolFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function AuthenticatorTransportProtocolToJSON(value) {
    return value;
}

exports.AuthenticatorTransportProtocolFromJSON = AuthenticatorTransportProtocolFromJSON;
exports.AuthenticatorTransportProtocolFromJSONTyped = AuthenticatorTransportProtocolFromJSONTyped;
exports.AuthenticatorTransportProtocolToJSON = AuthenticatorTransportProtocolToJSON;
