// @dynamic-labs/sdk-api-core/src/models/ExternalWalletFundingDefaultChain.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
var ExternalWalletFundingDefaultChain;
(function (ExternalWalletFundingDefaultChain) {
    ExternalWalletFundingDefaultChain["Ethereum"] = "ethereum";
    ExternalWalletFundingDefaultChain["Solana"] = "solana";
})(ExternalWalletFundingDefaultChain || (ExternalWalletFundingDefaultChain = {}));
function ExternalWalletFundingDefaultChainFromJSON(json) {
    return ExternalWalletFundingDefaultChainFromJSONTyped(json);
}
function ExternalWalletFundingDefaultChainFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function ExternalWalletFundingDefaultChainToJSON(value) {
    return value;
}

export { ExternalWalletFundingDefaultChain, ExternalWalletFundingDefaultChainFromJSON, ExternalWalletFundingDefaultChainFromJSONTyped, ExternalWalletFundingDefaultChainToJSON };
