// @dynamic-labs/sdk-api-core/src/models/GlobalWalletAccessControlTypeEnum.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.GlobalWalletAccessControlTypeEnum = void 0;
(function (GlobalWalletAccessControlTypeEnum) {
    GlobalWalletAccessControlTypeEnum["Allow"] = "allow";
    GlobalWalletAccessControlTypeEnum["Block"] = "block";
})(exports.GlobalWalletAccessControlTypeEnum || (exports.GlobalWalletAccessControlTypeEnum = {}));
function GlobalWalletAccessControlTypeEnumFromJSON(json) {
    return GlobalWalletAccessControlTypeEnumFromJSONTyped(json);
}
function GlobalWalletAccessControlTypeEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function GlobalWalletAccessControlTypeEnumToJSON(value) {
    return value;
}

exports.GlobalWalletAccessControlTypeEnumFromJSON = GlobalWalletAccessControlTypeEnumFromJSON;
exports.GlobalWalletAccessControlTypeEnumFromJSONTyped = GlobalWalletAccessControlTypeEnumFromJSONTyped;
exports.GlobalWalletAccessControlTypeEnumToJSON = GlobalWalletAccessControlTypeEnumToJSON;
