// @dynamic-labs/sdk-api-core/src/models/HardwareWalletEnum.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.HardwareWalletEnum = void 0;
(function (HardwareWalletEnum) {
    HardwareWalletEnum["Ledger"] = "ledger";
})(exports.HardwareWalletEnum || (exports.HardwareWalletEnum = {}));
function HardwareWalletEnumFromJSON(json) {
    return HardwareWalletEnumFromJSONTyped(json);
}
function HardwareWalletEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function HardwareWalletEnumToJSON(value) {
    return value;
}

exports.HardwareWalletEnumFromJSON = HardwareWalletEnumFromJSON;
exports.HardwareWalletEnumFromJSONTyped = HardwareWalletEnumFromJSONTyped;
exports.HardwareWalletEnumToJSON = HardwareWalletEnumToJSON;
