// @dynamic-labs/sdk-api-core/src/models/KycFieldType.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.KycFieldType = void 0;
(function (KycFieldType) {
    KycFieldType["Standard"] = "standard";
    KycFieldType["Custom"] = "custom";
})(exports.KycFieldType || (exports.KycFieldType = {}));
function KycFieldTypeFromJSON(json) {
    return KycFieldTypeFromJSONTyped(json);
}
function KycFieldTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function KycFieldTypeToJSON(value) {
    return value;
}

exports.KycFieldTypeFromJSON = KycFieldTypeFromJSON;
exports.KycFieldTypeFromJSONTyped = KycFieldTypeFromJSONTyped;
exports.KycFieldTypeToJSON = KycFieldTypeToJSON;
