// @dynamic-labs/sdk-api-core/src/models/MFADeviceType.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.MFADeviceType = void 0;
(function (MFADeviceType) {
    MFADeviceType["Totp"] = "totp";
    MFADeviceType["Passkey"] = "passkey";
    MFADeviceType["Email"] = "email";
})(exports.MFADeviceType || (exports.MFADeviceType = {}));
function MFADeviceTypeFromJSON(json) {
    return MFADeviceTypeFromJSONTyped(json);
}
function MFADeviceTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function MFADeviceTypeToJSON(value) {
    return value;
}

exports.MFADeviceTypeFromJSON = MFADeviceTypeFromJSON;
exports.MFADeviceTypeFromJSONTyped = MFADeviceTypeFromJSONTyped;
exports.MFADeviceTypeToJSON = MFADeviceTypeToJSON;
