// @dynamic-labs/sdk-api-core/src/models/PublicKeyCredentialType.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.PublicKeyCredentialType = void 0;
(function (PublicKeyCredentialType) {
    PublicKeyCredentialType["PublicKey"] = "public-key";
})(exports.PublicKeyCredentialType || (exports.PublicKeyCredentialType = {}));
function PublicKeyCredentialTypeFromJSON(json) {
    return PublicKeyCredentialTypeFromJSONTyped(json);
}
function PublicKeyCredentialTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function PublicKeyCredentialTypeToJSON(value) {
    return value;
}

exports.PublicKeyCredentialTypeFromJSON = PublicKeyCredentialTypeFromJSON;
exports.PublicKeyCredentialTypeFromJSONTyped = PublicKeyCredentialTypeFromJSONTyped;
exports.PublicKeyCredentialTypeToJSON = PublicKeyCredentialTypeToJSON;
