// @dynamic-labs/sdk-api-core/src/models/ZerodevBundlerProvider.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.ZerodevBundlerProvider = void 0;
(function (ZerodevBundlerProvider) {
    ZerodevBundlerProvider["Stackup"] = "STACKUP";
    ZerodevBundlerProvider["Pimlico"] = "PIMLICO";
    ZerodevBundlerProvider["Alchemy"] = "ALCHEMY";
    ZerodevBundlerProvider["Gelato"] = "GELATO";
})(exports.ZerodevBundlerProvider || (exports.ZerodevBundlerProvider = {}));
function ZerodevBundlerProviderFromJSON(json) {
    return ZerodevBundlerProviderFromJSONTyped(json);
}
function ZerodevBundlerProviderFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function ZerodevBundlerProviderToJSON(value) {
    return value;
}

exports.ZerodevBundlerProviderFromJSON = ZerodevBundlerProviderFromJSON;
exports.ZerodevBundlerProviderFromJSONTyped = ZerodevBundlerProviderFromJSONTyped;
exports.ZerodevBundlerProviderToJSON = ZerodevBundlerProviderToJSON;
