// @dynamic-labs/sdk-api-core/src/apis/WaasApi.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../../_virtual/_tslib.cjs');
var runtime = require('../runtime.cjs');
require('../models/ExchangeKeyEnum.cjs');
require('../models/AttestationConveyancePreference.cjs');
require('../models/AuthModeEnum.cjs');
require('../models/AuthStorageEnum.cjs');
require('../models/AuthenticatorAttachment.cjs');
require('../models/ResidentKeyRequirement.cjs');
require('../models/UserVerificationRequirement.cjs');
require('../models/AuthenticatorTransportProtocol.cjs');
require('../models/WaasBackupOptionsEnum.cjs');
require('../models/ThresholdSignatureScheme.cjs');
require('../models/JwtVerifiedCredentialFormatEnum.cjs');
require('../models/ProviderEnum.cjs');
require('../models/WalletAddressType.cjs');
require('../models/EcdsaValidatorOptions.cjs');
require('../models/EmbeddedWalletVersionEnum.cjs');
require('../models/HardwareWalletEnum.cjs');
require('../models/PasswordSourceTypeEnum.cjs');
require('../models/ProviderEntryPointVersionEnum.cjs');
require('../models/ProviderKernelVersionEnum.cjs');
require('../models/WalletProviderEnum.cjs');
require('../models/MfaBackupCodeAcknowledgement.cjs');
require('../models/CustomFieldType.cjs');
require('../models/KycFieldType.cjs');
require('../models/ChainEnum.cjs');
require('../models/EmbeddedWalletChainEnum.cjs');
require('../models/EmbeddedWalletProviderEnum.cjs');
require('../models/UserIdentifierTypeEnum.cjs');
require('../models/WaasChainEnum.cjs');
require('../models/CurrencyType.cjs');
require('../models/DelegatedAccessEncryptionPublicKey.cjs');
var DelegatedAccessEncryptionPublicKeyResponse = require('../models/DelegatedAccessEncryptionPublicKeyResponse.cjs');
require('../models/DelegatedShareDeliveryResponse.cjs');
require('../models/TimeUnitEnum.cjs');
require('../models/EmbeddedWalletAuthType.cjs');
require('../models/EmbeddedWalletSecurityMethod.cjs');
require('../models/EnvironmentEnum.cjs');
require('../models/ExternalWalletFundingDefaultChain.cjs');
require('../models/ExternalWalletFundingTokenRule.cjs');
require('../models/OnrampProviders.cjs');
require('../models/PasskeyTransport.cjs');
require('../models/PasskeyCredentialHint.cjs');
require('../models/PasskeyCredentialType.cjs');
require('../models/PublicKeyCredentialType.cjs');
require('../models/GlobalWalletAccessControlTypeEnum.cjs');
require('../models/GlobalWalletConnectionStatusEnum.cjs');
require('../models/HealthcheckStatus.cjs');
require('../models/MFAAction.cjs');
require('../models/MFADeviceType.cjs');
require('../models/Network.cjs');
require('../models/NextViewEnum.cjs');
require('../models/OauthResultStatus.cjs');
require('../models/SignInProviderEnum.cjs');
require('../models/SocialSignInProviderEnum.cjs');
require('../models/SdkViewSectionAlignment.cjs');
require('../models/SdkViewSectionType.cjs');
require('../models/SdkViewType.cjs');
require('../models/ZerodevBundlerProvider.cjs');
require('../models/PublishEventsEvents.cjs');
require('../models/SsoProviderEnum.cjs');
require('../models/SsoProviderCheckResponse.cjs');
require('../models/UnprocessableEntityErrorCode.cjs');
require('../models/UserFieldsCheckEnum.cjs');
require('../models/WaasDelegatedAccessEncryptionPublicKeyType.cjs');

/* tslint:disable */
/**
 *
 */
class WaasApi extends runtime.BaseAPI {
    /**
     * Fetches the encryption public key used for delegated access in a WAAS environment. By default, returns only the latest active key.
     * Get delegated access encryption public key for an environment
     */
    getDelegatedAccessEncryptionPublicKeyRaw(requestParameters, initOverrides) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            if (requestParameters.environmentId === null || requestParameters.environmentId === undefined) {
                throw new runtime.RequiredError('environmentId', 'Required parameter requestParameters.environmentId was null or undefined when calling getDelegatedAccessEncryptionPublicKey.');
            }
            const queryParameters = {};
            const headerParameters = {};
            const response = yield this.request({
                path: `/sdk/{environmentId}/waas/delegatedAccess/encryptionPublicKey`.replace(`{${"environmentId"}}`, encodeURIComponent(String(requestParameters.environmentId))),
                method: 'GET',
                headers: headerParameters,
                query: queryParameters,
            }, initOverrides);
            return new runtime.JSONApiResponse(response, (jsonValue) => DelegatedAccessEncryptionPublicKeyResponse.DelegatedAccessEncryptionPublicKeyResponseFromJSON(jsonValue));
        });
    }
    /**
     * Fetches the encryption public key used for delegated access in a WAAS environment. By default, returns only the latest active key.
     * Get delegated access encryption public key for an environment
     */
    getDelegatedAccessEncryptionPublicKey(requestParameters, initOverrides) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            const response = yield this.getDelegatedAccessEncryptionPublicKeyRaw(requestParameters, initOverrides);
            return yield response.value();
        });
    }
    /**
     * Returns the allowed HTTP methods and CORS headers for this endpoint.
     * Get CORS and allowed methods for delegated access encryption public keys endpoint
     */
    optionsDelegatedAccessEncryptionPublicKeyRaw(requestParameters, initOverrides) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            if (requestParameters.environmentId === null || requestParameters.environmentId === undefined) {
                throw new runtime.RequiredError('environmentId', 'Required parameter requestParameters.environmentId was null or undefined when calling optionsDelegatedAccessEncryptionPublicKey.');
            }
            const queryParameters = {};
            const headerParameters = {};
            const response = yield this.request({
                path: `/sdk/{environmentId}/waas/delegatedAccess/encryptionPublicKey`.replace(`{${"environmentId"}}`, encodeURIComponent(String(requestParameters.environmentId))),
                method: 'OPTIONS',
                headers: headerParameters,
                query: queryParameters,
            }, initOverrides);
            return new runtime.VoidApiResponse(response);
        });
    }
    /**
     * Returns the allowed HTTP methods and CORS headers for this endpoint.
     * Get CORS and allowed methods for delegated access encryption public keys endpoint
     */
    optionsDelegatedAccessEncryptionPublicKey(requestParameters, initOverrides) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            yield this.optionsDelegatedAccessEncryptionPublicKeyRaw(requestParameters, initOverrides);
        });
    }
}

exports.WaasApi = WaasApi;
