// @dynamic-labs-sdk/client/isMfaRequiredForAction.esm.js
import { g as getDefaultClient, a as getCore, d as assertDefined, c as createApiClient } from './constants.esm.js';

/**
 * Converts an ArrayBuffer to a hex-encoded string
 */ const bufferToHex = (buffer)=>[
        ...new Uint8Array(buffer)
    ].map((x)=>x.toString(16).padStart(2, '0')).join('');

/**
 * Consumes and clears the current MFA token from the client state.
 *
 * This function retrieves the MFA token obtained from a successful MFA authentication,
 * and removes it from the client, to prevent it from being used again.
 *
 * @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
 * @returns The MFA token that was consumed.
 * @throws Error if no MFA token is found.
 */ const consumeMfaToken = (client = getDefaultClient())=>{
    const core = getCore(client);
    const mfaToken = core.state.get().mfaToken;
    assertDefined(mfaToken, 'No MFA token found');
    core.state.set({
        mfaToken: null
    });
    return mfaToken;
};

/**
 * Retrieves the available MFA methods for the current user.
 *
 * This function fetches information about which multi-factor authentication
 * methods are available and configured for the user's account.
 *
 * @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
 * @returns A promise that resolves to the user's MFA methods configuration.
 */ const getMfaMethods = async (client = getDefaultClient())=>{
    const core = getCore(client);
    const apiClient = createApiClient({}, client);
    const userMfaMethods = await apiClient.getUserMfaMethods({
        environmentId: core.environmentId,
        verifiedOnly: true
    });
    return userMfaMethods;
};

/**
 * Checks if MFA is required for a specific action.
 *
 * This function determines whether multi-factor authentication is required
 * for the specified action based on project settings and user configuration.
 *
 * @param params.mfaAction - The action to check MFA requirements for.
 * @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
 * @returns A promise that resolves to true if MFA is required for the action, false otherwise.
 */ const isMfaRequiredForAction = async ({ mfaAction }, client = getDefaultClient())=>{
    var _projectSettings_security_mfa_actions, _projectSettings_security_mfa, _projectSettings_security, _projectSettings_security_mfa1, _projectSettings_security1;
    const projectSettings = client.projectSettings;
    const isRequiredForAction = projectSettings == null ? void 0 : (_projectSettings_security = projectSettings.security) == null ? void 0 : (_projectSettings_security_mfa = _projectSettings_security.mfa) == null ? void 0 : (_projectSettings_security_mfa_actions = _projectSettings_security_mfa.actions) == null ? void 0 : _projectSettings_security_mfa_actions.some((action)=>action.action === mfaAction && action.required);
    // if mfa is not required for the action, return false
    // no matter if general mfa is set to required or not
    if (!isRequiredForAction) {
        return false;
    }
    const isMfaMethodRequired = projectSettings == null ? void 0 : (_projectSettings_security1 = projectSettings.security) == null ? void 0 : (_projectSettings_security_mfa1 = _projectSettings_security1.mfa) == null ? void 0 : _projectSettings_security_mfa1.required;
    // if mfa is required for the action and also set as required in general, return true
    if (isMfaMethodRequired) {
        return true;
    }
    const mfaMethods = await getMfaMethods(client);
    // if general mfa is optional, but user has registered methods,
    // then it should be required for the action
    return mfaMethods.userHasVerifiedMfaMethods;
};

export { bufferToHex as b, consumeMfaToken as c, getMfaMethods as g, isMfaRequiredForAction as i };
