// @dynamic-labs-sdk/client/NotWaasWalletAccountError.cjs.js
'use strict';

var constants = require('./constants.cjs.js');
var getWalletProviderByKey = require('./getWalletProviderByKey.cjs.js');

/**
 * Refreshes the current user's data from the server.
 *
 * This function fetches the latest user information from the backend,
 * updating the local user state with any changes.
 *
 * @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
 * @returns A promise that resolves when the user data is refreshed.
 */ const refreshUser = async (client = constants.getDefaultClient())=>{
    const core = constants.getCore(client);
    const apiClient = constants.createApiClient({}, client);
    const response = await apiClient.refreshAuth({
        environmentId: core.environmentId
    });
    getWalletProviderByKey.updateAuthFromVerifyResponse({
        response
    }, client);
};

class NotWaasWalletAccountError extends constants.BaseError {
    constructor({ walletAddress }){
        super({
            cause: null,
            code: 'not_waas_wallet_account_error',
            docsUrl: null,
            name: 'NotWaasWalletAccountError',
            shortMessage: `Wallet account ${walletAddress} is not a Dynamic WaaS wallet account`
        });
    }
}

exports.NotWaasWalletAccountError = NotWaasWalletAccountError;
exports.refreshUser = refreshUser;
