// @dynamic-labs-sdk/client/NotWaasWalletAccountError.esm.js
import { g as getDefaultClient, a as getCore, c as createApiClient, B as BaseError } from './constants.esm.js';
import { u as updateAuthFromVerifyResponse } from './getWalletProviderByKey.esm.js';

/**
 * Refreshes the current user's data from the server.
 *
 * This function fetches the latest user information from the backend,
 * updating the local user state with any changes.
 *
 * @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
 * @returns A promise that resolves when the user data is refreshed.
 */ const refreshUser = async (client = getDefaultClient())=>{
    const core = getCore(client);
    const apiClient = createApiClient({}, client);
    const response = await apiClient.refreshAuth({
        environmentId: core.environmentId
    });
    updateAuthFromVerifyResponse({
        response
    }, client);
};

class NotWaasWalletAccountError extends BaseError {
    constructor({ walletAddress }){
        super({
            cause: null,
            code: 'not_waas_wallet_account_error',
            docsUrl: null,
            name: 'NotWaasWalletAccountError',
            shortMessage: `Wallet account ${walletAddress} is not a Dynamic WaaS wallet account`
        });
    }
}

export { NotWaasWalletAccountError as N, refreshUser as r };
