// @dynamic-labs-sdk/client/waas.esm.js
import { assertPackageVersion } from '@dynamic-labs-sdk/assert-package-version';
import { B as BaseError, g as getDefaultClient, k as getChainFromVerifiedCredentialChain, d as assertDefined, n as name, v as version } from './constants.esm.js';
import { N as NotWaasWalletAccountError, r as refreshUser } from './NotWaasWalletAccountError.esm.js';
import { a as getWalletProviderFromWalletAccount, l as formatWalletProviderKey, b as getWalletProviderByKey, N as NoWalletProviderFoundError, h as getWalletProviders } from './getWalletProviderByKey.esm.js';
import { D as DYNAMIC_WAAS_METADATA } from './constants.esm2.js';
import { WalletProviderEnum, EmbeddedWalletVersionEnum, JwtVerifiedCredentialFormatEnum } from '@dynamic-labs/sdk-api-core';
import { f as filterDuplicates } from './filterDuplicates.esm.js';
import 'buffer';

class NotWaasWalletProviderError extends BaseError {
    constructor({ walletProviderKey }){
        super({
            cause: null,
            code: 'not_waas_wallet_provider_error',
            docsUrl: null,
            name: 'NotWaasWalletProviderError',
            shortMessage: `Wallet provider ${walletProviderKey} is not a Dynamic WaaS wallet provider`
        });
    }
}

const isWaasWalletProvider = (walletProvider)=>{
    return walletProvider.key.includes(DYNAMIC_WAAS_METADATA.normalizedWalletName);
};

const getWaasWalletProviderFromWalletAccount = ({ walletAccount }, client)=>{
    const walletProvider = getWalletProviderFromWalletAccount({
        walletAccount
    }, client);
    if (!isWaasWalletProvider(walletProvider)) {
        throw new NotWaasWalletAccountError({
            walletAddress: walletAccount.address
        });
    }
    return walletProvider;
};

/**
 * Backs up WaaS wallet key shares to Google Drive for secure recovery.
 *
 * This function creates a backup of the wallet's cryptographic key shares
 * to Google Drive, allowing users to recover their wallet later.
 *
 * @param params.walletAccount - The WaaS wallet account to backup.
 * @param [params.password] - Optional password to encrypt the backup.
 * @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
 * @returns A promise that resolves when the backup process is complete.
 */ const backupWaasKeySharesToGoogleDrive = async ({ password, walletAccount }, client = getDefaultClient())=>{
    const provider = getWaasWalletProviderFromWalletAccount({
        walletAccount
    }, client);
    return provider.backupKeySharesToGoogleDrive({
        password,
        walletAccount
    });
};

/**
 * Creates wallet accounts for the specified chains.
 *
 * This function creates wallet accounts for the specified chains.
 * If the same chain is passed multiple times, just as many wallet accounts will be created
 * for that chain.
 *
 * @example
 * ```ts
 * // Will create 2 wallet accounts for EVM and 1 for SOL
 * await createWaasWalletAccounts({ chains: ['EVM', 'EVM', 'SOL'] });
 * ```
 *
 * @param params.chains - The chains to create wallet accounts for.
 * @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
 * @returns A promise that resolves when all wallet accounts are created.
 */ const createWaasWalletAccounts = async ({ chains }, client = getDefaultClient())=>{
    // we need this right now so we don't try to create wallet accounts for chains
    // that we don't support Dynamid Waas yet in the JS SDK, like SUI or BTC
    const filteredSupportedChains = chains.filter((chain)=>chain === 'EVM' || chain === 'SOL');
    if (filteredSupportedChains.length === 0) {
        return;
    }
    const walletProviderKeys = filteredSupportedChains.map((chain)=>formatWalletProviderKey({
            chain,
            displayName: DYNAMIC_WAAS_METADATA.displayName,
            walletProviderType: WalletProviderEnum.EmbeddedWallet
        }));
    const walletProviderEntries = filterDuplicates(walletProviderKeys).map((walletProviderKey)=>{
        const walletProvider = getWalletProviderByKey({
            walletProviderKey
        }, client);
        if (!walletProvider) {
            throw new NoWalletProviderFoundError({
                walletProviderKey
            });
        }
        if (!isWaasWalletProvider(walletProvider)) {
            throw new NotWaasWalletProviderError({
                walletProviderKey
            });
        }
        return [
            walletProviderKey,
            walletProvider
        ];
    });
    const walletProviderMap = Object.fromEntries(walletProviderEntries);
    await Promise.all(walletProviderKeys.map((walletProviderKey)=>walletProviderMap[walletProviderKey].createWalletAccount()));
    await refreshUser(client);
};

/**
 * Delegates cryptographic key shares for a WaaS wallet account.
 *
 * This function allows the delegation of key management responsibilities,
 * enabling distributed key management for enhanced security.
 *
 * @param params.walletAccount - The WaaS wallet account to delegate shares for.
 * @param [params.password] - Optional password for key share operations.
 * @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
 * @returns A promise that resolves when the key delegation is complete.
 */ const delegateWaasKeyShares = async ({ password, walletAccount }, client = getDefaultClient())=>{
    const provider = getWaasWalletProviderFromWalletAccount({
        walletAccount
    }, client);
    return provider.delegateKeyShares({
        password,
        walletAccount
    });
};

/**
 * Exports the client-side key shares for a WaaS wallet account.
 *
 * This function retrieves the cryptographic key shares that are stored
 * on the client side for backup or migration purposes.
 *
 * @param params.walletAccount - The WaaS wallet account to export shares from.
 * @param [params.password] - Optional password to encrypt the exported shares.
 * @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
 * @returns A promise that resolves to the exported key shares.
 */ const exportWaasClientKeyshares = async ({ password, walletAccount }, client = getDefaultClient())=>{
    const provider = getWaasWalletProviderFromWalletAccount({
        walletAccount
    }, client);
    return provider.exportClientKeyshares({
        password,
        walletAccount
    });
};

/**
 * Exports the private key for a WaaS wallet account with secure display.
 *
 * This function reveals the private key for a WaaS wallet and displays it
 * securely within an iframe dynamically inserted into the specified HTML container element.
 *
 * @param params.walletAccount - The WaaS wallet account to export the private key from.
 * @param params.displayContainer - The HTML element where the iframe will be injected.
 * @param [params.password] - Optional password to decrypt the private key.
 * @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
 * @returns A promise that resolves when the private key export is complete.
 */ const exportWaasPrivateKey = async ({ displayContainer, password, walletAccount }, client = getDefaultClient())=>{
    const provider = getWaasWalletProviderFromWalletAccount({
        walletAccount
    }, client);
    return provider.exportPrivateKey({
        displayContainer,
        password,
        walletAccount
    });
};

const userHasEmbeddedWalletForChain = ({ user, chain })=>{
    const hasEmbeddedWalletForChain = user.verifiedCredentials.some((vc)=>vc.walletProvider === WalletProviderEnum.EmbeddedWallet && vc.chain && getChainFromVerifiedCredentialChain(vc.chain) === chain);
    return hasEmbeddedWalletForChain;
};

/**
 * Checks if Dynamic WaaS (Wallet as a Service) is enabled for the current project.
 *
 * This function examines the project settings to determine if Dynamic WaaS is
 * enabled in the Dynamic Dashboard or not.
 *
 * @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
 * @returns True if Dynamic WaaS is enabled, false otherwise.
 */ const isDynamicWaasEnabled = (client = getDefaultClient())=>{
    const projectSettings = client.projectSettings;
    assertDefined(projectSettings, 'Project settings are not available');
    const embeddedWalletSettings = projectSettings.sdk.embeddedWallets;
    const dynamicWaasIsEnabled = (embeddedWalletSettings == null ? void 0 : embeddedWalletSettings.defaultWalletVersion) === EmbeddedWalletVersionEnum.V3;
    return dynamicWaasIsEnabled;
};

/**
 * Gets all chains missing a waas wallet account for the current user.
 *
 * @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
 * @returns An array of all chains that have waas wallet accounts enabled
 * but for which the current user does not yet have an embedded wallet account.
 */ const getChainsMissingWaasWalletAccounts = (client = getDefaultClient())=>{
    var _embeddedWalletSettings_chainConfigurations;
    const projectSettings = client.projectSettings;
    assertDefined(projectSettings, 'Project settings are not available');
    const embeddedWalletSettings = projectSettings.sdk.embeddedWallets;
    const isWaasEnabled = isDynamicWaasEnabled(client);
    const user = client.user;
    if (!user || !isWaasEnabled) {
        return [];
    }
    var _embeddedWalletSettings_chainConfigurations_filter;
    const enabledChainForWaas = ((_embeddedWalletSettings_chainConfigurations_filter = embeddedWalletSettings == null ? void 0 : (_embeddedWalletSettings_chainConfigurations = embeddedWalletSettings.chainConfigurations) == null ? void 0 : _embeddedWalletSettings_chainConfigurations.filter((chain)=>chain.enabled)) != null ? _embeddedWalletSettings_chainConfigurations_filter : []).map((chain)=>chain.name);
    // For each enabled chain, if user does NOT have an embedded wallet, add to missing
    return enabledChainForWaas.filter((chain)=>!userHasEmbeddedWalletForChain({
            chain,
            user
        }));
};

const getWaasWalletProviderByChain = ({ chain }, client)=>{
    const providers = getWalletProviders(client);
    const waasProviderKey = formatWalletProviderKey({
        chain,
        displayName: DYNAMIC_WAAS_METADATA.displayName,
        walletProviderType: WalletProviderEnum.EmbeddedWallet
    });
    const waasProvider = providers.find((provider)=>provider.key === waasProviderKey && provider.chain === chain);
    if (!waasProvider || !isWaasWalletProvider(waasProvider)) {
        throw new NoWalletProviderFoundError({
            walletProviderKey: waasProviderKey
        });
    }
    return waasProvider;
};

/**
 * Imports an existing private key to create a new WaaS wallet account.
 *
 * This function allows users to import their existing private key into
 * Dynamic's WaaS infrastructure for secure management.
 *
 * @param params.chain - The chain for the imported wallet.
 * @param params.privateKey - The private key to import.
 * @param [params.thresholdSignatureScheme] - Optional threshold signature scheme configuration. Default is TWO_OF_TWO.
 * @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
 * @returns A promise that resolves when the private key is successfully imported.
 */ const importWaasPrivateKey = async ({ chain, privateKey, thresholdSignatureScheme }, client = getDefaultClient())=>{
    const provider = getWaasWalletProviderByChain({
        chain
    }, client);
    return provider.importPrivateKey({
        privateKey,
        thresholdSignatureScheme
    });
};

/**
 * This function determines whether the provided wallet account is a Dynamic
 * WaaS wallet account.
 *
 * @param params.walletAccount - The wallet account to check.
 * @returns True if the wallet account is a WaaS wallet account, false otherwise.
 */ const isWaasWalletAccount = ({ walletAccount })=>{
    return walletAccount.walletProviderKey.includes(DYNAMIC_WAAS_METADATA.normalizedWalletName);
};

/**
 * Refreshes the cryptographic key shares for a WaaS wallet account.
 *
 * This function generates new key shares for the wallet while maintaining
 * the same threshold signature scheme.
 *
 * @param params.walletAccount - The WaaS wallet account to refresh shares for.
 * @param [params.password] - Optional password for key share operations.
 * @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
 * @returns A promise that resolves when the key shares are refreshed.
 */ const refreshWaasWalletAccountShares = async ({ walletAccount, password }, client = getDefaultClient())=>{
    const provider = getWaasWalletProviderFromWalletAccount({
        walletAccount
    }, client);
    return provider.refreshWalletAccountShares({
        password,
        walletAccount
    });
};

const userHasExternalWallet = (user)=>{
    const hasExternalWallet = user == null ? void 0 : user.verifiedCredentials.some((vc)=>vc.format === JwtVerifiedCredentialFormatEnum.Blockchain && vc.walletProvider !== WalletProviderEnum.EmbeddedWallet);
    return hasExternalWallet;
};

/**
 * Checks if a WaaS wallet should be automatically created for the specified blockchain.
 *
 * This function determines whether the current user configuration and project
 * settings require automatic wallet creation for the given chain, considering
 * whether the user already has an embedded wallet for the enabled chains or an external wallet.
 *
 * @param params.chain - The blockchain network to check.
 * @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
 * @returns True if a wallet should be auto-created for the chain, false otherwise.
 */ const shouldAutoCreateWalletForChain = ({ chain }, client = getDefaultClient())=>{
    assertDefined(client.projectSettings, 'Project settings are not available');
    const embeddedWalletSettings = client.projectSettings.sdk.embeddedWallets;
    const automaticEmbeddedWalletCreationEnabled = embeddedWalletSettings == null ? void 0 : embeddedWalletSettings.automaticEmbeddedWalletCreation;
    if (!automaticEmbeddedWalletCreationEnabled || !client.user) {
        return false;
    }
    const automaticEmbeddedWalletCreationForExternalEnabled = embeddedWalletSettings == null ? void 0 : embeddedWalletSettings.automaticEmbeddedWalletCreationForExternal;
    // If user has an external wallet (e.g metamask, phantom, etc), should only
    // create if automaticEmbeddedWalletCreationForExternalEnabled is true
    const shouldCreateDynamicWaasWallet = !userHasExternalWallet(client.user) || automaticEmbeddedWalletCreationForExternalEnabled;
    if (!shouldCreateDynamicWaasWallet) {
        return false;
    }
    const chainsMissingAutoCreatedWallet = getChainsMissingWaasWalletAccounts(client);
    return chainsMissingAutoCreatedWallet.includes(chain);
};

/**
 * Updates the password for a WaaS wallet account.
 *
 * This function changes the password used to encrypt and protect the
 * cryptographic key shares for the specified wallet account.
 *
 * @param params.walletAccount - The WaaS wallet account to update the password for.
 * @param params.currentPassword - The current password for the wallet.
 * @param params.newPassword - The new password to set for the wallet.
 * @param [client] - The Dynamic client instance. Only required when using multiple Dynamic clients.
 * @returns A promise that resolves when the password is successfully updated.
 */ const updateWaasPassword = async ({ currentPassword, newPassword, walletAccount }, client = getDefaultClient())=>{
    const provider = getWaasWalletProviderFromWalletAccount({
        walletAccount
    }, client);
    return provider.updatePassword({
        currentPassword,
        newPassword,
        walletAccount
    });
};

assertPackageVersion(name, version);

export { NotWaasWalletProviderError, backupWaasKeySharesToGoogleDrive, createWaasWalletAccounts, delegateWaasKeyShares, exportWaasClientKeyshares, exportWaasPrivateKey, getChainsMissingWaasWalletAccounts, importWaasPrivateKey, isDynamicWaasEnabled, isWaasWalletAccount, refreshWaasWalletAccountShares, shouldAutoCreateWalletForChain, updateWaasPassword };
