// @dynamic-labs/sui-core/src/index.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var assertPackageVersion = require('@dynamic-labs/assert-package-version');
var _package = require('../package.cjs');
var constants = require('./utils/constants/constants.cjs');
var SuiWalletConnector = require('./SuiWalletConnector.cjs');
var SuiWallet = require('./wallet/SuiWallet.cjs');
var isSuiWallet = require('./wallet/isSuiWallet/isSuiWallet.cjs');
var SuiUiTransaction = require('./utils/SuiUiTransaction/SuiUiTransaction.cjs');
var fetchSuiInjectedWalletConnectors = require('./wallets/injected/fetchSuiInjectedWalletConnectors.cjs');

/* eslint-disable @typescript-eslint/no-unused-vars */
assertPackageVersion.assertPackageVersion('@dynamic-labs/sui-core', _package.version);

exports.MIST_PER_SUI = constants.MIST_PER_SUI;
exports.SuiWalletConnector = SuiWalletConnector.SuiWalletConnector;
exports.SuiWallet = SuiWallet.SuiWallet;
exports.isSuiWallet = isSuiWallet.isSuiWallet;
exports.SuiUiTransaction = SuiUiTransaction.SuiUiTransaction;
exports.fetchSuiInjectedWalletConnectors = fetchSuiInjectedWalletConnectors.fetchSuiInjectedWalletConnectors;
