// @dynamic-labs/sdk-react-core/src/lib/shared/assets/globe.js
'use client'
import * as React from 'react';

var _path;
var _excluded = ["title", "titleId"];
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
function _objectWithoutProperties(e, t) { if (null == e) return {}; var o, r, i = _objectWithoutPropertiesLoose(e, t); if (Object.getOwnPropertySymbols) { var s = Object.getOwnPropertySymbols(e); for (r = 0; r < s.length; r++) o = s[r], t.includes(o) || {}.propertyIsEnumerable.call(e, o) && (i[o] = e[o]); } return i; }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (e.includes(n)) continue; t[n] = r[n]; } return t; }
var SvgGlobe = function SvgGlobe(_ref) {
  var title = _ref.title,
    titleId = _ref.titleId,
    props = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/React.createElement("svg", _extends({
    width: 28,
    height: 28,
    viewBox: "0 0 28 28",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg",
    "aria-labelledby": titleId
  }, props), title ? /*#__PURE__*/React.createElement("title", {
    id: titleId
  }, title) : null, _path || (_path = /*#__PURE__*/React.createElement("path", {
    fillRule: "evenodd",
    clipRule: "evenodd",
    d: "M4.738 12.84h4.616c.09-2.54.47-4.854 1.064-6.634.114-.342.238-.671.373-.984a9.352 9.352 0 0 0-6.053 7.617ZM14 2.332C7.55 2.333 2.334 7.563 2.334 14c0 6.437 5.217 11.667 11.666 11.667 6.45 0 11.667-5.23 11.667-11.667C25.667 7.563 20.45 2.333 14 2.333Zm0 2.322s.007 0 .016.003a.343.343 0 0 1 .077.037c.08.05.202.151.351.34.304.387.63 1.018.924 1.901.502 1.51.855 3.568.944 5.903h-4.623c.089-2.335.441-4.394.944-5.903.294-.883.619-1.514.923-1.9.15-.19.271-.291.352-.34a.343.343 0 0 1 .076-.038L14 4.655ZM9.354 15.161H4.738a9.352 9.352 0 0 0 6.052 7.616 12.295 12.295 0 0 1-.372-.983c-.593-1.78-.974-4.095-1.064-6.633ZM14 23.345s-.006 0-.016-.003a.34.34 0 0 1-.076-.037c-.081-.05-.202-.151-.352-.341-.304-.386-.629-1.017-.923-1.9-.503-1.51-.855-3.568-.944-5.903h4.623c-.09 2.335-.442 4.394-.944 5.902-.295.884-.48 1.3-.784 1.687-.15.19-.41.505-.491.555a.34.34 0 0 1-.077.037l-.016.003c1.127 0 2.209-.2 3.21-.568a9.351 9.351 0 0 0 6.052-7.616h-4.616c-.09 2.539-.47 4.853-1.064 6.633-.113.342-.238.671-.372.983m1.436-9.938c-.09-2.539-.47-4.853-1.064-6.633a12.305 12.305 0 0 0-.372-.984 9.351 9.351 0 0 1 6.052 7.617h-4.616Z",
    fill: "currentColor"
  })));
};

export { SvgGlobe as ReactComponent };
