// @dynamic-labs/sdk-react-core/src/lib/shared/assets/iphone-face-id.js
'use client'
import * as React from 'react';

var _path;
var _excluded = ["title", "titleId"];
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
function _objectWithoutProperties(e, t) { if (null == e) return {}; var o, r, i = _objectWithoutPropertiesLoose(e, t); if (Object.getOwnPropertySymbols) { var s = Object.getOwnPropertySymbols(e); for (r = 0; r < s.length; r++) o = s[r], t.includes(o) || {}.propertyIsEnumerable.call(e, o) && (i[o] = e[o]); } return i; }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (e.includes(n)) continue; t[n] = r[n]; } return t; }
var SvgIphoneFaceId = function SvgIphoneFaceId(_ref) {
  var title = _ref.title,
    titleId = _ref.titleId,
    props = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/React.createElement("svg", _extends({
    width: 64,
    height: 64,
    viewBox: "0 0 64 64",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg",
    "aria-labelledby": titleId
  }, props), title ? /*#__PURE__*/React.createElement("title", {
    id: titleId
  }, title) : null, _path || (_path = /*#__PURE__*/React.createElement("path", {
    fillRule: "evenodd",
    clipRule: "evenodd",
    d: "M3.25 17.563v-7.188a7.133 7.133 0 0 1 7.125-7.125h7.188a1.625 1.625 0 0 0 0-3.25h-7.188C4.655 0 0 4.654 0 10.375v7.188a1.625 1.625 0 0 0 3.25 0ZM17.563 60.75a1.625 1.625 0 0 1 0 3.25h-7.188C4.655 64 0 59.346 0 53.625v-7.188a1.625 1.625 0 0 1 3.25 0v7.188a7.133 7.133 0 0 0 7.125 7.125h7.188ZM64 46.437v7.188C64 59.345 59.346 64 53.625 64h-7.188a1.625 1.625 0 0 1 0-3.25h7.188a7.133 7.133 0 0 0 7.125-7.125v-7.188a1.625 1.625 0 0 1 3.25 0Zm0-36.062v7.188a1.625 1.625 0 0 1-3.25 0v-7.188a7.133 7.133 0 0 0-7.125-7.125h-7.188a1.625 1.625 0 0 1 0-3.25h7.188C59.345 0 64 4.654 64 10.375ZM43.214 47.302a1.687 1.687 0 0 0-2.29-2.48A13.115 13.115 0 0 1 32 48.313c-3.319 0-6.488-1.239-8.924-3.49a1.688 1.688 0 0 0-2.29 2.48A16.481 16.481 0 0 0 32 51.687c4.17 0 8.153-1.557 11.214-4.385Zm-7.839-23.177v12a4.317 4.317 0 0 1-4.313 4.313h-1.375a1.687 1.687 0 1 1 0-3.376h1.375c.517 0 .938-.42.938-.937v-12a1.687 1.687 0 1 1 3.375 0ZM47 28.844V24.03a1.594 1.594 0 0 0-3.188 0v4.813a1.594 1.594 0 0 0 3.188 0Zm-29.625 0a1.594 1.594 0 0 0 3.188 0V24.03a1.594 1.594 0 0 0-3.188 0v4.813Z",
    fill: "#000"
  })));
};

export { SvgIphoneFaceId as ReactComponent };
