// @dynamic-labs/sdk-react-core/src/lib/shared/assets/link-circle.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);

var _path, _g, _defs;
var _excluded = ["title", "titleId"];
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
function _objectWithoutProperties(e, t) { if (null == e) return {}; var o, r, i = _objectWithoutPropertiesLoose(e, t); if (Object.getOwnPropertySymbols) { var s = Object.getOwnPropertySymbols(e); for (r = 0; r < s.length; r++) o = s[r], t.includes(o) || {}.propertyIsEnumerable.call(e, o) && (i[o] = e[o]); } return i; }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (e.includes(n)) continue; t[n] = r[n]; } return t; }
var SvgLinkCircle = function SvgLinkCircle(_ref) {
  var title = _ref.title,
    titleId = _ref.titleId,
    props = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/React__namespace.createElement("svg", _extends({
    width: 24,
    height: 24,
    viewBox: "0 0 24 24",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg",
    "aria-labelledby": titleId
  }, props), title ? /*#__PURE__*/React__namespace.createElement("title", {
    id: titleId
  }, title) : null, _path || (_path = /*#__PURE__*/React__namespace.createElement("path", {
    fillRule: "evenodd",
    clipRule: "evenodd",
    d: "M12 20a8 8 0 1 0 0-16 8 8 0 0 0 0 16Zm0 2c5.523 0 10-4.477 10-10S17.523 2 12 2 2 6.477 2 12s4.477 10 10 10Z",
    fill: "#ABACB2"
  })), _g || (_g = /*#__PURE__*/React__namespace.createElement("g", {
    clipPath: "url(#link-circle_svg__a)"
  }, /*#__PURE__*/React__namespace.createElement("path", {
    d: "M13.58 8.58a1.3 1.3 0 1 1 1.84 1.84l-2 2a1.3 1.3 0 0 1-1.84 0 .7.7 0 0 0-.99.99 2.7 2.7 0 0 0 3.82 0l2-2a2.7 2.7 0 0 0-3.82-3.82l-.549.55a.7.7 0 1 0 .99.99l.55-.55Zm-3 3a1.3 1.3 0 0 1 1.84 0 .7.7 0 0 0 .99-.99 2.7 2.7 0 0 0-3.82 0l-2 2a2.7 2.7 0 0 0 3.82 3.82l.55-.552a.7.7 0 0 0-.99-.99l-.55.551a1.3 1.3 0 1 1-1.84-1.838l2-2Z",
    fill: "#ABACB2",
    stroke: "#ABACB2",
    strokeWidth: 0.4,
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }))), _defs || (_defs = /*#__PURE__*/React__namespace.createElement("defs", null, /*#__PURE__*/React__namespace.createElement("clipPath", {
    id: "link-circle_svg__a"
  }, /*#__PURE__*/React__namespace.createElement("path", {
    fill: "#fff",
    transform: "translate(6 6)",
    d: "M0 0h12v12H0z"
  })))));
};

exports.ReactComponent = SvgLinkCircle;
