// @dynamic-labs/sdk-react-core/src/lib/shared/assets/profile-outline.js
'use client'
import * as React from 'react';

var _path;
var _excluded = ["title", "titleId"];
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
function _objectWithoutProperties(e, t) { if (null == e) return {}; var o, r, i = _objectWithoutPropertiesLoose(e, t); if (Object.getOwnPropertySymbols) { var s = Object.getOwnPropertySymbols(e); for (r = 0; r < s.length; r++) o = s[r], t.includes(o) || {}.propertyIsEnumerable.call(e, o) && (i[o] = e[o]); } return i; }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (e.includes(n)) continue; t[n] = r[n]; } return t; }
var SvgProfileOutline = function SvgProfileOutline(_ref) {
  var title = _ref.title,
    titleId = _ref.titleId,
    props = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/React.createElement("svg", _extends({
    width: 25,
    height: 24,
    viewBox: "0 0 25 24",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg",
    "aria-labelledby": titleId
  }, props), title ? /*#__PURE__*/React.createElement("title", {
    id: titleId
  }, title) : null, _path || (_path = /*#__PURE__*/React.createElement("path", {
    fillRule: "evenodd",
    clipRule: "evenodd",
    d: "M12.5 4a8 8 0 0 0-6.598 12.526A14.943 14.943 0 0 1 12.5 15c2.366 0 4.606.548 6.598 1.526A8 8 0 0 0 12.5 4Zm7.654 14.436A9.96 9.96 0 0 0 22.5 12c0-5.523-4.477-10-10-10s-10 4.477-10 10a9.96 9.96 0 0 0 2.37 6.464A9.978 9.978 0 0 0 12.5 22a9.978 9.978 0 0 0 7.654-3.564ZM17.7 18.08A12.954 12.954 0 0 0 12.5 17c-1.85 0-3.607.386-5.199 1.08A7.968 7.968 0 0 0 12.5 20c1.985 0 3.8-.723 5.199-1.92ZM12.5 8a2 2 0 1 0 0 4 2 2 0 0 0 0-4Zm-4 2a4 4 0 1 1 8 0 4 4 0 0 1-8 0Z",
    fill: "currentColor"
  })));
};

export { SvgProfileOutline as ReactComponent };
