// @dynamic-labs/sdk-react-core/src/lib/shared/assets/qr-code.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);

var _path;
var _excluded = ["title", "titleId"];
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
function _objectWithoutProperties(e, t) { if (null == e) return {}; var o, r, i = _objectWithoutPropertiesLoose(e, t); if (Object.getOwnPropertySymbols) { var s = Object.getOwnPropertySymbols(e); for (r = 0; r < s.length; r++) o = s[r], t.includes(o) || {}.propertyIsEnumerable.call(e, o) && (i[o] = e[o]); } return i; }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (e.includes(n)) continue; t[n] = r[n]; } return t; }
var SvgQrCode = function SvgQrCode(_ref) {
  var title = _ref.title,
    titleId = _ref.titleId,
    props = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/React__namespace.createElement("svg", _extends({
    width: 28,
    height: 28,
    viewBox: "0 0 28 28",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg",
    "aria-labelledby": titleId
  }, props), title ? /*#__PURE__*/React__namespace.createElement("title", {
    id: titleId
  }, title) : null, _path || (_path = /*#__PURE__*/React__namespace.createElement("path", {
    fillRule: "evenodd",
    clipRule: "evenodd",
    d: "M3.5 5.833A2.333 2.333 0 0 1 5.833 3.5h2.334A2.333 2.333 0 0 1 10.5 5.833v2.334A2.333 2.333 0 0 1 8.167 10.5H5.833A2.333 2.333 0 0 1 3.5 8.167V5.833Zm4.667 0H5.833v2.334h2.334V5.833ZM14 3.5c.644 0 1.167.522 1.167 1.167v1.166a1.167 1.167 0 0 1-2.334 0V4.667c0-.645.523-1.167 1.167-1.167Zm3.5 2.333A2.333 2.333 0 0 1 19.833 3.5h2.334A2.333 2.333 0 0 1 24.5 5.833v2.334a2.333 2.333 0 0 1-2.333 2.333h-2.334A2.333 2.333 0 0 1 17.5 8.167V5.833Zm4.667 0h-2.334v2.334h2.334V5.833ZM14 9.333c.644 0 1.167.523 1.167 1.167v2.333h3.511a1.167 1.167 0 0 1 0 2.334H14A1.167 1.167 0 0 1 12.833 14v-3.5c0-.644.523-1.167 1.167-1.167ZM3.5 14c0-.644.522-1.167 1.167-1.167h4.666a1.167 1.167 0 0 1 0 2.334H4.667A1.167 1.167 0 0 1 3.5 14Zm18.667 0c0-.644.522-1.167 1.166-1.167h.012a1.167 1.167 0 0 1 0 2.334h-.012A1.167 1.167 0 0 1 22.167 14ZM3.5 19.833A2.333 2.333 0 0 1 5.833 17.5h2.334a2.333 2.333 0 0 1 2.333 2.333v2.334A2.333 2.333 0 0 1 8.167 24.5H5.833A2.333 2.333 0 0 1 3.5 22.167v-2.334Zm4.667 0H5.833v2.334h2.334v-2.334Zm4.666-1.166c0-.645.523-1.167 1.167-1.167h2.333a1.167 1.167 0 1 1 0 2.333h-1.166v3.5a1.167 1.167 0 0 1-2.334 0v-4.666Zm7 0c0-.645.523-1.167 1.167-1.167h2.333a1.167 1.167 0 1 1 0 2.333H21a1.167 1.167 0 0 1-1.167-1.166ZM17.5 23.333c0-.644.522-1.166 1.167-1.166h4.666a1.167 1.167 0 1 1 0 2.333h-4.666a1.167 1.167 0 0 1-1.167-1.167Z",
    fill: "currentColor"
  })));
};

exports.ReactComponent = SvgQrCode;
