// @dynamic-labs/ethereum/src/utils/getConnectorConstructorForEip6963Wallet/getConnectorConstructorForEip6963Wallet.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var utils = require('@dynamic-labs/utils');
var InjectedWalletBase = require('../../injected/InjectedWalletBase.cjs');

const getConnectorConstructorForEip6963Wallet = (walletDetail) => {
    const { info, provider } = walletDetail;
    const sanitizedName = utils.sanitizeName(info.name);
    return class extends InjectedWalletBase.InjectedWalletBase {
        constructor(props) {
            super(Object.assign(Object.assign({}, props), { metadata: {
                    groupKey: sanitizedName,
                    icon: info.icon,
                    id: info.uuid,
                    name: info.name,
                    rdns: info.rdns,
                } }));
            this.name = info.name;
            this.overrideKey = `${sanitizedName}evm`;
        }
        findProvider() {
            return provider;
        }
    };
};

exports.getConnectorConstructorForEip6963Wallet = getConnectorConstructorForEip6963Wallet;
