// @dynamic-labs/wallet-connector-core/src/utils/getChainInfo/getChainInfo.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const createChainInfo = (blockchainName, name, symbol, displayName) => ({
    blockchainName,
    displayName: displayName !== null && displayName !== void 0 ? displayName : blockchainName,
    name,
    symbol: symbol,
});
const chainsInfo = [
    createChainInfo('Algorand', 'algorand', 'ALGO'),
    createChainInfo('Aptos', 'aptos', 'APTOS'),
    createChainInfo('Bitcoin', 'bitcoin', 'BTC'),
    createChainInfo('Cosmos', 'cosmos', 'COSMOS'),
    createChainInfo('Eclipse', 'eclipse', 'ECLIPSE'),
    createChainInfo('Ethereum', 'evm', 'ETH', 'EVM'),
    createChainInfo('Flow', 'flow', 'FLOW'),
    createChainInfo('Solana', 'solana', 'SOL'),
    createChainInfo('Spark', 'spark', 'SPARK'),
    createChainInfo('Starknet', 'starknet', 'ETH'),
    createChainInfo('Sui', 'sui', 'SUI'),
    createChainInfo('Tron', 'tron', 'TRON'),
];
const chainOverrides = {
    algo: 'algorand',
    bip122: 'bitcoin',
    btc: 'bitcoin',
    eip155: 'evm',
    eth: 'evm',
    sol: 'solana',
    stark: 'starknet',
    sui: 'sui',
};
let chainInfoOverrides;
const setChainInfoOverrides = (overrides) => {
    chainInfoOverrides = overrides;
};
const getChainInfo = (chain) => {
    var _a;
    const lowerCasedChain = chain.toLowerCase();
    const normalizedChain = (_a = chainOverrides[lowerCasedChain]) !== null && _a !== void 0 ? _a : lowerCasedChain;
    const chainInfo = chainsInfo.find((info) => info.name === normalizedChain ||
        info.symbol.toLocaleLowerCase() === normalizedChain);
    if (!chainInfo) {
        return;
    }
    return chainInfo;
};
const getChainInfoWithOverrides = (chain) => {
    var _a, _b;
    const chainInfo = getChainInfo(chain);
    if (!chainInfo) {
        return;
    }
    const overrides = chainInfoOverrides === null || chainInfoOverrides === void 0 ? void 0 : chainInfoOverrides[chainInfo.name];
    const chainInfoClone = Object.assign({}, chainInfo);
    if (overrides) {
        chainInfoClone.blockchainName =
            (_a = overrides.displayName) !== null && _a !== void 0 ? _a : chainInfo.blockchainName;
        chainInfoClone.displayName = (_b = overrides.displayName) !== null && _b !== void 0 ? _b : chainInfo.displayName;
    }
    return chainInfoClone;
};

exports.getChainInfo = getChainInfo;
exports.getChainInfoWithOverrides = getChainInfoWithOverrides;
exports.setChainInfoOverrides = setChainInfoOverrides;
