// @dynamic-labs/sdk-api-core/src/models/EmbeddedWalletChainEnum.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.EmbeddedWalletChainEnum = void 0;
(function (EmbeddedWalletChainEnum) {
    EmbeddedWalletChainEnum["Evm"] = "EVM";
    EmbeddedWalletChainEnum["Sol"] = "SOL";
    EmbeddedWalletChainEnum["Sui"] = "SUI";
    EmbeddedWalletChainEnum["Btc"] = "BTC";
})(exports.EmbeddedWalletChainEnum || (exports.EmbeddedWalletChainEnum = {}));
function EmbeddedWalletChainEnumFromJSON(json) {
    return EmbeddedWalletChainEnumFromJSONTyped(json);
}
function EmbeddedWalletChainEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function EmbeddedWalletChainEnumToJSON(value) {
    return value;
}

exports.EmbeddedWalletChainEnumFromJSON = EmbeddedWalletChainEnumFromJSON;
exports.EmbeddedWalletChainEnumFromJSONTyped = EmbeddedWalletChainEnumFromJSONTyped;
exports.EmbeddedWalletChainEnumToJSON = EmbeddedWalletChainEnumToJSON;
