// @dynamic-labs/sdk-api-core/src/models/GlobalWalletConnectionStatusEnum.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
var GlobalWalletConnectionStatusEnum;
(function (GlobalWalletConnectionStatusEnum) {
    GlobalWalletConnectionStatusEnum["Connected"] = "connected";
    GlobalWalletConnectionStatusEnum["Disconnected"] = "disconnected";
})(GlobalWalletConnectionStatusEnum || (GlobalWalletConnectionStatusEnum = {}));
function GlobalWalletConnectionStatusEnumFromJSON(json) {
    return GlobalWalletConnectionStatusEnumFromJSONTyped(json);
}
function GlobalWalletConnectionStatusEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function GlobalWalletConnectionStatusEnumToJSON(value) {
    return value;
}

export { GlobalWalletConnectionStatusEnum, GlobalWalletConnectionStatusEnumFromJSON, GlobalWalletConnectionStatusEnumFromJSONTyped, GlobalWalletConnectionStatusEnumToJSON };
