// @dynamic-labs/sdk-api-core/src/models/JwtVerifiedCredentialFormatEnum.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
var JwtVerifiedCredentialFormatEnum;
(function (JwtVerifiedCredentialFormatEnum) {
    JwtVerifiedCredentialFormatEnum["Blockchain"] = "blockchain";
    JwtVerifiedCredentialFormatEnum["Email"] = "email";
    JwtVerifiedCredentialFormatEnum["Oauth"] = "oauth";
    JwtVerifiedCredentialFormatEnum["Passkey"] = "passkey";
    JwtVerifiedCredentialFormatEnum["PhoneNumber"] = "phoneNumber";
    JwtVerifiedCredentialFormatEnum["ExternalUser"] = "externalUser";
})(JwtVerifiedCredentialFormatEnum || (JwtVerifiedCredentialFormatEnum = {}));
function JwtVerifiedCredentialFormatEnumFromJSON(json) {
    return JwtVerifiedCredentialFormatEnumFromJSONTyped(json);
}
function JwtVerifiedCredentialFormatEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function JwtVerifiedCredentialFormatEnumToJSON(value) {
    return value;
}

export { JwtVerifiedCredentialFormatEnum, JwtVerifiedCredentialFormatEnumFromJSON, JwtVerifiedCredentialFormatEnumFromJSONTyped, JwtVerifiedCredentialFormatEnumToJSON };
