// @dynamic-labs/sdk-api-core/src/models/OauthResultStatus.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.OauthResultStatus = void 0;
(function (OauthResultStatus) {
    OauthResultStatus["Pending"] = "pending";
    OauthResultStatus["Completed"] = "completed";
})(exports.OauthResultStatus || (exports.OauthResultStatus = {}));
function OauthResultStatusFromJSON(json) {
    return OauthResultStatusFromJSONTyped(json);
}
function OauthResultStatusFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function OauthResultStatusToJSON(value) {
    return value;
}

exports.OauthResultStatusFromJSON = OauthResultStatusFromJSON;
exports.OauthResultStatusFromJSONTyped = OauthResultStatusFromJSONTyped;
exports.OauthResultStatusToJSON = OauthResultStatusToJSON;
