// @dynamic-labs/sdk-api-core/src/models/ProviderEntryPointVersionEnum.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.ProviderEntryPointVersionEnum = void 0;
(function (ProviderEntryPointVersionEnum) {
    ProviderEntryPointVersionEnum["V6"] = "v6";
    ProviderEntryPointVersionEnum["V7"] = "v7";
})(exports.ProviderEntryPointVersionEnum || (exports.ProviderEntryPointVersionEnum = {}));
function ProviderEntryPointVersionEnumFromJSON(json) {
    return ProviderEntryPointVersionEnumFromJSONTyped(json);
}
function ProviderEntryPointVersionEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function ProviderEntryPointVersionEnumToJSON(value) {
    return value;
}

exports.ProviderEntryPointVersionEnumFromJSON = ProviderEntryPointVersionEnumFromJSON;
exports.ProviderEntryPointVersionEnumFromJSONTyped = ProviderEntryPointVersionEnumFromJSONTyped;
exports.ProviderEntryPointVersionEnumToJSON = ProviderEntryPointVersionEnumToJSON;
