// @dynamic-labs/sdk-api-core/src/models/ResidentKeyRequirement.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.ResidentKeyRequirement = void 0;
(function (ResidentKeyRequirement) {
    ResidentKeyRequirement["Discouraged"] = "discouraged";
    ResidentKeyRequirement["Preferred"] = "preferred";
    ResidentKeyRequirement["Required"] = "required";
})(exports.ResidentKeyRequirement || (exports.ResidentKeyRequirement = {}));
function ResidentKeyRequirementFromJSON(json) {
    return ResidentKeyRequirementFromJSONTyped(json);
}
function ResidentKeyRequirementFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function ResidentKeyRequirementToJSON(value) {
    return value;
}

exports.ResidentKeyRequirementFromJSON = ResidentKeyRequirementFromJSON;
exports.ResidentKeyRequirementFromJSONTyped = ResidentKeyRequirementFromJSONTyped;
exports.ResidentKeyRequirementToJSON = ResidentKeyRequirementToJSON;
