// @dynamic-labs/sdk-api-core/src/models/UnprocessableEntityErrorCode.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.UnprocessableEntityErrorCode = void 0;
(function (UnprocessableEntityErrorCode) {
    UnprocessableEntityErrorCode["EmailCannotBeNull"] = "email_cannot_be_null";
    UnprocessableEntityErrorCode["ExchangeCannotBeEnabled"] = "exchange_cannot_be_enabled";
    UnprocessableEntityErrorCode["TransferAmountTooSmall"] = "transfer_amount_too_small";
    UnprocessableEntityErrorCode["TransferAddressNotWhitelisted"] = "transfer_address_not_whitelisted";
    UnprocessableEntityErrorCode["InvalidScopes"] = "invalid_scopes";
    UnprocessableEntityErrorCode["UnknownTransferError"] = "unknown_transfer_error";
    UnprocessableEntityErrorCode["TransferMfaRequired"] = "transfer_mfa_required";
    UnprocessableEntityErrorCode["TransferMfaFailed"] = "transfer_mfa_failed";
    UnprocessableEntityErrorCode["InvalidTransferFunds"] = "invalid_transfer_funds";
    UnprocessableEntityErrorCode["InvalidTransferCurrency"] = "invalid_transfer_currency";
    UnprocessableEntityErrorCode["InvalidExchangeProvider"] = "invalid_exchange_provider";
    UnprocessableEntityErrorCode["InvalidTransferNetwork"] = "invalid_transfer_network";
    UnprocessableEntityErrorCode["InvalidEmail"] = "invalid_email";
    UnprocessableEntityErrorCode["EmailAlreadyExists"] = "email_already_exists";
    UnprocessableEntityErrorCode["AllowlistAlreadyExists"] = "allowlist_already_exists";
    UnprocessableEntityErrorCode["AllowlistEntryAlreadyExists"] = "allowlist_entry_already_exists";
    UnprocessableEntityErrorCode["ReassignWalletError"] = "reassign_wallet_error";
    UnprocessableEntityErrorCode["ReassignWalletConfirm"] = "reassign_wallet_confirm";
    UnprocessableEntityErrorCode["MembersCannotDeleteThemself"] = "members_cannot_delete_themself";
    UnprocessableEntityErrorCode["UsernameAlreadyExists"] = "username_already_exists";
    UnprocessableEntityErrorCode["WrongEmailVerificationToken"] = "wrong_email_verification_token";
    UnprocessableEntityErrorCode["WrongSmsVerificationToken"] = "wrong_sms_verification_token";
    UnprocessableEntityErrorCode["InvalidEmailVerification"] = "invalid_email_verification";
    UnprocessableEntityErrorCode["InvalidSmsVerification"] = "invalid_sms_verification";
    UnprocessableEntityErrorCode["InvalidVerification"] = "invalid_verification";
    UnprocessableEntityErrorCode["InvalidPosition"] = "invalid_position";
    UnprocessableEntityErrorCode["TooManySmsVerificationAttempts"] = "too_many_sms_verification_attempts";
    UnprocessableEntityErrorCode["TooManyEmailVerificationAttempts"] = "too_many_email_verification_attempts";
    UnprocessableEntityErrorCode["TooManyVerificationAttempts"] = "too_many_verification_attempts";
    UnprocessableEntityErrorCode["OrganizationNameAlreadyExists"] = "organization_name_already_exists";
    UnprocessableEntityErrorCode["ProjectNameAlreadyExists"] = "project_name_already_exists";
    UnprocessableEntityErrorCode["WalletNotDeployed"] = "wallet_not_deployed";
    UnprocessableEntityErrorCode["EmailVerificationRequired"] = "email_verification_required";
    UnprocessableEntityErrorCode["PhoneVerificationRequired"] = "phone_verification_required";
    UnprocessableEntityErrorCode["InviteAddressRequired"] = "invite_address_required";
    UnprocessableEntityErrorCode["ProviderNotAvailable"] = "provider_not_available";
    UnprocessableEntityErrorCode["ForbiddenUnlinkRequest"] = "forbidden_unlink_request";
    UnprocessableEntityErrorCode["InvalidUnlinkRequest"] = "invalid_unlink_request";
    UnprocessableEntityErrorCode["TooManyApiTokens"] = "too_many_api_tokens";
    UnprocessableEntityErrorCode["LockTimeout"] = "lock_timeout";
    UnprocessableEntityErrorCode["LockTooManyAttempts"] = "lock_too_many_attempts";
    UnprocessableEntityErrorCode["NftTokenGatingNotSupportedForChain"] = "nft_token_gating_not_supported_for_chain";
    UnprocessableEntityErrorCode["EmptyChainName"] = "empty_chain_name";
    UnprocessableEntityErrorCode["NoEnabledEmailProvider"] = "no_enabled_email_provider";
    UnprocessableEntityErrorCode["NoEnabledSmsProvider"] = "no_enabled_sms_provider";
    UnprocessableEntityErrorCode["InvalidKeyExportUrl"] = "invalid_key_export_url";
    UnprocessableEntityErrorCode["InvalidDynamicProps"] = "invalid_dynamic_props";
    UnprocessableEntityErrorCode["TooManyRequests"] = "too_many_requests";
    UnprocessableEntityErrorCode["TooManyOrganizationsForUser"] = "too_many_organizations_for_user";
    UnprocessableEntityErrorCode["TooManyProjectsForOrganization"] = "too_many_projects_for_organization";
    UnprocessableEntityErrorCode["EmailAssociatedWithDifferentProvider"] = "email_associated_with_different_provider";
    UnprocessableEntityErrorCode["UserHasAlreadyAccountWithEmail"] = "user_has_already_account_with_email";
    UnprocessableEntityErrorCode["UserHasAlreadyAccountWithPhoneNumber"] = "user_has_already_account_with_phone_number";
    UnprocessableEntityErrorCode["OtherVerifyFailure"] = "other_verify_failure";
    UnprocessableEntityErrorCode["EmailTiedToEmbeddedWallet"] = "email_tied_to_embedded_wallet";
    UnprocessableEntityErrorCode["InvalidInvite"] = "invalid_invite";
    UnprocessableEntityErrorCode["SocialAccountAlreadyExists"] = "social_account_already_exists";
    UnprocessableEntityErrorCode["InvalidEmailAddress"] = "invalid_email_address";
    UnprocessableEntityErrorCode["InvalidGate"] = "invalid_gate";
    UnprocessableEntityErrorCode["ConflictingEmbeddedWalletProviders"] = "conflicting_embedded_wallet_providers";
    UnprocessableEntityErrorCode["InvalidUser"] = "invalid_user";
    UnprocessableEntityErrorCode["InvalidCorsOrigins"] = "invalid_cors_origins";
    UnprocessableEntityErrorCode["InvalidMobileDeeplinkUrls"] = "invalid_mobile_deeplink_urls";
    UnprocessableEntityErrorCode["UnauthorizedMobileDeeplinkUrl"] = "unauthorized_mobile_deeplink_url";
    UnprocessableEntityErrorCode["InvalidPrivateKeyFormat"] = "invalid_private_key_format";
    UnprocessableEntityErrorCode["InvalidEmbeddedWalletSettings"] = "invalid_embedded_wallet_settings";
    UnprocessableEntityErrorCode["WalletNotSupportPasskey"] = "wallet_not_support_passkey";
    UnprocessableEntityErrorCode["RecoveryEmailUnavailableOrInvalid"] = "recovery_email_unavailable_or_invalid";
    UnprocessableEntityErrorCode["ConnectError"] = "connect_error";
    UnprocessableEntityErrorCode["InvalidWalletName"] = "invalid_wallet_name";
    UnprocessableEntityErrorCode["InvalidWalletAddress"] = "invalid_wallet_address";
    UnprocessableEntityErrorCode["EmailRecoveryDisabled"] = "email_recovery_disabled";
    UnprocessableEntityErrorCode["NoCompatibleWalletServiceEnabled"] = "no_compatible_wallet_service_enabled";
    UnprocessableEntityErrorCode["MissingAaProjectId"] = "missing_aa_project_id";
    UnprocessableEntityErrorCode["LinkedEmbeddedWallet"] = "linked_embedded_wallet";
    UnprocessableEntityErrorCode["InvalidMpcEnvironment"] = "invalid_mpc_environment";
    UnprocessableEntityErrorCode["InvalidRegex"] = "invalid_regex";
    UnprocessableEntityErrorCode["RepeatedOptions"] = "repeated_options";
    UnprocessableEntityErrorCode["MissingPhoneNumberOrEmail"] = "missing_phone_number_or_email";
    UnprocessableEntityErrorCode["DuplicateExists"] = "duplicate_exists";
    UnprocessableEntityErrorCode["MfaDeviceNotFound"] = "mfa_device_not_found";
    UnprocessableEntityErrorCode["MfaInvalidCode"] = "mfa_invalid_code";
    UnprocessableEntityErrorCode["MfaInvalidRequest"] = "mfa_invalid_request";
    UnprocessableEntityErrorCode["InvalidPhone"] = "invalid_phone";
    UnprocessableEntityErrorCode["GateExists"] = "gate_exists";
    UnprocessableEntityErrorCode["InvalidEmbeddedWalletChainsConfiguration"] = "invalid_embedded_wallet_chains_configuration";
    UnprocessableEntityErrorCode["MergeAccountsConfirmation"] = "merge_accounts_confirmation";
    UnprocessableEntityErrorCode["MergeAccountsInvalid"] = "merge_accounts_invalid";
    UnprocessableEntityErrorCode["InvalidWalletId"] = "invalid_wallet_id";
    UnprocessableEntityErrorCode["UnableToFetchBalances"] = "unable_to_fetch_balances";
    UnprocessableEntityErrorCode["CannotSwitchToUnique"] = "cannot_switch_to_unique";
    UnprocessableEntityErrorCode["CustomFieldDataNotUnique"] = "custom_field_data_not_unique";
    UnprocessableEntityErrorCode["MetadataSizeExceeded"] = "metadata_size_exceeded";
    UnprocessableEntityErrorCode["InvalidCustomHostname"] = "invalid_custom_hostname";
    UnprocessableEntityErrorCode["BlockedUserCannotTransferWallet"] = "blocked_user_cannot_transfer_wallet";
    UnprocessableEntityErrorCode["FeatureNotEnabledInSuborg"] = "feature_not_enabled_in_suborg";
    UnprocessableEntityErrorCode["InvalidUsername"] = "invalid_username";
    UnprocessableEntityErrorCode["InvalidExternalAuth"] = "invalid_external_auth";
    UnprocessableEntityErrorCode["InvalidChainAddressMatch"] = "invalid_chain_address_match";
    UnprocessableEntityErrorCode["InvalidUpdate"] = "invalid_update";
    UnprocessableEntityErrorCode["CannotDeleteLastProject"] = "cannot_delete_last_project";
    UnprocessableEntityErrorCode["NoHdWalletFound"] = "no_hd_wallet_found";
    UnprocessableEntityErrorCode["WalletAccountExistsForChain"] = "wallet_account_exists_for_chain";
    UnprocessableEntityErrorCode["TooManyEmbeddedWalletsForUser"] = "too_many_embedded_wallets_for_user";
    UnprocessableEntityErrorCode["TooManyEmbeddedWalletsPerChainForUser"] = "too_many_embedded_wallets_per_chain_for_user";
    UnprocessableEntityErrorCode["InvalidSessionPublicKey"] = "invalid_session_public_key";
    UnprocessableEntityErrorCode["CustomFieldNotUnique"] = "custom_field_not_unique";
    UnprocessableEntityErrorCode["InvalidAaProjectConfiguration"] = "invalid_aa_project_configuration";
    UnprocessableEntityErrorCode["NoV1EmbeddedWalletFound"] = "no_v1_embedded_wallet_found";
    UnprocessableEntityErrorCode["EmbeddedWalletAddressAlreadyLinked"] = "embedded_wallet_address_already_linked";
    UnprocessableEntityErrorCode["KeyShareAlreadyBackedUp"] = "key_share_already_backed_up";
    UnprocessableEntityErrorCode["InvalidKeySharesLength"] = "invalid_key_shares_length";
    UnprocessableEntityErrorCode["WalletServiceConnectionError"] = "wallet_service_connection_error";
    UnprocessableEntityErrorCode["NoGlobalWalletFound"] = "no_global_wallet_found";
    UnprocessableEntityErrorCode["InvalidMetadataValue"] = "invalid_metadata_value";
    UnprocessableEntityErrorCode["MaxPasskeysPerUser"] = "max_passkeys_per_user";
    UnprocessableEntityErrorCode["PasskeyNotFound"] = "passkey_not_found";
    UnprocessableEntityErrorCode["CannotDeletePasskeyLinkedToEmbeddedWallet"] = "cannot_delete_passkey_linked_to_embedded_wallet";
    UnprocessableEntityErrorCode["CannotDeleteLastSignInVerifiedCredential"] = "cannot_delete_last_sign_in_verified_credential";
})(exports.UnprocessableEntityErrorCode || (exports.UnprocessableEntityErrorCode = {}));
function UnprocessableEntityErrorCodeFromJSON(json) {
    return UnprocessableEntityErrorCodeFromJSONTyped(json);
}
function UnprocessableEntityErrorCodeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function UnprocessableEntityErrorCodeToJSON(value) {
    return value;
}

exports.UnprocessableEntityErrorCodeFromJSON = UnprocessableEntityErrorCodeFromJSON;
exports.UnprocessableEntityErrorCodeFromJSONTyped = UnprocessableEntityErrorCodeFromJSONTyped;
exports.UnprocessableEntityErrorCodeToJSON = UnprocessableEntityErrorCodeToJSON;
