// @dynamic-labs/sdk-api-core/src/models/UserIdentifierTypeEnum.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.UserIdentifierTypeEnum = void 0;
(function (UserIdentifierTypeEnum) {
    UserIdentifierTypeEnum["Email"] = "email";
    UserIdentifierTypeEnum["Id"] = "id";
    UserIdentifierTypeEnum["ExternalUserId"] = "externalUserId";
    UserIdentifierTypeEnum["PhoneNumber"] = "phoneNumber";
    UserIdentifierTypeEnum["SocialUsername"] = "socialUsername";
    UserIdentifierTypeEnum["SocialAccountId"] = "socialAccountId";
})(exports.UserIdentifierTypeEnum || (exports.UserIdentifierTypeEnum = {}));
function UserIdentifierTypeEnumFromJSON(json) {
    return UserIdentifierTypeEnumFromJSONTyped(json);
}
function UserIdentifierTypeEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function UserIdentifierTypeEnumToJSON(value) {
    return value;
}

exports.UserIdentifierTypeEnumFromJSON = UserIdentifierTypeEnumFromJSON;
exports.UserIdentifierTypeEnumFromJSONTyped = UserIdentifierTypeEnumFromJSONTyped;
exports.UserIdentifierTypeEnumToJSON = UserIdentifierTypeEnumToJSON;
