// @dynamic-labs/sdk-api-core/src/models/WaasBackupOptionsEnum.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
var WaasBackupOptionsEnum;
(function (WaasBackupOptionsEnum) {
    WaasBackupOptionsEnum["GoogleDrive"] = "googleDrive";
    WaasBackupOptionsEnum["Icloud"] = "icloud";
    WaasBackupOptionsEnum["Dynamic"] = "dynamic";
    WaasBackupOptionsEnum["External"] = "external";
    WaasBackupOptionsEnum["Delegated"] = "delegated";
})(WaasBackupOptionsEnum || (WaasBackupOptionsEnum = {}));
function WaasBackupOptionsEnumFromJSON(json) {
    return WaasBackupOptionsEnumFromJSONTyped(json);
}
function WaasBackupOptionsEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function WaasBackupOptionsEnumToJSON(value) {
    return value;
}

export { WaasBackupOptionsEnum, WaasBackupOptionsEnumFromJSON, WaasBackupOptionsEnumFromJSONTyped, WaasBackupOptionsEnumToJSON };
