// @dynamic-labs/sdk-api-core/src/models/WaasChainEnum.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.WaasChainEnum = void 0;
(function (WaasChainEnum) {
    WaasChainEnum["Evm"] = "EVM";
    WaasChainEnum["Svm"] = "SVM";
    WaasChainEnum["Sui"] = "SUI";
})(exports.WaasChainEnum || (exports.WaasChainEnum = {}));
function WaasChainEnumFromJSON(json) {
    return WaasChainEnumFromJSONTyped(json);
}
function WaasChainEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function WaasChainEnumToJSON(value) {
    return value;
}

exports.WaasChainEnumFromJSON = WaasChainEnumFromJSON;
exports.WaasChainEnumFromJSONTyped = WaasChainEnumFromJSONTyped;
exports.WaasChainEnumToJSON = WaasChainEnumToJSON;
