// @dynamic-labs/sdk-api-core/src/models/WaasDelegatedAccessEncryptionPublicKeyType.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.WaasDelegatedAccessEncryptionPublicKeyType = void 0;
(function (WaasDelegatedAccessEncryptionPublicKeyType) {
    WaasDelegatedAccessEncryptionPublicKeyType["Dynamic"] = "dynamic";
    WaasDelegatedAccessEncryptionPublicKeyType["External"] = "external";
})(exports.WaasDelegatedAccessEncryptionPublicKeyType || (exports.WaasDelegatedAccessEncryptionPublicKeyType = {}));
function WaasDelegatedAccessEncryptionPublicKeyTypeFromJSON(json) {
    return WaasDelegatedAccessEncryptionPublicKeyTypeFromJSONTyped(json);
}
function WaasDelegatedAccessEncryptionPublicKeyTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function WaasDelegatedAccessEncryptionPublicKeyTypeToJSON(value) {
    return value;
}

exports.WaasDelegatedAccessEncryptionPublicKeyTypeFromJSON = WaasDelegatedAccessEncryptionPublicKeyTypeFromJSON;
exports.WaasDelegatedAccessEncryptionPublicKeyTypeFromJSONTyped = WaasDelegatedAccessEncryptionPublicKeyTypeFromJSONTyped;
exports.WaasDelegatedAccessEncryptionPublicKeyTypeToJSON = WaasDelegatedAccessEncryptionPublicKeyTypeToJSON;
