// @dynamic-labs/sdk-react-core/src/lib/context/DynamicContext/useDynamicContext/useDynamicContext.js
'use client'
import { useContext, useCallback, useRef } from 'react';
import { DynamicContext } from '../DynamicContext.js';
import { useLoadingAndLifecycle } from '../../../store/state/loadingAndLifecycle/loadingAndLifecycle.js';
import { useOverrides } from './useOverrides/useOverrides.js';

// This is external context hook which is exposed to customers
const useDynamicContext = () => {
    const context = useContext(DynamicContext);
    if (context === undefined) {
        throw new Error('useDynamicContext must be used within a DynamicContextProvider');
    }
    const { accountSwitchState, awaitingSignatureState, getNameService, handleLogOut, removeWallet, loadingNetwork, locale, loginWithEmail, multiWalletWidgetState, network, networkConfigurations, qrcodeUri, refetchProjectSettings, sdkHasLoaded, setLogInWithEmail, setMultiWalletWidgetState, setShowAuthFlow, setShowDynamicUserProfile, setShowQrcodeModal, shadowDOMEnabled, showAuthFlow, showQrcodeModal, user, userWithMissingInfo, bridgeChains, bridgeChainsToConnect, primaryWallet, setShowBridgeWidget, sendWagmiSettings, showDynamicUserProfile, selectedTabIndex, setSelectedTabIndex, } = context;
    const { authMode, setAuthMode } = useOverrides(context);
    const externalSetShowAuthFlow = useCallback((value) => {
        setShowAuthFlow(value, {
            clearErrors: true,
            emitCancelAuth: true,
            ignoreIfIsEmbeddedWidget: true,
            initializeWalletConnect: true,
            performMultiWalletChecks: true,
        });
    }, [setShowAuthFlow]);
    const { initialWalletVerificationInProgress } = useLoadingAndLifecycle();
    // This ref ensures we don't return a value for primary wallet until useConnectAndSign has completely finished running
    // Read UserWalletsContext.tsx explanation for more context
    const verifiedPrimaryWallet = useRef(null);
    if (!initialWalletVerificationInProgress || !primaryWallet) {
        verifiedPrimaryWallet.current = primaryWallet;
    }
    return {
        accountSwitchState,
        authMode,
        awaitingSignatureState,
        bridgeChains,
        bridgeChainsToConnect,
        getNameService,
        handleLogOut,
        handleUnlinkWallet: removeWallet,
        loadingNetwork,
        locale,
        loginWithEmail,
        multiWalletWidgetState,
        network,
        networkConfigurations,
        primaryWallet: verifiedPrimaryWallet.current,
        qrcodeUri,
        refetchProjectSettings,
        removeWallet,
        sdkHasLoaded,
        selectedTabIndex,
        sendWagmiSettings,
        setAuthMode,
        setLogInWithEmail,
        setMultiWalletWidgetState,
        setSelectedTabIndex,
        setShowAuthFlow: externalSetShowAuthFlow,
        setShowBridgeWidget,
        setShowDynamicUserProfile,
        setShowQrcodeModal,
        shadowDOMEnabled,
        showAuthFlow,
        showDynamicUserProfile,
        showQrcodeModal,
        user,
        userWithMissingInfo,
    };
};

export { useDynamicContext };
