// @dynamic-labs/wallet-connector-core/src/utils/getWalletMetadataFromWalletBook/getWalletMetadataFromWalletBook.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var walletBook = require('@dynamic-labs/wallet-book');
var logger = require('../logger.cjs');
var getDeepLinks = require('./getDeepLinks/getDeepLinks.cjs');
var getDownloadLinks = require('./getDownloadLinks/getDownloadLinks.cjs');
var getIconUrl = require('./getIconUrl/getIconUrl.cjs');
var getValidHexColor = require('./getValidHexColor/getValidHexColor.cjs');
var getWalletLimitations = require('./getWalletLimitations/getWalletLimitations.cjs');

const getWalletMetadataFromWalletBook = ({ walletKey, walletBookWallet, walletBook: walletBook$1, walletFallback, }) => {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j;
    if (!walletBook.isWalletBookPopulated(walletBook$1) && !walletFallback) {
        logger.logger.warn('Wallet book is required');
        return;
    }
    try {
        const wallet = walletBookWallet !== null && walletBookWallet !== void 0 ? walletBookWallet : walletBook.getWalletBookWallet(walletBook$1, walletKey, walletFallback);
        // if no walletBookWallet is provided and no wallet is found in the wallet book, throw an error
        if (!wallet) {
            logger.logger.warn('Wallet not found in wallet book');
            return;
        }
        return {
            brandColor: getValidHexColor.getValidHexColor((_a = wallet === null || wallet === void 0 ? void 0 : wallet.brand) === null || _a === void 0 ? void 0 : _a.primaryColor),
            deepLinks: getDeepLinks.getDeepLinks({
                desktop: wallet === null || wallet === void 0 ? void 0 : wallet.desktop,
                mobile: wallet === null || wallet === void 0 ? void 0 : wallet.mobile,
            }),
            downloadLinks: getDownloadLinks.getDownloadLinks(wallet),
            groupKey: (_b = wallet === null || wallet === void 0 ? void 0 : wallet.chainGroup) !== null && _b !== void 0 ? _b : wallet === null || wallet === void 0 ? void 0 : wallet.walletGroup,
            icon: (_d = getIconUrl.getIconUrl((_c = wallet === null || wallet === void 0 ? void 0 : wallet.brand) === null || _c === void 0 ? void 0 : _c.spriteId)) !== null && _d !== void 0 ? _d : '',
            id: walletKey,
            inAppBrowserUrl: (_h = (_f = (_e = wallet.mobile) === null || _e === void 0 ? void 0 : _e.inAppBrowserV2) !== null && _f !== void 0 ? _f : (_g = wallet.mobile) === null || _g === void 0 ? void 0 : _g.inAppBrowser) !== null && _h !== void 0 ? _h : undefined,
            name: wallet.name,
            rdns: (_j = wallet.eip6963Config) === null || _j === void 0 ? void 0 : _j.rdns,
            supportedHardwareWallets: wallet.hardwareWallets,
            walletLimitations: getWalletLimitations.getWalletLimitations(wallet.walletLimitations),
        };
    }
    catch (e) {
        logger.logger.warn('Error getting wallet metadata from wallet book', {
            error: e,
        });
        return;
    }
};

exports.getWalletMetadataFromWalletBook = getWalletMetadataFromWalletBook;
