// @dynamic-labs/sdk-api-core/src/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var runtime = require('./runtime.cjs');
var SDKApi = require('./apis/SDKApi.cjs');
var Account = require('./models/Account.cjs');
var AccountBalances = require('./models/AccountBalances.cjs');
var Asset = require('./models/Asset.cjs');
var AssetDiff = require('./models/AssetDiff.cjs');
var AssetExposure = require('./models/AssetExposure.cjs');
var AssetExposureSpenderToInfo = require('./models/AssetExposureSpenderToInfo.cjs');
var AssetTransfer = require('./models/AssetTransfer.cjs');
var AttestationConveyancePreference = require('./models/AttestationConveyancePreference.cjs');
var AuthModeEnum = require('./models/AuthModeEnum.cjs');
var AuthSettings = require('./models/AuthSettings.cjs');
var AuthStorageEnum = require('./models/AuthStorageEnum.cjs');
var AuthenticationExtensionsClientInputs = require('./models/AuthenticationExtensionsClientInputs.cjs');
var AuthenticationExtensionsClientOutputs = require('./models/AuthenticationExtensionsClientOutputs.cjs');
var AuthenticatorAssertionResponse = require('./models/AuthenticatorAssertionResponse.cjs');
var AuthenticatorAttachment = require('./models/AuthenticatorAttachment.cjs');
var AuthenticatorAttestationResponse = require('./models/AuthenticatorAttestationResponse.cjs');
var AuthenticatorSelectionCriteria = require('./models/AuthenticatorSelectionCriteria.cjs');
var AuthenticatorTransportProtocol = require('./models/AuthenticatorTransportProtocol.cjs');
var BackupKeySharesToLocationRequest = require('./models/BackupKeySharesToLocationRequest.cjs');
var BackupKeySharesToLocationResponse = require('./models/BackupKeySharesToLocationResponse.cjs');
var BackupKeySharesToLocationsRequest = require('./models/BackupKeySharesToLocationsRequest.cjs');
var BackupKeySharesToLocationsRequestLocations = require('./models/BackupKeySharesToLocationsRequestLocations.cjs');
var BackupKeySharesToLocationsResponse = require('./models/BackupKeySharesToLocationsResponse.cjs');
var BackupKeySharesToLocationsResponseLocationsWithKeyShares = require('./models/BackupKeySharesToLocationsResponseLocationsWithKeyShares.cjs');
var BackupMultipleClientKeySharesRequest = require('./models/BackupMultipleClientKeySharesRequest.cjs');
var BackupMultipleClientKeySharesResponse = require('./models/BackupMultipleClientKeySharesResponse.cjs');
var BadGateway = require('./models/BadGateway.cjs');
var BadRequest = require('./models/BadRequest.cjs');
var BaseUser = require('./models/BaseUser.cjs');
var BlockaidValidation = require('./models/BlockaidValidation.cjs');
var ChainConfiguration = require('./models/ChainConfiguration.cjs');
var ChainEnum = require('./models/ChainEnum.cjs');
var CoinbaseMpcWalletProperties = require('./models/CoinbaseMpcWalletProperties.cjs');
var CompletePasskeyRecoveryRequest = require('./models/CompletePasskeyRecoveryRequest.cjs');
var ConnectRequest = require('./models/ConnectRequest.cjs');
var CreateEmbeddedWalletParams = require('./models/CreateEmbeddedWalletParams.cjs');
var CreateEmbeddedWalletSpecificOpts = require('./models/CreateEmbeddedWalletSpecificOpts.cjs');
var CreateEmbeddedWalletsRequest = require('./models/CreateEmbeddedWalletsRequest.cjs');
var CreateExchangeTransferRequest = require('./models/CreateExchangeTransferRequest.cjs');
var CreateExchangeTransferRequestNetworkObject = require('./models/CreateExchangeTransferRequestNetworkObject.cjs');
var CreateMfaToken = require('./models/CreateMfaToken.cjs');
var CreateTurnkeyEmbeddedWalletSpecificOpts = require('./models/CreateTurnkeyEmbeddedWalletSpecificOpts.cjs');
var CreateUserEmbeddedWalletsFromFarcasterRequest = require('./models/CreateUserEmbeddedWalletsFromFarcasterRequest.cjs');
var CreateUserEmbeddedWalletsRequest = require('./models/CreateUserEmbeddedWalletsRequest.cjs');
var CreateWaasAccountRequest = require('./models/CreateWaasAccountRequest.cjs');
var CreateWalletAccountRequest = require('./models/CreateWalletAccountRequest.cjs');
var Currency = require('./models/Currency.cjs');
var CurrencyType = require('./models/CurrencyType.cjs');
var CustomFieldType = require('./models/CustomFieldType.cjs');
var CustomFieldValidValue = require('./models/CustomFieldValidValue.cjs');
var CustomFieldValidationRules = require('./models/CustomFieldValidationRules.cjs');
var DeleteEmbeddedWalletsRequest = require('./models/DeleteEmbeddedWalletsRequest.cjs');
var DeleteUserPasskeyRequest = require('./models/DeleteUserPasskeyRequest.cjs');
var Duration = require('./models/Duration.cjs');
var DynamicJwt = require('./models/DynamicJwt.cjs');
var EcdsaValidatorOptions = require('./models/EcdsaValidatorOptions.cjs');
var Eip712Domain = require('./models/Eip712Domain.cjs');
var Eip712TypeMember = require('./models/Eip712TypeMember.cjs');
var EmailProviderResponse = require('./models/EmailProviderResponse.cjs');
var EmailVerificationCreateRequest = require('./models/EmailVerificationCreateRequest.cjs');
var EmailVerificationCreateResponse = require('./models/EmailVerificationCreateResponse.cjs');
var EmailVerificationMfaRequest = require('./models/EmailVerificationMfaRequest.cjs');
var EmailVerificationRetryRequest = require('./models/EmailVerificationRetryRequest.cjs');
var EmailVerificationVerifyRequest = require('./models/EmailVerificationVerifyRequest.cjs');
var EmbeddedWalletAuthToken = require('./models/EmbeddedWalletAuthToken.cjs');
var EmbeddedWalletAuthType = require('./models/EmbeddedWalletAuthType.cjs');
var EmbeddedWalletChainEnum = require('./models/EmbeddedWalletChainEnum.cjs');
var EmbeddedWalletPasscodeClaimRequest = require('./models/EmbeddedWalletPasscodeClaimRequest.cjs');
var EmbeddedWalletProviderEnum = require('./models/EmbeddedWalletProviderEnum.cjs');
var EmbeddedWalletSecret = require('./models/EmbeddedWalletSecret.cjs');
var EmbeddedWalletSecretWithUpdatedJwt = require('./models/EmbeddedWalletSecretWithUpdatedJwt.cjs');
var EmbeddedWalletSecretWithUpdatedJwtAllOf = require('./models/EmbeddedWalletSecretWithUpdatedJwtAllOf.cjs');
var EmbeddedWalletSecurityMethod = require('./models/EmbeddedWalletSecurityMethod.cjs');
var EmbeddedWalletVersionEnum = require('./models/EmbeddedWalletVersionEnum.cjs');
var EnvironmentEnum = require('./models/EnvironmentEnum.cjs');
var ErrorMessageWithCode = require('./models/ErrorMessageWithCode.cjs');
var Exchange = require('./models/Exchange.cjs');
var ExchangeKeyEnum = require('./models/ExchangeKeyEnum.cjs');
var ExchangeOption = require('./models/ExchangeOption.cjs');
var ExchangeTransaction = require('./models/ExchangeTransaction.cjs');
var ExchangeTransferResponse = require('./models/ExchangeTransferResponse.cjs');
var ExportEmbeddedWalletResponse = require('./models/ExportEmbeddedWalletResponse.cjs');
var ExportWaasWalletPrivateKeyRequest = require('./models/ExportWaasWalletPrivateKeyRequest.cjs');
var ExternalAuth = require('./models/ExternalAuth.cjs');
var ExternalAuthSigninRequest = require('./models/ExternalAuthSigninRequest.cjs');
var ExternalWalletFundingDefaultChain = require('./models/ExternalWalletFundingDefaultChain.cjs');
var ExternalWalletFundingDefaultSettings = require('./models/ExternalWalletFundingDefaultSettings.cjs');
var ExternalWalletFundingDefaultSettingsToken = require('./models/ExternalWalletFundingDefaultSettingsToken.cjs');
var ExternalWalletFundingTokenRule = require('./models/ExternalWalletFundingTokenRule.cjs');
var FarcasterSignInRequest = require('./models/FarcasterSignInRequest.cjs');
var FeatureFlags = require('./models/FeatureFlags.cjs');
var Forbidden = require('./models/Forbidden.cjs');
var ForbiddenErrorPayload = require('./models/ForbiddenErrorPayload.cjs');
var ForbiddenWithErrorAndPayload = require('./models/ForbiddenWithErrorAndPayload.cjs');
var FrameworkSettings = require('./models/FrameworkSettings.cjs');
var Funding = require('./models/Funding.cjs');
var FundingExternalWallets = require('./models/FundingExternalWallets.cjs');
var FundingExternalWalletsDefaultSettings = require('./models/FundingExternalWalletsDefaultSettings.cjs');
var FundingExternalWalletsMinAmount = require('./models/FundingExternalWalletsMinAmount.cjs');
var GetPasskeyAuthenticationOptionsResponse = require('./models/GetPasskeyAuthenticationOptionsResponse.cjs');
var GetPasskeyRegistrationOptionsResponse = require('./models/GetPasskeyRegistrationOptionsResponse.cjs');
var GetUserPasskeysResponse = require('./models/GetUserPasskeysResponse.cjs');
var GlobalWalletAccessControlTypeEnum = require('./models/GlobalWalletAccessControlTypeEnum.cjs');
var GlobalWalletConnection = require('./models/GlobalWalletConnection.cjs');
var GlobalWalletConnectionCreateRequest = require('./models/GlobalWalletConnectionCreateRequest.cjs');
var GlobalWalletConnectionStatusEnum = require('./models/GlobalWalletConnectionStatusEnum.cjs');
var GlobalWalletSettings = require('./models/GlobalWalletSettings.cjs');
var GlobalWalletSettingsCustomMenuLinks = require('./models/GlobalWalletSettingsCustomMenuLinks.cjs');
var HCaptchaSettings = require('./models/HCaptchaSettings.cjs');
var HardwareWalletEnum = require('./models/HardwareWalletEnum.cjs');
var HardwareWalletProperties = require('./models/HardwareWalletProperties.cjs');
var HealthcheckResponse = require('./models/HealthcheckResponse.cjs');
var HealthcheckStatus = require('./models/HealthcheckStatus.cjs');
var ImportWaasPrivateKeyRequest = require('./models/ImportWaasPrivateKeyRequest.cjs');
var InitEmailAuthRequest = require('./models/InitEmailAuthRequest.cjs');
var InitEmailAuthResponse = require('./models/InitEmailAuthResponse.cjs');
var InitPasskeyRecoveryRequest = require('./models/InitPasskeyRecoveryRequest.cjs');
var InitPasskeyRecoveryResponse = require('./models/InitPasskeyRecoveryResponse.cjs');
var IntegrationSetting = require('./models/IntegrationSetting.cjs');
var InternalServerError = require('./models/InternalServerError.cjs');
var JwksKey = require('./models/JwksKey.cjs');
var JwksResponse = require('./models/JwksResponse.cjs');
var JwtBlockchainAccount = require('./models/JwtBlockchainAccount.cjs');
var JwtPayloadDeprecatedInfo = require('./models/JwtPayloadDeprecatedInfo.cjs');
var JwtVerifiedCredential = require('./models/JwtVerifiedCredential.cjs');
var JwtVerifiedCredentialFormatEnum = require('./models/JwtVerifiedCredentialFormatEnum.cjs');
var JwtVerifiedCredentialHashes = require('./models/JwtVerifiedCredentialHashes.cjs');
var KycFieldType = require('./models/KycFieldType.cjs');
var MFAAction = require('./models/MFAAction.cjs');
var MFAAuthPasskeyDeviceGetResponse = require('./models/MFAAuthPasskeyDeviceGetResponse.cjs');
var MFAAuthPasskeyDeviceGetResponseAssertion = require('./models/MFAAuthPasskeyDeviceGetResponseAssertion.cjs');
var MFAAuthPasskeyDevicePostRequest = require('./models/MFAAuthPasskeyDevicePostRequest.cjs');
var MFAAuthRecoveryDevicePostRequest = require('./models/MFAAuthRecoveryDevicePostRequest.cjs');
var MFAAuthTotpDevicePostRequest = require('./models/MFAAuthTotpDevicePostRequest.cjs');
var MFADevice = require('./models/MFADevice.cjs');
var MFADeviceType = require('./models/MFADeviceType.cjs');
var MFAGetRecoveryCodesResponse = require('./models/MFAGetRecoveryCodesResponse.cjs');
var MFAListDevicesResponse = require('./models/MFAListDevicesResponse.cjs');
var MFAMethodsResponse = require('./models/MFAMethodsResponse.cjs');
var MFARegenRecoveryCodesResponse = require('./models/MFARegenRecoveryCodesResponse.cjs');
var MFARegisterPasskeyDeviceGetResponse = require('./models/MFARegisterPasskeyDeviceGetResponse.cjs');
var MFARegisterPasskeyDeviceGetResponseRegistration = require('./models/MFARegisterPasskeyDeviceGetResponseRegistration.cjs');
var MFARegisterPasskeyDevicePostRequest = require('./models/MFARegisterPasskeyDevicePostRequest.cjs');
var MFARegisterTotpDeviceGetResponse = require('./models/MFARegisterTotpDeviceGetResponse.cjs');
var MFARegisterTotpDevicePostRequest = require('./models/MFARegisterTotpDevicePostRequest.cjs');
var MFASettings = require('./models/MFASettings.cjs');
var MFASettingsActions = require('./models/MFASettingsActions.cjs');
var MFASettingsMethods = require('./models/MFASettingsMethods.cjs');
var MFAUpdateDeviceRequest = require('./models/MFAUpdateDeviceRequest.cjs');
var MergeConflicts = require('./models/MergeConflicts.cjs');
var MergeUser = require('./models/MergeUser.cjs');
var MergeUserConflict = require('./models/MergeUserConflict.cjs');
var MergeUserConflictResolution = require('./models/MergeUserConflictResolution.cjs');
var MergeUserConflictResolutions = require('./models/MergeUserConflictResolutions.cjs');
var MethodNotAllowed = require('./models/MethodNotAllowed.cjs');
var MfaBackupCodeAcknowledgement = require('./models/MfaBackupCodeAcknowledgement.cjs');
var MinifiedDynamicJwt = require('./models/MinifiedDynamicJwt.cjs');
var MobileSettings = require('./models/MobileSettings.cjs');
var MultichainAccountBalanceResponse = require('./models/MultichainAccountBalanceResponse.cjs');
var MultichainAccountBalanceResponseChainBalances = require('./models/MultichainAccountBalanceResponseChainBalances.cjs');
var MultichainAccountBalanceResponseNetworks = require('./models/MultichainAccountBalanceResponseNetworks.cjs');
var MultichainAccountBalancesRequest = require('./models/MultichainAccountBalancesRequest.cjs');
var MultichainAccountBalancesRequestBalanceRequests = require('./models/MultichainAccountBalancesRequestBalanceRequests.cjs');
var NameService = require('./models/NameService.cjs');
var NameServiceData = require('./models/NameServiceData.cjs');
var NameServiceSdkSettings = require('./models/NameServiceSdkSettings.cjs');
var NameServiceSdkSettingsEvm = require('./models/NameServiceSdkSettingsEvm.cjs');
var NativeCurrency = require('./models/NativeCurrency.cjs');
var Network = require('./models/Network.cjs');
var NetworkConfiguration = require('./models/NetworkConfiguration.cjs');
var NetworkConfigurationResponse = require('./models/NetworkConfigurationResponse.cjs');
var NextJsSettings = require('./models/NextJsSettings.cjs');
var NextViewEnum = require('./models/NextViewEnum.cjs');
var NonceResponse = require('./models/NonceResponse.cjs');
var NotFound = require('./models/NotFound.cjs');
var OAuthAccount = require('./models/OAuthAccount.cjs');
var OauthInitAuthRequest = require('./models/OauthInitAuthRequest.cjs');
var OauthProviderLoginUrl = require('./models/OauthProviderLoginUrl.cjs');
var OauthProviderRequest = require('./models/OauthProviderRequest.cjs');
var OauthRequest = require('./models/OauthRequest.cjs');
var OauthResultRequest = require('./models/OauthResultRequest.cjs');
var OauthResultResponse = require('./models/OauthResultResponse.cjs');
var OauthResultStatus = require('./models/OauthResultStatus.cjs');
var OnrampProviders = require('./models/OnrampProviders.cjs');
var OpenRoomResponse = require('./models/OpenRoomResponse.cjs');
var OpenRoomResponseForReshare = require('./models/OpenRoomResponseForReshare.cjs');
var OpenRoomResponseWithServerKeygenIds = require('./models/OpenRoomResponseWithServerKeygenIds.cjs');
var PasskeyAllowCredential = require('./models/PasskeyAllowCredential.cjs');
var PasskeyAuthRequest = require('./models/PasskeyAuthRequest.cjs');
var PasskeyCredentialHint = require('./models/PasskeyCredentialHint.cjs');
var PasskeyCredentialType = require('./models/PasskeyCredentialType.cjs');
var PasskeyExcludeCredential = require('./models/PasskeyExcludeCredential.cjs');
var PasskeyExtension = require('./models/PasskeyExtension.cjs');
var PasskeyRegisterRequest = require('./models/PasskeyRegisterRequest.cjs');
var PasskeyRegisterVerifyRequest = require('./models/PasskeyRegisterVerifyRequest.cjs');
var PasskeyRegistrationCredential = require('./models/PasskeyRegistrationCredential.cjs');
var PasskeyRegistrationCredentialV2 = require('./models/PasskeyRegistrationCredentialV2.cjs');
var PasskeyRelyingParty = require('./models/PasskeyRelyingParty.cjs');
var PasskeyStorage = require('./models/PasskeyStorage.cjs');
var PasskeyTransport = require('./models/PasskeyTransport.cjs');
var PasskeyUser = require('./models/PasskeyUser.cjs');
var PasswordSourceTypeEnum = require('./models/PasswordSourceTypeEnum.cjs');
var PrefetchRequest = require('./models/PrefetchRequest.cjs');
var PriceData = require('./models/PriceData.cjs');
var ProjectSettings = require('./models/ProjectSettings.cjs');
var ProjectSettingsChains = require('./models/ProjectSettingsChains.cjs');
var ProjectSettingsDesign = require('./models/ProjectSettingsDesign.cjs');
var ProjectSettingsDesignButton = require('./models/ProjectSettingsDesignButton.cjs');
var ProjectSettingsDesignModal = require('./models/ProjectSettingsDesignModal.cjs');
var ProjectSettingsDesignWidget = require('./models/ProjectSettingsDesignWidget.cjs');
var ProjectSettingsGeneral = require('./models/ProjectSettingsGeneral.cjs');
var ProjectSettingsKyc = require('./models/ProjectSettingsKyc.cjs');
var ProjectSettingsPrivacy = require('./models/ProjectSettingsPrivacy.cjs');
var ProjectSettingsSdk = require('./models/ProjectSettingsSdk.cjs');
var ProjectSettingsSdkAccountAbstraction = require('./models/ProjectSettingsSdkAccountAbstraction.cjs');
var ProjectSettingsSdkEmailSignIn = require('./models/ProjectSettingsSdkEmailSignIn.cjs');
var ProjectSettingsSdkEmbeddedWallets = require('./models/ProjectSettingsSdkEmbeddedWallets.cjs');
var ProjectSettingsSdkSocialSignIn = require('./models/ProjectSettingsSdkSocialSignIn.cjs');
var ProjectSettingsSdkWaas = require('./models/ProjectSettingsSdkWaas.cjs');
var ProjectSettingsSdkWaasDelegatedAccess = require('./models/ProjectSettingsSdkWaasDelegatedAccess.cjs');
var ProjectSettingsSdkWaasOnSignUp = require('./models/ProjectSettingsSdkWaasOnSignUp.cjs');
var ProjectSettingsSdkWalletConnect = require('./models/ProjectSettingsSdkWalletConnect.cjs');
var ProjectSettingsSecurity = require('./models/ProjectSettingsSecurity.cjs');
var Provider = require('./models/Provider.cjs');
var ProviderAgreement = require('./models/ProviderAgreement.cjs');
var ProviderEntryPointVersionEnum = require('./models/ProviderEntryPointVersionEnum.cjs');
var ProviderEnum = require('./models/ProviderEnum.cjs');
var ProviderKernelVersionEnum = require('./models/ProviderKernelVersionEnum.cjs');
var ProviderMultichainAccountAbstractionProviders = require('./models/ProviderMultichainAccountAbstractionProviders.cjs');
var PublicKeyCredentialDescriptor = require('./models/PublicKeyCredentialDescriptor.cjs');
var PublicKeyCredentialParameters = require('./models/PublicKeyCredentialParameters.cjs');
var PublicKeyCredentialRpEntity = require('./models/PublicKeyCredentialRpEntity.cjs');
var PublicKeyCredentialType = require('./models/PublicKeyCredentialType.cjs');
var PublicKeyCredentialUserEntity = require('./models/PublicKeyCredentialUserEntity.cjs');
var PublishEvents = require('./models/PublishEvents.cjs');
var PublishEventsEvents = require('./models/PublishEventsEvents.cjs');
var RampConfiguration = require('./models/RampConfiguration.cjs');
var ReactSettings = require('./models/ReactSettings.cjs');
var RecoverMultipleClientKeySharesRequest = require('./models/RecoverMultipleClientKeySharesRequest.cjs');
var RecoverMultipleClientKeySharesResponse = require('./models/RecoverMultipleClientKeySharesResponse.cjs');
var RegisterEmbeddedWalletSessionKeyResponse = require('./models/RegisterEmbeddedWalletSessionKeyResponse.cjs');
var RegisterSessionKeyRequest = require('./models/RegisterSessionKeyRequest.cjs');
var ReshareRequest = require('./models/ReshareRequest.cjs');
var ResidentKeyRequirement = require('./models/ResidentKeyRequirement.cjs');
var ScanWebsiteUrlRequest = require('./models/ScanWebsiteUrlRequest.cjs');
var ScanWebsiteUrlResponse = require('./models/ScanWebsiteUrlResponse.cjs');
var SdkSettingsRequest = require('./models/SdkSettingsRequest.cjs');
var SdkUser = require('./models/SdkUser.cjs');
var SdkView = require('./models/SdkView.cjs');
var SdkViewSection = require('./models/SdkViewSection.cjs');
var SdkViewSectionAlignment = require('./models/SdkViewSectionAlignment.cjs');
var SdkViewSectionType = require('./models/SdkViewSectionType.cjs');
var SdkViewType = require('./models/SdkViewType.cjs');
var SdkViewUpdateRequest = require('./models/SdkViewUpdateRequest.cjs');
var SdkViewsResponse = require('./models/SdkViewsResponse.cjs');
var SignInProviderEnum = require('./models/SignInProviderEnum.cjs');
var SignMessageAuthorizationSignature = require('./models/SignMessageAuthorizationSignature.cjs');
var SignMessageContext = require('./models/SignMessageContext.cjs');
var SignMessageEip7702Auth = require('./models/SignMessageEip7702Auth.cjs');
var SignMessageEvmMessage = require('./models/SignMessageEvmMessage.cjs');
var SignMessageEvmMessageAnyOf = require('./models/SignMessageEvmMessageAnyOf.cjs');
var SignMessageEvmTransaction = require('./models/SignMessageEvmTransaction.cjs');
var SignMessageEvmTypedData = require('./models/SignMessageEvmTypedData.cjs');
var SignMessageEvmUserOperation = require('./models/SignMessageEvmUserOperation.cjs');
var SignMessageSuiTransaction = require('./models/SignMessageSuiTransaction.cjs');
var SignMessageSvmTransaction = require('./models/SignMessageSvmTransaction.cjs');
var SignMessageUserOperationData = require('./models/SignMessageUserOperationData.cjs');
var SignMessageWithWaasRequest = require('./models/SignMessageWithWaasRequest.cjs');
var SignTransactionWithWaasRequest = require('./models/SignTransactionWithWaasRequest.cjs');
var SimulateEVMTransactionRequest = require('./models/SimulateEVMTransactionRequest.cjs');
var SimulateSVMTransactionRequest = require('./models/SimulateSVMTransactionRequest.cjs');
var SimulateTransactionResponse = require('./models/SimulateTransactionResponse.cjs');
var SimulateUserOpRequest = require('./models/SimulateUserOpRequest.cjs');
var SmartWalletProperties = require('./models/SmartWalletProperties.cjs');
var SmsCountryCode = require('./models/SmsCountryCode.cjs');
var SmsVerificationCreateRequest = require('./models/SmsVerificationCreateRequest.cjs');
var SmsVerificationCreateResponse = require('./models/SmsVerificationCreateResponse.cjs');
var SmsVerificationRetryRequest = require('./models/SmsVerificationRetryRequest.cjs');
var SmsVerificationVerifyRequest = require('./models/SmsVerificationVerifyRequest.cjs');
var SocialSignInProvider = require('./models/SocialSignInProvider.cjs');
var SocialSignInProviderEnum = require('./models/SocialSignInProviderEnum.cjs');
var SolanaTransactionOptimizationRequest = require('./models/SolanaTransactionOptimizationRequest.cjs');
var SolanaTransactionOptimizationResponse = require('./models/SolanaTransactionOptimizationResponse.cjs');
var SupportedOfframpsResponse = require('./models/SupportedOfframpsResponse.cjs');
var SupportedOnrampsResponse = require('./models/SupportedOnrampsResponse.cjs');
var SupportedSecurityMethod = require('./models/SupportedSecurityMethod.cjs');
var SupportedSecurityMethods = require('./models/SupportedSecurityMethods.cjs');
var TelegramPostRequest = require('./models/TelegramPostRequest.cjs');
var TelegramUser = require('./models/TelegramUser.cjs');
var ThresholdSignatureScheme = require('./models/ThresholdSignatureScheme.cjs');
var TimeUnitEnum = require('./models/TimeUnitEnum.cjs');
var TokenBalance = require('./models/TokenBalance.cjs');
var TooManyRequests = require('./models/TooManyRequests.cjs');
var TransferDestination = require('./models/TransferDestination.cjs');
var TransferDestinationResponse = require('./models/TransferDestinationResponse.cjs');
var TurnkeyCreateWalletAccountsRequestBody = require('./models/TurnkeyCreateWalletAccountsRequestBody.cjs');
var TurnkeyCreateWalletAccountsRequestBodyParameters = require('./models/TurnkeyCreateWalletAccountsRequestBodyParameters.cjs');
var TurnkeyDeleteEmbeddedWalletsRequestBody = require('./models/TurnkeyDeleteEmbeddedWalletsRequestBody.cjs');
var TurnkeyDeleteEmbeddedWalletsRequestBodyParameters = require('./models/TurnkeyDeleteEmbeddedWalletsRequestBodyParameters.cjs');
var TurnkeySignedRequest = require('./models/TurnkeySignedRequest.cjs');
var TurnkeyStamp = require('./models/TurnkeyStamp.cjs');
var TurnkeyWalletAccount = require('./models/TurnkeyWalletAccount.cjs');
var TurnkeyWalletProperties = require('./models/TurnkeyWalletProperties.cjs');
var Unauthorized = require('./models/Unauthorized.cjs');
var UnprocessableEntity = require('./models/UnprocessableEntity.cjs');
var UnprocessableEntityErrorCode = require('./models/UnprocessableEntityErrorCode.cjs');
var UnprocessableEntityErrorPayload = require('./models/UnprocessableEntityErrorPayload.cjs');
var UpdateRecoveryEmailRequest = require('./models/UpdateRecoveryEmailRequest.cjs');
var UpdateSelfResponse = require('./models/UpdateSelfResponse.cjs');
var UpdateSelfResponseAllOf = require('./models/UpdateSelfResponseAllOf.cjs');
var UpdateUserPasskeyRequest = require('./models/UpdateUserPasskeyRequest.cjs');
var UpgradeEmbeddedWalletToV2Request = require('./models/UpgradeEmbeddedWalletToV2Request.cjs');
var UserFields = require('./models/UserFields.cjs');
var UserFieldsCheckEnum = require('./models/UserFieldsCheckEnum.cjs');
var UserFieldsCheckParams = require('./models/UserFieldsCheckParams.cjs');
var UserFieldsCheckResponse = require('./models/UserFieldsCheckResponse.cjs');
var UserIdentifierTypeEnum = require('./models/UserIdentifierTypeEnum.cjs');
var UserOauthAccessTokenResponse = require('./models/UserOauthAccessTokenResponse.cjs');
var UserPasskey = require('./models/UserPasskey.cjs');
var UserVerificationRequirement = require('./models/UserVerificationRequirement.cjs');
var UserWalletSelectionRequest = require('./models/UserWalletSelectionRequest.cjs');
var V6UserOp = require('./models/V6UserOp.cjs');
var V7UserOp = require('./models/V7UserOp.cjs');
var VerifyRequest = require('./models/VerifyRequest.cjs');
var VerifyResponse = require('./models/VerifyResponse.cjs');
var VerifyUnlinkRequest = require('./models/VerifyUnlinkRequest.cjs');
var WaasBackupOptionsEnum = require('./models/WaasBackupOptionsEnum.cjs');
var WaasChainEnum = require('./models/WaasChainEnum.cjs');
var WaasWalletProperties = require('./models/WaasWalletProperties.cjs');
var WalletAdditionalAddress = require('./models/WalletAdditionalAddress.cjs');
var WalletAddressType = require('./models/WalletAddressType.cjs');
var WalletKeyShareInfo = require('./models/WalletKeyShareInfo.cjs');
var WalletKeyShareInfoWithEncryptedAccountCredential = require('./models/WalletKeyShareInfoWithEncryptedAccountCredential.cjs');
var WalletKeyShareInfoWithEncryptedAccountCredentialAllOf = require('./models/WalletKeyShareInfoWithEncryptedAccountCredentialAllOf.cjs');
var WalletProperties = require('./models/WalletProperties.cjs');
var WalletProviderEnum = require('./models/WalletProviderEnum.cjs');
var ZerodevBundlerProvider = require('./models/ZerodevBundlerProvider.cjs');



exports.BASE_PATH = runtime.BASE_PATH;
exports.BaseAPI = runtime.BaseAPI;
exports.BlobApiResponse = runtime.BlobApiResponse;
exports.COLLECTION_FORMATS = runtime.COLLECTION_FORMATS;
exports.Configuration = runtime.Configuration;
exports.JSONApiResponse = runtime.JSONApiResponse;
exports.RequiredError = runtime.RequiredError;
exports.TextApiResponse = runtime.TextApiResponse;
exports.VoidApiResponse = runtime.VoidApiResponse;
exports.canConsumeForm = runtime.canConsumeForm;
exports.exists = runtime.exists;
exports.mapValues = runtime.mapValues;
exports.querystring = runtime.querystring;
exports.SDKApi = SDKApi.SDKApi;
exports.AccountFromJSON = Account.AccountFromJSON;
exports.AccountFromJSONTyped = Account.AccountFromJSONTyped;
exports.AccountToJSON = Account.AccountToJSON;
exports.AccountBalancesFromJSON = AccountBalances.AccountBalancesFromJSON;
exports.AccountBalancesFromJSONTyped = AccountBalances.AccountBalancesFromJSONTyped;
exports.AccountBalancesToJSON = AccountBalances.AccountBalancesToJSON;
exports.AssetFromJSON = Asset.AssetFromJSON;
exports.AssetFromJSONTyped = Asset.AssetFromJSONTyped;
exports.AssetToJSON = Asset.AssetToJSON;
exports.AssetDiffFromJSON = AssetDiff.AssetDiffFromJSON;
exports.AssetDiffFromJSONTyped = AssetDiff.AssetDiffFromJSONTyped;
exports.AssetDiffToJSON = AssetDiff.AssetDiffToJSON;
exports.AssetExposureFromJSON = AssetExposure.AssetExposureFromJSON;
exports.AssetExposureFromJSONTyped = AssetExposure.AssetExposureFromJSONTyped;
exports.AssetExposureToJSON = AssetExposure.AssetExposureToJSON;
exports.AssetExposureSpenderToInfoFromJSON = AssetExposureSpenderToInfo.AssetExposureSpenderToInfoFromJSON;
exports.AssetExposureSpenderToInfoFromJSONTyped = AssetExposureSpenderToInfo.AssetExposureSpenderToInfoFromJSONTyped;
exports.AssetExposureSpenderToInfoToJSON = AssetExposureSpenderToInfo.AssetExposureSpenderToInfoToJSON;
exports.AssetTransferFromJSON = AssetTransfer.AssetTransferFromJSON;
exports.AssetTransferFromJSONTyped = AssetTransfer.AssetTransferFromJSONTyped;
exports.AssetTransferToJSON = AssetTransfer.AssetTransferToJSON;
Object.defineProperty(exports, 'AttestationConveyancePreference', {
	enumerable: true,
	get: function () { return AttestationConveyancePreference.AttestationConveyancePreference; }
});
exports.AttestationConveyancePreferenceFromJSON = AttestationConveyancePreference.AttestationConveyancePreferenceFromJSON;
exports.AttestationConveyancePreferenceFromJSONTyped = AttestationConveyancePreference.AttestationConveyancePreferenceFromJSONTyped;
exports.AttestationConveyancePreferenceToJSON = AttestationConveyancePreference.AttestationConveyancePreferenceToJSON;
Object.defineProperty(exports, 'AuthModeEnum', {
	enumerable: true,
	get: function () { return AuthModeEnum.AuthModeEnum; }
});
exports.AuthModeEnumFromJSON = AuthModeEnum.AuthModeEnumFromJSON;
exports.AuthModeEnumFromJSONTyped = AuthModeEnum.AuthModeEnumFromJSONTyped;
exports.AuthModeEnumToJSON = AuthModeEnum.AuthModeEnumToJSON;
exports.AuthSettingsFromJSON = AuthSettings.AuthSettingsFromJSON;
exports.AuthSettingsFromJSONTyped = AuthSettings.AuthSettingsFromJSONTyped;
exports.AuthSettingsToJSON = AuthSettings.AuthSettingsToJSON;
Object.defineProperty(exports, 'AuthStorageEnum', {
	enumerable: true,
	get: function () { return AuthStorageEnum.AuthStorageEnum; }
});
exports.AuthStorageEnumFromJSON = AuthStorageEnum.AuthStorageEnumFromJSON;
exports.AuthStorageEnumFromJSONTyped = AuthStorageEnum.AuthStorageEnumFromJSONTyped;
exports.AuthStorageEnumToJSON = AuthStorageEnum.AuthStorageEnumToJSON;
exports.AuthenticationExtensionsClientInputsFromJSON = AuthenticationExtensionsClientInputs.AuthenticationExtensionsClientInputsFromJSON;
exports.AuthenticationExtensionsClientInputsFromJSONTyped = AuthenticationExtensionsClientInputs.AuthenticationExtensionsClientInputsFromJSONTyped;
exports.AuthenticationExtensionsClientInputsToJSON = AuthenticationExtensionsClientInputs.AuthenticationExtensionsClientInputsToJSON;
exports.AuthenticationExtensionsClientOutputsFromJSON = AuthenticationExtensionsClientOutputs.AuthenticationExtensionsClientOutputsFromJSON;
exports.AuthenticationExtensionsClientOutputsFromJSONTyped = AuthenticationExtensionsClientOutputs.AuthenticationExtensionsClientOutputsFromJSONTyped;
exports.AuthenticationExtensionsClientOutputsToJSON = AuthenticationExtensionsClientOutputs.AuthenticationExtensionsClientOutputsToJSON;
exports.AuthenticatorAssertionResponseFromJSON = AuthenticatorAssertionResponse.AuthenticatorAssertionResponseFromJSON;
exports.AuthenticatorAssertionResponseFromJSONTyped = AuthenticatorAssertionResponse.AuthenticatorAssertionResponseFromJSONTyped;
exports.AuthenticatorAssertionResponseToJSON = AuthenticatorAssertionResponse.AuthenticatorAssertionResponseToJSON;
Object.defineProperty(exports, 'AuthenticatorAttachment', {
	enumerable: true,
	get: function () { return AuthenticatorAttachment.AuthenticatorAttachment; }
});
exports.AuthenticatorAttachmentFromJSON = AuthenticatorAttachment.AuthenticatorAttachmentFromJSON;
exports.AuthenticatorAttachmentFromJSONTyped = AuthenticatorAttachment.AuthenticatorAttachmentFromJSONTyped;
exports.AuthenticatorAttachmentToJSON = AuthenticatorAttachment.AuthenticatorAttachmentToJSON;
exports.AuthenticatorAttestationResponseFromJSON = AuthenticatorAttestationResponse.AuthenticatorAttestationResponseFromJSON;
exports.AuthenticatorAttestationResponseFromJSONTyped = AuthenticatorAttestationResponse.AuthenticatorAttestationResponseFromJSONTyped;
exports.AuthenticatorAttestationResponseToJSON = AuthenticatorAttestationResponse.AuthenticatorAttestationResponseToJSON;
exports.AuthenticatorSelectionCriteriaFromJSON = AuthenticatorSelectionCriteria.AuthenticatorSelectionCriteriaFromJSON;
exports.AuthenticatorSelectionCriteriaFromJSONTyped = AuthenticatorSelectionCriteria.AuthenticatorSelectionCriteriaFromJSONTyped;
exports.AuthenticatorSelectionCriteriaToJSON = AuthenticatorSelectionCriteria.AuthenticatorSelectionCriteriaToJSON;
Object.defineProperty(exports, 'AuthenticatorTransportProtocol', {
	enumerable: true,
	get: function () { return AuthenticatorTransportProtocol.AuthenticatorTransportProtocol; }
});
exports.AuthenticatorTransportProtocolFromJSON = AuthenticatorTransportProtocol.AuthenticatorTransportProtocolFromJSON;
exports.AuthenticatorTransportProtocolFromJSONTyped = AuthenticatorTransportProtocol.AuthenticatorTransportProtocolFromJSONTyped;
exports.AuthenticatorTransportProtocolToJSON = AuthenticatorTransportProtocol.AuthenticatorTransportProtocolToJSON;
exports.BackupKeySharesToLocationRequestFromJSON = BackupKeySharesToLocationRequest.BackupKeySharesToLocationRequestFromJSON;
exports.BackupKeySharesToLocationRequestFromJSONTyped = BackupKeySharesToLocationRequest.BackupKeySharesToLocationRequestFromJSONTyped;
exports.BackupKeySharesToLocationRequestToJSON = BackupKeySharesToLocationRequest.BackupKeySharesToLocationRequestToJSON;
exports.BackupKeySharesToLocationResponseFromJSON = BackupKeySharesToLocationResponse.BackupKeySharesToLocationResponseFromJSON;
exports.BackupKeySharesToLocationResponseFromJSONTyped = BackupKeySharesToLocationResponse.BackupKeySharesToLocationResponseFromJSONTyped;
exports.BackupKeySharesToLocationResponseToJSON = BackupKeySharesToLocationResponse.BackupKeySharesToLocationResponseToJSON;
exports.BackupKeySharesToLocationsRequestFromJSON = BackupKeySharesToLocationsRequest.BackupKeySharesToLocationsRequestFromJSON;
exports.BackupKeySharesToLocationsRequestFromJSONTyped = BackupKeySharesToLocationsRequest.BackupKeySharesToLocationsRequestFromJSONTyped;
exports.BackupKeySharesToLocationsRequestToJSON = BackupKeySharesToLocationsRequest.BackupKeySharesToLocationsRequestToJSON;
exports.BackupKeySharesToLocationsRequestLocationsFromJSON = BackupKeySharesToLocationsRequestLocations.BackupKeySharesToLocationsRequestLocationsFromJSON;
exports.BackupKeySharesToLocationsRequestLocationsFromJSONTyped = BackupKeySharesToLocationsRequestLocations.BackupKeySharesToLocationsRequestLocationsFromJSONTyped;
exports.BackupKeySharesToLocationsRequestLocationsToJSON = BackupKeySharesToLocationsRequestLocations.BackupKeySharesToLocationsRequestLocationsToJSON;
exports.BackupKeySharesToLocationsResponseFromJSON = BackupKeySharesToLocationsResponse.BackupKeySharesToLocationsResponseFromJSON;
exports.BackupKeySharesToLocationsResponseFromJSONTyped = BackupKeySharesToLocationsResponse.BackupKeySharesToLocationsResponseFromJSONTyped;
exports.BackupKeySharesToLocationsResponseToJSON = BackupKeySharesToLocationsResponse.BackupKeySharesToLocationsResponseToJSON;
exports.BackupKeySharesToLocationsResponseLocationsWithKeySharesFromJSON = BackupKeySharesToLocationsResponseLocationsWithKeyShares.BackupKeySharesToLocationsResponseLocationsWithKeySharesFromJSON;
exports.BackupKeySharesToLocationsResponseLocationsWithKeySharesFromJSONTyped = BackupKeySharesToLocationsResponseLocationsWithKeyShares.BackupKeySharesToLocationsResponseLocationsWithKeySharesFromJSONTyped;
exports.BackupKeySharesToLocationsResponseLocationsWithKeySharesToJSON = BackupKeySharesToLocationsResponseLocationsWithKeyShares.BackupKeySharesToLocationsResponseLocationsWithKeySharesToJSON;
exports.BackupMultipleClientKeySharesRequestFromJSON = BackupMultipleClientKeySharesRequest.BackupMultipleClientKeySharesRequestFromJSON;
exports.BackupMultipleClientKeySharesRequestFromJSONTyped = BackupMultipleClientKeySharesRequest.BackupMultipleClientKeySharesRequestFromJSONTyped;
exports.BackupMultipleClientKeySharesRequestToJSON = BackupMultipleClientKeySharesRequest.BackupMultipleClientKeySharesRequestToJSON;
exports.BackupMultipleClientKeySharesResponseFromJSON = BackupMultipleClientKeySharesResponse.BackupMultipleClientKeySharesResponseFromJSON;
exports.BackupMultipleClientKeySharesResponseFromJSONTyped = BackupMultipleClientKeySharesResponse.BackupMultipleClientKeySharesResponseFromJSONTyped;
exports.BackupMultipleClientKeySharesResponseToJSON = BackupMultipleClientKeySharesResponse.BackupMultipleClientKeySharesResponseToJSON;
exports.BadGatewayFromJSON = BadGateway.BadGatewayFromJSON;
exports.BadGatewayFromJSONTyped = BadGateway.BadGatewayFromJSONTyped;
exports.BadGatewayToJSON = BadGateway.BadGatewayToJSON;
exports.BadRequestFromJSON = BadRequest.BadRequestFromJSON;
exports.BadRequestFromJSONTyped = BadRequest.BadRequestFromJSONTyped;
exports.BadRequestToJSON = BadRequest.BadRequestToJSON;
exports.BaseUserFromJSON = BaseUser.BaseUserFromJSON;
exports.BaseUserFromJSONTyped = BaseUser.BaseUserFromJSONTyped;
exports.BaseUserToJSON = BaseUser.BaseUserToJSON;
exports.BlockaidValidationFromJSON = BlockaidValidation.BlockaidValidationFromJSON;
exports.BlockaidValidationFromJSONTyped = BlockaidValidation.BlockaidValidationFromJSONTyped;
exports.BlockaidValidationToJSON = BlockaidValidation.BlockaidValidationToJSON;
exports.ChainConfigurationFromJSON = ChainConfiguration.ChainConfigurationFromJSON;
exports.ChainConfigurationFromJSONTyped = ChainConfiguration.ChainConfigurationFromJSONTyped;
exports.ChainConfigurationToJSON = ChainConfiguration.ChainConfigurationToJSON;
Object.defineProperty(exports, 'ChainEnum', {
	enumerable: true,
	get: function () { return ChainEnum.ChainEnum; }
});
exports.ChainEnumFromJSON = ChainEnum.ChainEnumFromJSON;
exports.ChainEnumFromJSONTyped = ChainEnum.ChainEnumFromJSONTyped;
exports.ChainEnumToJSON = ChainEnum.ChainEnumToJSON;
exports.CoinbaseMpcWalletPropertiesFromJSON = CoinbaseMpcWalletProperties.CoinbaseMpcWalletPropertiesFromJSON;
exports.CoinbaseMpcWalletPropertiesFromJSONTyped = CoinbaseMpcWalletProperties.CoinbaseMpcWalletPropertiesFromJSONTyped;
exports.CoinbaseMpcWalletPropertiesToJSON = CoinbaseMpcWalletProperties.CoinbaseMpcWalletPropertiesToJSON;
exports.CompletePasskeyRecoveryRequestFromJSON = CompletePasskeyRecoveryRequest.CompletePasskeyRecoveryRequestFromJSON;
exports.CompletePasskeyRecoveryRequestFromJSONTyped = CompletePasskeyRecoveryRequest.CompletePasskeyRecoveryRequestFromJSONTyped;
exports.CompletePasskeyRecoveryRequestToJSON = CompletePasskeyRecoveryRequest.CompletePasskeyRecoveryRequestToJSON;
exports.ConnectRequestFromJSON = ConnectRequest.ConnectRequestFromJSON;
exports.ConnectRequestFromJSONTyped = ConnectRequest.ConnectRequestFromJSONTyped;
exports.ConnectRequestToJSON = ConnectRequest.ConnectRequestToJSON;
exports.CreateEmbeddedWalletParamsFromJSON = CreateEmbeddedWalletParams.CreateEmbeddedWalletParamsFromJSON;
exports.CreateEmbeddedWalletParamsFromJSONTyped = CreateEmbeddedWalletParams.CreateEmbeddedWalletParamsFromJSONTyped;
exports.CreateEmbeddedWalletParamsToJSON = CreateEmbeddedWalletParams.CreateEmbeddedWalletParamsToJSON;
exports.CreateEmbeddedWalletSpecificOptsFromJSON = CreateEmbeddedWalletSpecificOpts.CreateEmbeddedWalletSpecificOptsFromJSON;
exports.CreateEmbeddedWalletSpecificOptsFromJSONTyped = CreateEmbeddedWalletSpecificOpts.CreateEmbeddedWalletSpecificOptsFromJSONTyped;
exports.CreateEmbeddedWalletSpecificOptsToJSON = CreateEmbeddedWalletSpecificOpts.CreateEmbeddedWalletSpecificOptsToJSON;
exports.CreateEmbeddedWalletsRequestFromJSON = CreateEmbeddedWalletsRequest.CreateEmbeddedWalletsRequestFromJSON;
exports.CreateEmbeddedWalletsRequestFromJSONTyped = CreateEmbeddedWalletsRequest.CreateEmbeddedWalletsRequestFromJSONTyped;
exports.CreateEmbeddedWalletsRequestToJSON = CreateEmbeddedWalletsRequest.CreateEmbeddedWalletsRequestToJSON;
exports.CreateExchangeTransferRequestFromJSON = CreateExchangeTransferRequest.CreateExchangeTransferRequestFromJSON;
exports.CreateExchangeTransferRequestFromJSONTyped = CreateExchangeTransferRequest.CreateExchangeTransferRequestFromJSONTyped;
exports.CreateExchangeTransferRequestToJSON = CreateExchangeTransferRequest.CreateExchangeTransferRequestToJSON;
exports.CreateExchangeTransferRequestNetworkObjectFromJSON = CreateExchangeTransferRequestNetworkObject.CreateExchangeTransferRequestNetworkObjectFromJSON;
exports.CreateExchangeTransferRequestNetworkObjectFromJSONTyped = CreateExchangeTransferRequestNetworkObject.CreateExchangeTransferRequestNetworkObjectFromJSONTyped;
exports.CreateExchangeTransferRequestNetworkObjectToJSON = CreateExchangeTransferRequestNetworkObject.CreateExchangeTransferRequestNetworkObjectToJSON;
exports.CreateMfaTokenFromJSON = CreateMfaToken.CreateMfaTokenFromJSON;
exports.CreateMfaTokenFromJSONTyped = CreateMfaToken.CreateMfaTokenFromJSONTyped;
exports.CreateMfaTokenToJSON = CreateMfaToken.CreateMfaTokenToJSON;
exports.CreateTurnkeyEmbeddedWalletSpecificOptsFromJSON = CreateTurnkeyEmbeddedWalletSpecificOpts.CreateTurnkeyEmbeddedWalletSpecificOptsFromJSON;
exports.CreateTurnkeyEmbeddedWalletSpecificOptsFromJSONTyped = CreateTurnkeyEmbeddedWalletSpecificOpts.CreateTurnkeyEmbeddedWalletSpecificOptsFromJSONTyped;
exports.CreateTurnkeyEmbeddedWalletSpecificOptsToJSON = CreateTurnkeyEmbeddedWalletSpecificOpts.CreateTurnkeyEmbeddedWalletSpecificOptsToJSON;
exports.CreateUserEmbeddedWalletsFromFarcasterRequestFromJSON = CreateUserEmbeddedWalletsFromFarcasterRequest.CreateUserEmbeddedWalletsFromFarcasterRequestFromJSON;
exports.CreateUserEmbeddedWalletsFromFarcasterRequestFromJSONTyped = CreateUserEmbeddedWalletsFromFarcasterRequest.CreateUserEmbeddedWalletsFromFarcasterRequestFromJSONTyped;
exports.CreateUserEmbeddedWalletsFromFarcasterRequestToJSON = CreateUserEmbeddedWalletsFromFarcasterRequest.CreateUserEmbeddedWalletsFromFarcasterRequestToJSON;
exports.CreateUserEmbeddedWalletsRequestFromJSON = CreateUserEmbeddedWalletsRequest.CreateUserEmbeddedWalletsRequestFromJSON;
exports.CreateUserEmbeddedWalletsRequestFromJSONTyped = CreateUserEmbeddedWalletsRequest.CreateUserEmbeddedWalletsRequestFromJSONTyped;
exports.CreateUserEmbeddedWalletsRequestToJSON = CreateUserEmbeddedWalletsRequest.CreateUserEmbeddedWalletsRequestToJSON;
exports.CreateWaasAccountRequestFromJSON = CreateWaasAccountRequest.CreateWaasAccountRequestFromJSON;
exports.CreateWaasAccountRequestFromJSONTyped = CreateWaasAccountRequest.CreateWaasAccountRequestFromJSONTyped;
exports.CreateWaasAccountRequestToJSON = CreateWaasAccountRequest.CreateWaasAccountRequestToJSON;
exports.CreateWalletAccountRequestFromJSON = CreateWalletAccountRequest.CreateWalletAccountRequestFromJSON;
exports.CreateWalletAccountRequestFromJSONTyped = CreateWalletAccountRequest.CreateWalletAccountRequestFromJSONTyped;
exports.CreateWalletAccountRequestToJSON = CreateWalletAccountRequest.CreateWalletAccountRequestToJSON;
exports.CurrencyFromJSON = Currency.CurrencyFromJSON;
exports.CurrencyFromJSONTyped = Currency.CurrencyFromJSONTyped;
exports.CurrencyToJSON = Currency.CurrencyToJSON;
Object.defineProperty(exports, 'CurrencyType', {
	enumerable: true,
	get: function () { return CurrencyType.CurrencyType; }
});
exports.CurrencyTypeFromJSON = CurrencyType.CurrencyTypeFromJSON;
exports.CurrencyTypeFromJSONTyped = CurrencyType.CurrencyTypeFromJSONTyped;
exports.CurrencyTypeToJSON = CurrencyType.CurrencyTypeToJSON;
Object.defineProperty(exports, 'CustomFieldType', {
	enumerable: true,
	get: function () { return CustomFieldType.CustomFieldType; }
});
exports.CustomFieldTypeFromJSON = CustomFieldType.CustomFieldTypeFromJSON;
exports.CustomFieldTypeFromJSONTyped = CustomFieldType.CustomFieldTypeFromJSONTyped;
exports.CustomFieldTypeToJSON = CustomFieldType.CustomFieldTypeToJSON;
exports.CustomFieldValidValueFromJSON = CustomFieldValidValue.CustomFieldValidValueFromJSON;
exports.CustomFieldValidValueFromJSONTyped = CustomFieldValidValue.CustomFieldValidValueFromJSONTyped;
exports.CustomFieldValidValueToJSON = CustomFieldValidValue.CustomFieldValidValueToJSON;
exports.CustomFieldValidationRulesFromJSON = CustomFieldValidationRules.CustomFieldValidationRulesFromJSON;
exports.CustomFieldValidationRulesFromJSONTyped = CustomFieldValidationRules.CustomFieldValidationRulesFromJSONTyped;
exports.CustomFieldValidationRulesToJSON = CustomFieldValidationRules.CustomFieldValidationRulesToJSON;
exports.DeleteEmbeddedWalletsRequestFromJSON = DeleteEmbeddedWalletsRequest.DeleteEmbeddedWalletsRequestFromJSON;
exports.DeleteEmbeddedWalletsRequestFromJSONTyped = DeleteEmbeddedWalletsRequest.DeleteEmbeddedWalletsRequestFromJSONTyped;
exports.DeleteEmbeddedWalletsRequestToJSON = DeleteEmbeddedWalletsRequest.DeleteEmbeddedWalletsRequestToJSON;
exports.DeleteUserPasskeyRequestFromJSON = DeleteUserPasskeyRequest.DeleteUserPasskeyRequestFromJSON;
exports.DeleteUserPasskeyRequestFromJSONTyped = DeleteUserPasskeyRequest.DeleteUserPasskeyRequestFromJSONTyped;
exports.DeleteUserPasskeyRequestToJSON = DeleteUserPasskeyRequest.DeleteUserPasskeyRequestToJSON;
exports.DurationFromJSON = Duration.DurationFromJSON;
exports.DurationFromJSONTyped = Duration.DurationFromJSONTyped;
exports.DurationToJSON = Duration.DurationToJSON;
exports.DynamicJwtFromJSON = DynamicJwt.DynamicJwtFromJSON;
exports.DynamicJwtFromJSONTyped = DynamicJwt.DynamicJwtFromJSONTyped;
exports.DynamicJwtToJSON = DynamicJwt.DynamicJwtToJSON;
Object.defineProperty(exports, 'EcdsaValidatorOptions', {
	enumerable: true,
	get: function () { return EcdsaValidatorOptions.EcdsaValidatorOptions; }
});
exports.EcdsaValidatorOptionsFromJSON = EcdsaValidatorOptions.EcdsaValidatorOptionsFromJSON;
exports.EcdsaValidatorOptionsFromJSONTyped = EcdsaValidatorOptions.EcdsaValidatorOptionsFromJSONTyped;
exports.EcdsaValidatorOptionsToJSON = EcdsaValidatorOptions.EcdsaValidatorOptionsToJSON;
exports.Eip712DomainFromJSON = Eip712Domain.Eip712DomainFromJSON;
exports.Eip712DomainFromJSONTyped = Eip712Domain.Eip712DomainFromJSONTyped;
exports.Eip712DomainToJSON = Eip712Domain.Eip712DomainToJSON;
exports.Eip712TypeMemberFromJSON = Eip712TypeMember.Eip712TypeMemberFromJSON;
exports.Eip712TypeMemberFromJSONTyped = Eip712TypeMember.Eip712TypeMemberFromJSONTyped;
exports.Eip712TypeMemberToJSON = Eip712TypeMember.Eip712TypeMemberToJSON;
exports.EmailProviderResponseFromJSON = EmailProviderResponse.EmailProviderResponseFromJSON;
exports.EmailProviderResponseFromJSONTyped = EmailProviderResponse.EmailProviderResponseFromJSONTyped;
exports.EmailProviderResponseToJSON = EmailProviderResponse.EmailProviderResponseToJSON;
exports.EmailVerificationCreateRequestFromJSON = EmailVerificationCreateRequest.EmailVerificationCreateRequestFromJSON;
exports.EmailVerificationCreateRequestFromJSONTyped = EmailVerificationCreateRequest.EmailVerificationCreateRequestFromJSONTyped;
exports.EmailVerificationCreateRequestToJSON = EmailVerificationCreateRequest.EmailVerificationCreateRequestToJSON;
exports.EmailVerificationCreateResponseFromJSON = EmailVerificationCreateResponse.EmailVerificationCreateResponseFromJSON;
exports.EmailVerificationCreateResponseFromJSONTyped = EmailVerificationCreateResponse.EmailVerificationCreateResponseFromJSONTyped;
exports.EmailVerificationCreateResponseToJSON = EmailVerificationCreateResponse.EmailVerificationCreateResponseToJSON;
exports.EmailVerificationMfaRequestFromJSON = EmailVerificationMfaRequest.EmailVerificationMfaRequestFromJSON;
exports.EmailVerificationMfaRequestFromJSONTyped = EmailVerificationMfaRequest.EmailVerificationMfaRequestFromJSONTyped;
exports.EmailVerificationMfaRequestToJSON = EmailVerificationMfaRequest.EmailVerificationMfaRequestToJSON;
exports.EmailVerificationRetryRequestFromJSON = EmailVerificationRetryRequest.EmailVerificationRetryRequestFromJSON;
exports.EmailVerificationRetryRequestFromJSONTyped = EmailVerificationRetryRequest.EmailVerificationRetryRequestFromJSONTyped;
exports.EmailVerificationRetryRequestToJSON = EmailVerificationRetryRequest.EmailVerificationRetryRequestToJSON;
exports.EmailVerificationVerifyRequestFromJSON = EmailVerificationVerifyRequest.EmailVerificationVerifyRequestFromJSON;
exports.EmailVerificationVerifyRequestFromJSONTyped = EmailVerificationVerifyRequest.EmailVerificationVerifyRequestFromJSONTyped;
exports.EmailVerificationVerifyRequestToJSON = EmailVerificationVerifyRequest.EmailVerificationVerifyRequestToJSON;
exports.EmbeddedWalletAuthTokenFromJSON = EmbeddedWalletAuthToken.EmbeddedWalletAuthTokenFromJSON;
exports.EmbeddedWalletAuthTokenFromJSONTyped = EmbeddedWalletAuthToken.EmbeddedWalletAuthTokenFromJSONTyped;
exports.EmbeddedWalletAuthTokenToJSON = EmbeddedWalletAuthToken.EmbeddedWalletAuthTokenToJSON;
Object.defineProperty(exports, 'EmbeddedWalletAuthType', {
	enumerable: true,
	get: function () { return EmbeddedWalletAuthType.EmbeddedWalletAuthType; }
});
exports.EmbeddedWalletAuthTypeFromJSON = EmbeddedWalletAuthType.EmbeddedWalletAuthTypeFromJSON;
exports.EmbeddedWalletAuthTypeFromJSONTyped = EmbeddedWalletAuthType.EmbeddedWalletAuthTypeFromJSONTyped;
exports.EmbeddedWalletAuthTypeToJSON = EmbeddedWalletAuthType.EmbeddedWalletAuthTypeToJSON;
Object.defineProperty(exports, 'EmbeddedWalletChainEnum', {
	enumerable: true,
	get: function () { return EmbeddedWalletChainEnum.EmbeddedWalletChainEnum; }
});
exports.EmbeddedWalletChainEnumFromJSON = EmbeddedWalletChainEnum.EmbeddedWalletChainEnumFromJSON;
exports.EmbeddedWalletChainEnumFromJSONTyped = EmbeddedWalletChainEnum.EmbeddedWalletChainEnumFromJSONTyped;
exports.EmbeddedWalletChainEnumToJSON = EmbeddedWalletChainEnum.EmbeddedWalletChainEnumToJSON;
exports.EmbeddedWalletPasscodeClaimRequestFromJSON = EmbeddedWalletPasscodeClaimRequest.EmbeddedWalletPasscodeClaimRequestFromJSON;
exports.EmbeddedWalletPasscodeClaimRequestFromJSONTyped = EmbeddedWalletPasscodeClaimRequest.EmbeddedWalletPasscodeClaimRequestFromJSONTyped;
exports.EmbeddedWalletPasscodeClaimRequestToJSON = EmbeddedWalletPasscodeClaimRequest.EmbeddedWalletPasscodeClaimRequestToJSON;
Object.defineProperty(exports, 'EmbeddedWalletProviderEnum', {
	enumerable: true,
	get: function () { return EmbeddedWalletProviderEnum.EmbeddedWalletProviderEnum; }
});
exports.EmbeddedWalletProviderEnumFromJSON = EmbeddedWalletProviderEnum.EmbeddedWalletProviderEnumFromJSON;
exports.EmbeddedWalletProviderEnumFromJSONTyped = EmbeddedWalletProviderEnum.EmbeddedWalletProviderEnumFromJSONTyped;
exports.EmbeddedWalletProviderEnumToJSON = EmbeddedWalletProviderEnum.EmbeddedWalletProviderEnumToJSON;
exports.EmbeddedWalletSecretFromJSON = EmbeddedWalletSecret.EmbeddedWalletSecretFromJSON;
exports.EmbeddedWalletSecretFromJSONTyped = EmbeddedWalletSecret.EmbeddedWalletSecretFromJSONTyped;
exports.EmbeddedWalletSecretToJSON = EmbeddedWalletSecret.EmbeddedWalletSecretToJSON;
exports.EmbeddedWalletSecretWithUpdatedJwtFromJSON = EmbeddedWalletSecretWithUpdatedJwt.EmbeddedWalletSecretWithUpdatedJwtFromJSON;
exports.EmbeddedWalletSecretWithUpdatedJwtFromJSONTyped = EmbeddedWalletSecretWithUpdatedJwt.EmbeddedWalletSecretWithUpdatedJwtFromJSONTyped;
exports.EmbeddedWalletSecretWithUpdatedJwtToJSON = EmbeddedWalletSecretWithUpdatedJwt.EmbeddedWalletSecretWithUpdatedJwtToJSON;
exports.EmbeddedWalletSecretWithUpdatedJwtAllOfFromJSON = EmbeddedWalletSecretWithUpdatedJwtAllOf.EmbeddedWalletSecretWithUpdatedJwtAllOfFromJSON;
exports.EmbeddedWalletSecretWithUpdatedJwtAllOfFromJSONTyped = EmbeddedWalletSecretWithUpdatedJwtAllOf.EmbeddedWalletSecretWithUpdatedJwtAllOfFromJSONTyped;
exports.EmbeddedWalletSecretWithUpdatedJwtAllOfToJSON = EmbeddedWalletSecretWithUpdatedJwtAllOf.EmbeddedWalletSecretWithUpdatedJwtAllOfToJSON;
Object.defineProperty(exports, 'EmbeddedWalletSecurityMethod', {
	enumerable: true,
	get: function () { return EmbeddedWalletSecurityMethod.EmbeddedWalletSecurityMethod; }
});
exports.EmbeddedWalletSecurityMethodFromJSON = EmbeddedWalletSecurityMethod.EmbeddedWalletSecurityMethodFromJSON;
exports.EmbeddedWalletSecurityMethodFromJSONTyped = EmbeddedWalletSecurityMethod.EmbeddedWalletSecurityMethodFromJSONTyped;
exports.EmbeddedWalletSecurityMethodToJSON = EmbeddedWalletSecurityMethod.EmbeddedWalletSecurityMethodToJSON;
Object.defineProperty(exports, 'EmbeddedWalletVersionEnum', {
	enumerable: true,
	get: function () { return EmbeddedWalletVersionEnum.EmbeddedWalletVersionEnum; }
});
exports.EmbeddedWalletVersionEnumFromJSON = EmbeddedWalletVersionEnum.EmbeddedWalletVersionEnumFromJSON;
exports.EmbeddedWalletVersionEnumFromJSONTyped = EmbeddedWalletVersionEnum.EmbeddedWalletVersionEnumFromJSONTyped;
exports.EmbeddedWalletVersionEnumToJSON = EmbeddedWalletVersionEnum.EmbeddedWalletVersionEnumToJSON;
Object.defineProperty(exports, 'EnvironmentEnum', {
	enumerable: true,
	get: function () { return EnvironmentEnum.EnvironmentEnum; }
});
exports.EnvironmentEnumFromJSON = EnvironmentEnum.EnvironmentEnumFromJSON;
exports.EnvironmentEnumFromJSONTyped = EnvironmentEnum.EnvironmentEnumFromJSONTyped;
exports.EnvironmentEnumToJSON = EnvironmentEnum.EnvironmentEnumToJSON;
exports.ErrorMessageWithCodeFromJSON = ErrorMessageWithCode.ErrorMessageWithCodeFromJSON;
exports.ErrorMessageWithCodeFromJSONTyped = ErrorMessageWithCode.ErrorMessageWithCodeFromJSONTyped;
exports.ErrorMessageWithCodeToJSON = ErrorMessageWithCode.ErrorMessageWithCodeToJSON;
exports.ExchangeFromJSON = Exchange.ExchangeFromJSON;
exports.ExchangeFromJSONTyped = Exchange.ExchangeFromJSONTyped;
exports.ExchangeToJSON = Exchange.ExchangeToJSON;
Object.defineProperty(exports, 'ExchangeKeyEnum', {
	enumerable: true,
	get: function () { return ExchangeKeyEnum.ExchangeKeyEnum; }
});
exports.ExchangeKeyEnumFromJSON = ExchangeKeyEnum.ExchangeKeyEnumFromJSON;
exports.ExchangeKeyEnumFromJSONTyped = ExchangeKeyEnum.ExchangeKeyEnumFromJSONTyped;
exports.ExchangeKeyEnumToJSON = ExchangeKeyEnum.ExchangeKeyEnumToJSON;
exports.ExchangeOptionFromJSON = ExchangeOption.ExchangeOptionFromJSON;
exports.ExchangeOptionFromJSONTyped = ExchangeOption.ExchangeOptionFromJSONTyped;
exports.ExchangeOptionToJSON = ExchangeOption.ExchangeOptionToJSON;
exports.ExchangeTransactionFromJSON = ExchangeTransaction.ExchangeTransactionFromJSON;
exports.ExchangeTransactionFromJSONTyped = ExchangeTransaction.ExchangeTransactionFromJSONTyped;
exports.ExchangeTransactionToJSON = ExchangeTransaction.ExchangeTransactionToJSON;
exports.ExchangeTransferResponseFromJSON = ExchangeTransferResponse.ExchangeTransferResponseFromJSON;
exports.ExchangeTransferResponseFromJSONTyped = ExchangeTransferResponse.ExchangeTransferResponseFromJSONTyped;
exports.ExchangeTransferResponseToJSON = ExchangeTransferResponse.ExchangeTransferResponseToJSON;
exports.ExportEmbeddedWalletResponseFromJSON = ExportEmbeddedWalletResponse.ExportEmbeddedWalletResponseFromJSON;
exports.ExportEmbeddedWalletResponseFromJSONTyped = ExportEmbeddedWalletResponse.ExportEmbeddedWalletResponseFromJSONTyped;
exports.ExportEmbeddedWalletResponseToJSON = ExportEmbeddedWalletResponse.ExportEmbeddedWalletResponseToJSON;
exports.ExportWaasWalletPrivateKeyRequestFromJSON = ExportWaasWalletPrivateKeyRequest.ExportWaasWalletPrivateKeyRequestFromJSON;
exports.ExportWaasWalletPrivateKeyRequestFromJSONTyped = ExportWaasWalletPrivateKeyRequest.ExportWaasWalletPrivateKeyRequestFromJSONTyped;
exports.ExportWaasWalletPrivateKeyRequestToJSON = ExportWaasWalletPrivateKeyRequest.ExportWaasWalletPrivateKeyRequestToJSON;
exports.ExternalAuthFromJSON = ExternalAuth.ExternalAuthFromJSON;
exports.ExternalAuthFromJSONTyped = ExternalAuth.ExternalAuthFromJSONTyped;
exports.ExternalAuthToJSON = ExternalAuth.ExternalAuthToJSON;
exports.ExternalAuthSigninRequestFromJSON = ExternalAuthSigninRequest.ExternalAuthSigninRequestFromJSON;
exports.ExternalAuthSigninRequestFromJSONTyped = ExternalAuthSigninRequest.ExternalAuthSigninRequestFromJSONTyped;
exports.ExternalAuthSigninRequestToJSON = ExternalAuthSigninRequest.ExternalAuthSigninRequestToJSON;
Object.defineProperty(exports, 'ExternalWalletFundingDefaultChain', {
	enumerable: true,
	get: function () { return ExternalWalletFundingDefaultChain.ExternalWalletFundingDefaultChain; }
});
exports.ExternalWalletFundingDefaultChainFromJSON = ExternalWalletFundingDefaultChain.ExternalWalletFundingDefaultChainFromJSON;
exports.ExternalWalletFundingDefaultChainFromJSONTyped = ExternalWalletFundingDefaultChain.ExternalWalletFundingDefaultChainFromJSONTyped;
exports.ExternalWalletFundingDefaultChainToJSON = ExternalWalletFundingDefaultChain.ExternalWalletFundingDefaultChainToJSON;
exports.ExternalWalletFundingDefaultSettingsFromJSON = ExternalWalletFundingDefaultSettings.ExternalWalletFundingDefaultSettingsFromJSON;
exports.ExternalWalletFundingDefaultSettingsFromJSONTyped = ExternalWalletFundingDefaultSettings.ExternalWalletFundingDefaultSettingsFromJSONTyped;
exports.ExternalWalletFundingDefaultSettingsToJSON = ExternalWalletFundingDefaultSettings.ExternalWalletFundingDefaultSettingsToJSON;
exports.ExternalWalletFundingDefaultSettingsTokenFromJSON = ExternalWalletFundingDefaultSettingsToken.ExternalWalletFundingDefaultSettingsTokenFromJSON;
exports.ExternalWalletFundingDefaultSettingsTokenFromJSONTyped = ExternalWalletFundingDefaultSettingsToken.ExternalWalletFundingDefaultSettingsTokenFromJSONTyped;
exports.ExternalWalletFundingDefaultSettingsTokenToJSON = ExternalWalletFundingDefaultSettingsToken.ExternalWalletFundingDefaultSettingsTokenToJSON;
Object.defineProperty(exports, 'ExternalWalletFundingTokenRule', {
	enumerable: true,
	get: function () { return ExternalWalletFundingTokenRule.ExternalWalletFundingTokenRule; }
});
exports.ExternalWalletFundingTokenRuleFromJSON = ExternalWalletFundingTokenRule.ExternalWalletFundingTokenRuleFromJSON;
exports.ExternalWalletFundingTokenRuleFromJSONTyped = ExternalWalletFundingTokenRule.ExternalWalletFundingTokenRuleFromJSONTyped;
exports.ExternalWalletFundingTokenRuleToJSON = ExternalWalletFundingTokenRule.ExternalWalletFundingTokenRuleToJSON;
exports.FarcasterSignInRequestFromJSON = FarcasterSignInRequest.FarcasterSignInRequestFromJSON;
exports.FarcasterSignInRequestFromJSONTyped = FarcasterSignInRequest.FarcasterSignInRequestFromJSONTyped;
exports.FarcasterSignInRequestToJSON = FarcasterSignInRequest.FarcasterSignInRequestToJSON;
exports.FeatureFlagsFromJSON = FeatureFlags.FeatureFlagsFromJSON;
exports.FeatureFlagsFromJSONTyped = FeatureFlags.FeatureFlagsFromJSONTyped;
exports.FeatureFlagsToJSON = FeatureFlags.FeatureFlagsToJSON;
exports.ForbiddenFromJSON = Forbidden.ForbiddenFromJSON;
exports.ForbiddenFromJSONTyped = Forbidden.ForbiddenFromJSONTyped;
exports.ForbiddenToJSON = Forbidden.ForbiddenToJSON;
exports.ForbiddenErrorPayloadFromJSON = ForbiddenErrorPayload.ForbiddenErrorPayloadFromJSON;
exports.ForbiddenErrorPayloadFromJSONTyped = ForbiddenErrorPayload.ForbiddenErrorPayloadFromJSONTyped;
exports.ForbiddenErrorPayloadToJSON = ForbiddenErrorPayload.ForbiddenErrorPayloadToJSON;
exports.ForbiddenWithErrorAndPayloadFromJSON = ForbiddenWithErrorAndPayload.ForbiddenWithErrorAndPayloadFromJSON;
exports.ForbiddenWithErrorAndPayloadFromJSONTyped = ForbiddenWithErrorAndPayload.ForbiddenWithErrorAndPayloadFromJSONTyped;
exports.ForbiddenWithErrorAndPayloadToJSON = ForbiddenWithErrorAndPayload.ForbiddenWithErrorAndPayloadToJSON;
exports.FrameworkSettingsFromJSON = FrameworkSettings.FrameworkSettingsFromJSON;
exports.FrameworkSettingsFromJSONTyped = FrameworkSettings.FrameworkSettingsFromJSONTyped;
exports.FrameworkSettingsToJSON = FrameworkSettings.FrameworkSettingsToJSON;
exports.FundingFromJSON = Funding.FundingFromJSON;
exports.FundingFromJSONTyped = Funding.FundingFromJSONTyped;
exports.FundingToJSON = Funding.FundingToJSON;
exports.FundingExternalWalletsFromJSON = FundingExternalWallets.FundingExternalWalletsFromJSON;
exports.FundingExternalWalletsFromJSONTyped = FundingExternalWallets.FundingExternalWalletsFromJSONTyped;
exports.FundingExternalWalletsToJSON = FundingExternalWallets.FundingExternalWalletsToJSON;
exports.FundingExternalWalletsDefaultSettingsFromJSON = FundingExternalWalletsDefaultSettings.FundingExternalWalletsDefaultSettingsFromJSON;
exports.FundingExternalWalletsDefaultSettingsFromJSONTyped = FundingExternalWalletsDefaultSettings.FundingExternalWalletsDefaultSettingsFromJSONTyped;
exports.FundingExternalWalletsDefaultSettingsToJSON = FundingExternalWalletsDefaultSettings.FundingExternalWalletsDefaultSettingsToJSON;
exports.FundingExternalWalletsMinAmountFromJSON = FundingExternalWalletsMinAmount.FundingExternalWalletsMinAmountFromJSON;
exports.FundingExternalWalletsMinAmountFromJSONTyped = FundingExternalWalletsMinAmount.FundingExternalWalletsMinAmountFromJSONTyped;
exports.FundingExternalWalletsMinAmountToJSON = FundingExternalWalletsMinAmount.FundingExternalWalletsMinAmountToJSON;
exports.GetPasskeyAuthenticationOptionsResponseFromJSON = GetPasskeyAuthenticationOptionsResponse.GetPasskeyAuthenticationOptionsResponseFromJSON;
exports.GetPasskeyAuthenticationOptionsResponseFromJSONTyped = GetPasskeyAuthenticationOptionsResponse.GetPasskeyAuthenticationOptionsResponseFromJSONTyped;
exports.GetPasskeyAuthenticationOptionsResponseToJSON = GetPasskeyAuthenticationOptionsResponse.GetPasskeyAuthenticationOptionsResponseToJSON;
exports.GetPasskeyRegistrationOptionsResponseFromJSON = GetPasskeyRegistrationOptionsResponse.GetPasskeyRegistrationOptionsResponseFromJSON;
exports.GetPasskeyRegistrationOptionsResponseFromJSONTyped = GetPasskeyRegistrationOptionsResponse.GetPasskeyRegistrationOptionsResponseFromJSONTyped;
exports.GetPasskeyRegistrationOptionsResponseToJSON = GetPasskeyRegistrationOptionsResponse.GetPasskeyRegistrationOptionsResponseToJSON;
exports.GetUserPasskeysResponseFromJSON = GetUserPasskeysResponse.GetUserPasskeysResponseFromJSON;
exports.GetUserPasskeysResponseFromJSONTyped = GetUserPasskeysResponse.GetUserPasskeysResponseFromJSONTyped;
exports.GetUserPasskeysResponseToJSON = GetUserPasskeysResponse.GetUserPasskeysResponseToJSON;
Object.defineProperty(exports, 'GlobalWalletAccessControlTypeEnum', {
	enumerable: true,
	get: function () { return GlobalWalletAccessControlTypeEnum.GlobalWalletAccessControlTypeEnum; }
});
exports.GlobalWalletAccessControlTypeEnumFromJSON = GlobalWalletAccessControlTypeEnum.GlobalWalletAccessControlTypeEnumFromJSON;
exports.GlobalWalletAccessControlTypeEnumFromJSONTyped = GlobalWalletAccessControlTypeEnum.GlobalWalletAccessControlTypeEnumFromJSONTyped;
exports.GlobalWalletAccessControlTypeEnumToJSON = GlobalWalletAccessControlTypeEnum.GlobalWalletAccessControlTypeEnumToJSON;
exports.GlobalWalletConnectionFromJSON = GlobalWalletConnection.GlobalWalletConnectionFromJSON;
exports.GlobalWalletConnectionFromJSONTyped = GlobalWalletConnection.GlobalWalletConnectionFromJSONTyped;
exports.GlobalWalletConnectionToJSON = GlobalWalletConnection.GlobalWalletConnectionToJSON;
exports.GlobalWalletConnectionCreateRequestFromJSON = GlobalWalletConnectionCreateRequest.GlobalWalletConnectionCreateRequestFromJSON;
exports.GlobalWalletConnectionCreateRequestFromJSONTyped = GlobalWalletConnectionCreateRequest.GlobalWalletConnectionCreateRequestFromJSONTyped;
exports.GlobalWalletConnectionCreateRequestToJSON = GlobalWalletConnectionCreateRequest.GlobalWalletConnectionCreateRequestToJSON;
Object.defineProperty(exports, 'GlobalWalletConnectionStatusEnum', {
	enumerable: true,
	get: function () { return GlobalWalletConnectionStatusEnum.GlobalWalletConnectionStatusEnum; }
});
exports.GlobalWalletConnectionStatusEnumFromJSON = GlobalWalletConnectionStatusEnum.GlobalWalletConnectionStatusEnumFromJSON;
exports.GlobalWalletConnectionStatusEnumFromJSONTyped = GlobalWalletConnectionStatusEnum.GlobalWalletConnectionStatusEnumFromJSONTyped;
exports.GlobalWalletConnectionStatusEnumToJSON = GlobalWalletConnectionStatusEnum.GlobalWalletConnectionStatusEnumToJSON;
exports.GlobalWalletSettingsFromJSON = GlobalWalletSettings.GlobalWalletSettingsFromJSON;
exports.GlobalWalletSettingsFromJSONTyped = GlobalWalletSettings.GlobalWalletSettingsFromJSONTyped;
exports.GlobalWalletSettingsToJSON = GlobalWalletSettings.GlobalWalletSettingsToJSON;
exports.GlobalWalletSettingsCustomMenuLinksFromJSON = GlobalWalletSettingsCustomMenuLinks.GlobalWalletSettingsCustomMenuLinksFromJSON;
exports.GlobalWalletSettingsCustomMenuLinksFromJSONTyped = GlobalWalletSettingsCustomMenuLinks.GlobalWalletSettingsCustomMenuLinksFromJSONTyped;
exports.GlobalWalletSettingsCustomMenuLinksToJSON = GlobalWalletSettingsCustomMenuLinks.GlobalWalletSettingsCustomMenuLinksToJSON;
exports.HCaptchaSettingsFromJSON = HCaptchaSettings.HCaptchaSettingsFromJSON;
exports.HCaptchaSettingsFromJSONTyped = HCaptchaSettings.HCaptchaSettingsFromJSONTyped;
exports.HCaptchaSettingsToJSON = HCaptchaSettings.HCaptchaSettingsToJSON;
Object.defineProperty(exports, 'HardwareWalletEnum', {
	enumerable: true,
	get: function () { return HardwareWalletEnum.HardwareWalletEnum; }
});
exports.HardwareWalletEnumFromJSON = HardwareWalletEnum.HardwareWalletEnumFromJSON;
exports.HardwareWalletEnumFromJSONTyped = HardwareWalletEnum.HardwareWalletEnumFromJSONTyped;
exports.HardwareWalletEnumToJSON = HardwareWalletEnum.HardwareWalletEnumToJSON;
exports.HardwareWalletPropertiesFromJSON = HardwareWalletProperties.HardwareWalletPropertiesFromJSON;
exports.HardwareWalletPropertiesFromJSONTyped = HardwareWalletProperties.HardwareWalletPropertiesFromJSONTyped;
exports.HardwareWalletPropertiesToJSON = HardwareWalletProperties.HardwareWalletPropertiesToJSON;
exports.HealthcheckResponseFromJSON = HealthcheckResponse.HealthcheckResponseFromJSON;
exports.HealthcheckResponseFromJSONTyped = HealthcheckResponse.HealthcheckResponseFromJSONTyped;
exports.HealthcheckResponseToJSON = HealthcheckResponse.HealthcheckResponseToJSON;
Object.defineProperty(exports, 'HealthcheckStatus', {
	enumerable: true,
	get: function () { return HealthcheckStatus.HealthcheckStatus; }
});
exports.HealthcheckStatusFromJSON = HealthcheckStatus.HealthcheckStatusFromJSON;
exports.HealthcheckStatusFromJSONTyped = HealthcheckStatus.HealthcheckStatusFromJSONTyped;
exports.HealthcheckStatusToJSON = HealthcheckStatus.HealthcheckStatusToJSON;
exports.ImportWaasPrivateKeyRequestFromJSON = ImportWaasPrivateKeyRequest.ImportWaasPrivateKeyRequestFromJSON;
exports.ImportWaasPrivateKeyRequestFromJSONTyped = ImportWaasPrivateKeyRequest.ImportWaasPrivateKeyRequestFromJSONTyped;
exports.ImportWaasPrivateKeyRequestToJSON = ImportWaasPrivateKeyRequest.ImportWaasPrivateKeyRequestToJSON;
exports.InitEmailAuthRequestFromJSON = InitEmailAuthRequest.InitEmailAuthRequestFromJSON;
exports.InitEmailAuthRequestFromJSONTyped = InitEmailAuthRequest.InitEmailAuthRequestFromJSONTyped;
exports.InitEmailAuthRequestToJSON = InitEmailAuthRequest.InitEmailAuthRequestToJSON;
exports.InitEmailAuthResponseFromJSON = InitEmailAuthResponse.InitEmailAuthResponseFromJSON;
exports.InitEmailAuthResponseFromJSONTyped = InitEmailAuthResponse.InitEmailAuthResponseFromJSONTyped;
exports.InitEmailAuthResponseToJSON = InitEmailAuthResponse.InitEmailAuthResponseToJSON;
exports.InitPasskeyRecoveryRequestFromJSON = InitPasskeyRecoveryRequest.InitPasskeyRecoveryRequestFromJSON;
exports.InitPasskeyRecoveryRequestFromJSONTyped = InitPasskeyRecoveryRequest.InitPasskeyRecoveryRequestFromJSONTyped;
exports.InitPasskeyRecoveryRequestToJSON = InitPasskeyRecoveryRequest.InitPasskeyRecoveryRequestToJSON;
exports.InitPasskeyRecoveryResponseFromJSON = InitPasskeyRecoveryResponse.InitPasskeyRecoveryResponseFromJSON;
exports.InitPasskeyRecoveryResponseFromJSONTyped = InitPasskeyRecoveryResponse.InitPasskeyRecoveryResponseFromJSONTyped;
exports.InitPasskeyRecoveryResponseToJSON = InitPasskeyRecoveryResponse.InitPasskeyRecoveryResponseToJSON;
exports.IntegrationSettingFromJSON = IntegrationSetting.IntegrationSettingFromJSON;
exports.IntegrationSettingFromJSONTyped = IntegrationSetting.IntegrationSettingFromJSONTyped;
exports.IntegrationSettingToJSON = IntegrationSetting.IntegrationSettingToJSON;
exports.InternalServerErrorFromJSON = InternalServerError.InternalServerErrorFromJSON;
exports.InternalServerErrorFromJSONTyped = InternalServerError.InternalServerErrorFromJSONTyped;
exports.InternalServerErrorToJSON = InternalServerError.InternalServerErrorToJSON;
exports.JwksKeyFromJSON = JwksKey.JwksKeyFromJSON;
exports.JwksKeyFromJSONTyped = JwksKey.JwksKeyFromJSONTyped;
exports.JwksKeyToJSON = JwksKey.JwksKeyToJSON;
exports.JwksResponseFromJSON = JwksResponse.JwksResponseFromJSON;
exports.JwksResponseFromJSONTyped = JwksResponse.JwksResponseFromJSONTyped;
exports.JwksResponseToJSON = JwksResponse.JwksResponseToJSON;
exports.JwtBlockchainAccountFromJSON = JwtBlockchainAccount.JwtBlockchainAccountFromJSON;
exports.JwtBlockchainAccountFromJSONTyped = JwtBlockchainAccount.JwtBlockchainAccountFromJSONTyped;
exports.JwtBlockchainAccountToJSON = JwtBlockchainAccount.JwtBlockchainAccountToJSON;
exports.JwtPayloadDeprecatedInfoFromJSON = JwtPayloadDeprecatedInfo.JwtPayloadDeprecatedInfoFromJSON;
exports.JwtPayloadDeprecatedInfoFromJSONTyped = JwtPayloadDeprecatedInfo.JwtPayloadDeprecatedInfoFromJSONTyped;
exports.JwtPayloadDeprecatedInfoToJSON = JwtPayloadDeprecatedInfo.JwtPayloadDeprecatedInfoToJSON;
exports.JwtVerifiedCredentialFromJSON = JwtVerifiedCredential.JwtVerifiedCredentialFromJSON;
exports.JwtVerifiedCredentialFromJSONTyped = JwtVerifiedCredential.JwtVerifiedCredentialFromJSONTyped;
exports.JwtVerifiedCredentialToJSON = JwtVerifiedCredential.JwtVerifiedCredentialToJSON;
Object.defineProperty(exports, 'JwtVerifiedCredentialFormatEnum', {
	enumerable: true,
	get: function () { return JwtVerifiedCredentialFormatEnum.JwtVerifiedCredentialFormatEnum; }
});
exports.JwtVerifiedCredentialFormatEnumFromJSON = JwtVerifiedCredentialFormatEnum.JwtVerifiedCredentialFormatEnumFromJSON;
exports.JwtVerifiedCredentialFormatEnumFromJSONTyped = JwtVerifiedCredentialFormatEnum.JwtVerifiedCredentialFormatEnumFromJSONTyped;
exports.JwtVerifiedCredentialFormatEnumToJSON = JwtVerifiedCredentialFormatEnum.JwtVerifiedCredentialFormatEnumToJSON;
exports.JwtVerifiedCredentialHashesFromJSON = JwtVerifiedCredentialHashes.JwtVerifiedCredentialHashesFromJSON;
exports.JwtVerifiedCredentialHashesFromJSONTyped = JwtVerifiedCredentialHashes.JwtVerifiedCredentialHashesFromJSONTyped;
exports.JwtVerifiedCredentialHashesToJSON = JwtVerifiedCredentialHashes.JwtVerifiedCredentialHashesToJSON;
Object.defineProperty(exports, 'KycFieldType', {
	enumerable: true,
	get: function () { return KycFieldType.KycFieldType; }
});
exports.KycFieldTypeFromJSON = KycFieldType.KycFieldTypeFromJSON;
exports.KycFieldTypeFromJSONTyped = KycFieldType.KycFieldTypeFromJSONTyped;
exports.KycFieldTypeToJSON = KycFieldType.KycFieldTypeToJSON;
Object.defineProperty(exports, 'MFAAction', {
	enumerable: true,
	get: function () { return MFAAction.MFAAction; }
});
exports.MFAActionFromJSON = MFAAction.MFAActionFromJSON;
exports.MFAActionFromJSONTyped = MFAAction.MFAActionFromJSONTyped;
exports.MFAActionToJSON = MFAAction.MFAActionToJSON;
exports.MFAAuthPasskeyDeviceGetResponseFromJSON = MFAAuthPasskeyDeviceGetResponse.MFAAuthPasskeyDeviceGetResponseFromJSON;
exports.MFAAuthPasskeyDeviceGetResponseFromJSONTyped = MFAAuthPasskeyDeviceGetResponse.MFAAuthPasskeyDeviceGetResponseFromJSONTyped;
exports.MFAAuthPasskeyDeviceGetResponseToJSON = MFAAuthPasskeyDeviceGetResponse.MFAAuthPasskeyDeviceGetResponseToJSON;
exports.MFAAuthPasskeyDeviceGetResponseAssertionFromJSON = MFAAuthPasskeyDeviceGetResponseAssertion.MFAAuthPasskeyDeviceGetResponseAssertionFromJSON;
exports.MFAAuthPasskeyDeviceGetResponseAssertionFromJSONTyped = MFAAuthPasskeyDeviceGetResponseAssertion.MFAAuthPasskeyDeviceGetResponseAssertionFromJSONTyped;
exports.MFAAuthPasskeyDeviceGetResponseAssertionToJSON = MFAAuthPasskeyDeviceGetResponseAssertion.MFAAuthPasskeyDeviceGetResponseAssertionToJSON;
exports.MFAAuthPasskeyDevicePostRequestFromJSON = MFAAuthPasskeyDevicePostRequest.MFAAuthPasskeyDevicePostRequestFromJSON;
exports.MFAAuthPasskeyDevicePostRequestFromJSONTyped = MFAAuthPasskeyDevicePostRequest.MFAAuthPasskeyDevicePostRequestFromJSONTyped;
exports.MFAAuthPasskeyDevicePostRequestToJSON = MFAAuthPasskeyDevicePostRequest.MFAAuthPasskeyDevicePostRequestToJSON;
exports.MFAAuthRecoveryDevicePostRequestFromJSON = MFAAuthRecoveryDevicePostRequest.MFAAuthRecoveryDevicePostRequestFromJSON;
exports.MFAAuthRecoveryDevicePostRequestFromJSONTyped = MFAAuthRecoveryDevicePostRequest.MFAAuthRecoveryDevicePostRequestFromJSONTyped;
exports.MFAAuthRecoveryDevicePostRequestToJSON = MFAAuthRecoveryDevicePostRequest.MFAAuthRecoveryDevicePostRequestToJSON;
exports.MFAAuthTotpDevicePostRequestFromJSON = MFAAuthTotpDevicePostRequest.MFAAuthTotpDevicePostRequestFromJSON;
exports.MFAAuthTotpDevicePostRequestFromJSONTyped = MFAAuthTotpDevicePostRequest.MFAAuthTotpDevicePostRequestFromJSONTyped;
exports.MFAAuthTotpDevicePostRequestToJSON = MFAAuthTotpDevicePostRequest.MFAAuthTotpDevicePostRequestToJSON;
exports.MFADeviceFromJSON = MFADevice.MFADeviceFromJSON;
exports.MFADeviceFromJSONTyped = MFADevice.MFADeviceFromJSONTyped;
exports.MFADeviceToJSON = MFADevice.MFADeviceToJSON;
Object.defineProperty(exports, 'MFADeviceType', {
	enumerable: true,
	get: function () { return MFADeviceType.MFADeviceType; }
});
exports.MFADeviceTypeFromJSON = MFADeviceType.MFADeviceTypeFromJSON;
exports.MFADeviceTypeFromJSONTyped = MFADeviceType.MFADeviceTypeFromJSONTyped;
exports.MFADeviceTypeToJSON = MFADeviceType.MFADeviceTypeToJSON;
exports.MFAGetRecoveryCodesResponseFromJSON = MFAGetRecoveryCodesResponse.MFAGetRecoveryCodesResponseFromJSON;
exports.MFAGetRecoveryCodesResponseFromJSONTyped = MFAGetRecoveryCodesResponse.MFAGetRecoveryCodesResponseFromJSONTyped;
exports.MFAGetRecoveryCodesResponseToJSON = MFAGetRecoveryCodesResponse.MFAGetRecoveryCodesResponseToJSON;
exports.MFAListDevicesResponseFromJSON = MFAListDevicesResponse.MFAListDevicesResponseFromJSON;
exports.MFAListDevicesResponseFromJSONTyped = MFAListDevicesResponse.MFAListDevicesResponseFromJSONTyped;
exports.MFAListDevicesResponseToJSON = MFAListDevicesResponse.MFAListDevicesResponseToJSON;
exports.MFAMethodsResponseFromJSON = MFAMethodsResponse.MFAMethodsResponseFromJSON;
exports.MFAMethodsResponseFromJSONTyped = MFAMethodsResponse.MFAMethodsResponseFromJSONTyped;
exports.MFAMethodsResponseToJSON = MFAMethodsResponse.MFAMethodsResponseToJSON;
exports.MFARegenRecoveryCodesResponseFromJSON = MFARegenRecoveryCodesResponse.MFARegenRecoveryCodesResponseFromJSON;
exports.MFARegenRecoveryCodesResponseFromJSONTyped = MFARegenRecoveryCodesResponse.MFARegenRecoveryCodesResponseFromJSONTyped;
exports.MFARegenRecoveryCodesResponseToJSON = MFARegenRecoveryCodesResponse.MFARegenRecoveryCodesResponseToJSON;
exports.MFARegisterPasskeyDeviceGetResponseFromJSON = MFARegisterPasskeyDeviceGetResponse.MFARegisterPasskeyDeviceGetResponseFromJSON;
exports.MFARegisterPasskeyDeviceGetResponseFromJSONTyped = MFARegisterPasskeyDeviceGetResponse.MFARegisterPasskeyDeviceGetResponseFromJSONTyped;
exports.MFARegisterPasskeyDeviceGetResponseToJSON = MFARegisterPasskeyDeviceGetResponse.MFARegisterPasskeyDeviceGetResponseToJSON;
exports.MFARegisterPasskeyDeviceGetResponseRegistrationFromJSON = MFARegisterPasskeyDeviceGetResponseRegistration.MFARegisterPasskeyDeviceGetResponseRegistrationFromJSON;
exports.MFARegisterPasskeyDeviceGetResponseRegistrationFromJSONTyped = MFARegisterPasskeyDeviceGetResponseRegistration.MFARegisterPasskeyDeviceGetResponseRegistrationFromJSONTyped;
exports.MFARegisterPasskeyDeviceGetResponseRegistrationToJSON = MFARegisterPasskeyDeviceGetResponseRegistration.MFARegisterPasskeyDeviceGetResponseRegistrationToJSON;
exports.MFARegisterPasskeyDevicePostRequestFromJSON = MFARegisterPasskeyDevicePostRequest.MFARegisterPasskeyDevicePostRequestFromJSON;
exports.MFARegisterPasskeyDevicePostRequestFromJSONTyped = MFARegisterPasskeyDevicePostRequest.MFARegisterPasskeyDevicePostRequestFromJSONTyped;
exports.MFARegisterPasskeyDevicePostRequestToJSON = MFARegisterPasskeyDevicePostRequest.MFARegisterPasskeyDevicePostRequestToJSON;
exports.MFARegisterTotpDeviceGetResponseFromJSON = MFARegisterTotpDeviceGetResponse.MFARegisterTotpDeviceGetResponseFromJSON;
exports.MFARegisterTotpDeviceGetResponseFromJSONTyped = MFARegisterTotpDeviceGetResponse.MFARegisterTotpDeviceGetResponseFromJSONTyped;
exports.MFARegisterTotpDeviceGetResponseToJSON = MFARegisterTotpDeviceGetResponse.MFARegisterTotpDeviceGetResponseToJSON;
exports.MFARegisterTotpDevicePostRequestFromJSON = MFARegisterTotpDevicePostRequest.MFARegisterTotpDevicePostRequestFromJSON;
exports.MFARegisterTotpDevicePostRequestFromJSONTyped = MFARegisterTotpDevicePostRequest.MFARegisterTotpDevicePostRequestFromJSONTyped;
exports.MFARegisterTotpDevicePostRequestToJSON = MFARegisterTotpDevicePostRequest.MFARegisterTotpDevicePostRequestToJSON;
exports.MFASettingsFromJSON = MFASettings.MFASettingsFromJSON;
exports.MFASettingsFromJSONTyped = MFASettings.MFASettingsFromJSONTyped;
exports.MFASettingsToJSON = MFASettings.MFASettingsToJSON;
exports.MFASettingsActionsFromJSON = MFASettingsActions.MFASettingsActionsFromJSON;
exports.MFASettingsActionsFromJSONTyped = MFASettingsActions.MFASettingsActionsFromJSONTyped;
exports.MFASettingsActionsToJSON = MFASettingsActions.MFASettingsActionsToJSON;
exports.MFASettingsMethodsFromJSON = MFASettingsMethods.MFASettingsMethodsFromJSON;
exports.MFASettingsMethodsFromJSONTyped = MFASettingsMethods.MFASettingsMethodsFromJSONTyped;
exports.MFASettingsMethodsToJSON = MFASettingsMethods.MFASettingsMethodsToJSON;
exports.MFAUpdateDeviceRequestFromJSON = MFAUpdateDeviceRequest.MFAUpdateDeviceRequestFromJSON;
exports.MFAUpdateDeviceRequestFromJSONTyped = MFAUpdateDeviceRequest.MFAUpdateDeviceRequestFromJSONTyped;
exports.MFAUpdateDeviceRequestToJSON = MFAUpdateDeviceRequest.MFAUpdateDeviceRequestToJSON;
exports.MergeConflictsFromJSON = MergeConflicts.MergeConflictsFromJSON;
exports.MergeConflictsFromJSONTyped = MergeConflicts.MergeConflictsFromJSONTyped;
exports.MergeConflictsToJSON = MergeConflicts.MergeConflictsToJSON;
exports.MergeUserFromJSON = MergeUser.MergeUserFromJSON;
exports.MergeUserFromJSONTyped = MergeUser.MergeUserFromJSONTyped;
exports.MergeUserToJSON = MergeUser.MergeUserToJSON;
exports.MergeUserConflictFromJSON = MergeUserConflict.MergeUserConflictFromJSON;
exports.MergeUserConflictFromJSONTyped = MergeUserConflict.MergeUserConflictFromJSONTyped;
exports.MergeUserConflictToJSON = MergeUserConflict.MergeUserConflictToJSON;
exports.MergeUserConflictResolutionFromJSON = MergeUserConflictResolution.MergeUserConflictResolutionFromJSON;
exports.MergeUserConflictResolutionFromJSONTyped = MergeUserConflictResolution.MergeUserConflictResolutionFromJSONTyped;
exports.MergeUserConflictResolutionToJSON = MergeUserConflictResolution.MergeUserConflictResolutionToJSON;
exports.MergeUserConflictResolutionsFromJSON = MergeUserConflictResolutions.MergeUserConflictResolutionsFromJSON;
exports.MergeUserConflictResolutionsFromJSONTyped = MergeUserConflictResolutions.MergeUserConflictResolutionsFromJSONTyped;
exports.MergeUserConflictResolutionsToJSON = MergeUserConflictResolutions.MergeUserConflictResolutionsToJSON;
exports.MethodNotAllowedFromJSON = MethodNotAllowed.MethodNotAllowedFromJSON;
exports.MethodNotAllowedFromJSONTyped = MethodNotAllowed.MethodNotAllowedFromJSONTyped;
exports.MethodNotAllowedToJSON = MethodNotAllowed.MethodNotAllowedToJSON;
Object.defineProperty(exports, 'MfaBackupCodeAcknowledgement', {
	enumerable: true,
	get: function () { return MfaBackupCodeAcknowledgement.MfaBackupCodeAcknowledgement; }
});
exports.MfaBackupCodeAcknowledgementFromJSON = MfaBackupCodeAcknowledgement.MfaBackupCodeAcknowledgementFromJSON;
exports.MfaBackupCodeAcknowledgementFromJSONTyped = MfaBackupCodeAcknowledgement.MfaBackupCodeAcknowledgementFromJSONTyped;
exports.MfaBackupCodeAcknowledgementToJSON = MfaBackupCodeAcknowledgement.MfaBackupCodeAcknowledgementToJSON;
exports.MinifiedDynamicJwtFromJSON = MinifiedDynamicJwt.MinifiedDynamicJwtFromJSON;
exports.MinifiedDynamicJwtFromJSONTyped = MinifiedDynamicJwt.MinifiedDynamicJwtFromJSONTyped;
exports.MinifiedDynamicJwtToJSON = MinifiedDynamicJwt.MinifiedDynamicJwtToJSON;
exports.MobileSettingsFromJSON = MobileSettings.MobileSettingsFromJSON;
exports.MobileSettingsFromJSONTyped = MobileSettings.MobileSettingsFromJSONTyped;
exports.MobileSettingsToJSON = MobileSettings.MobileSettingsToJSON;
exports.MultichainAccountBalanceResponseFromJSON = MultichainAccountBalanceResponse.MultichainAccountBalanceResponseFromJSON;
exports.MultichainAccountBalanceResponseFromJSONTyped = MultichainAccountBalanceResponse.MultichainAccountBalanceResponseFromJSONTyped;
exports.MultichainAccountBalanceResponseToJSON = MultichainAccountBalanceResponse.MultichainAccountBalanceResponseToJSON;
exports.MultichainAccountBalanceResponseChainBalancesFromJSON = MultichainAccountBalanceResponseChainBalances.MultichainAccountBalanceResponseChainBalancesFromJSON;
exports.MultichainAccountBalanceResponseChainBalancesFromJSONTyped = MultichainAccountBalanceResponseChainBalances.MultichainAccountBalanceResponseChainBalancesFromJSONTyped;
exports.MultichainAccountBalanceResponseChainBalancesToJSON = MultichainAccountBalanceResponseChainBalances.MultichainAccountBalanceResponseChainBalancesToJSON;
exports.MultichainAccountBalanceResponseNetworksFromJSON = MultichainAccountBalanceResponseNetworks.MultichainAccountBalanceResponseNetworksFromJSON;
exports.MultichainAccountBalanceResponseNetworksFromJSONTyped = MultichainAccountBalanceResponseNetworks.MultichainAccountBalanceResponseNetworksFromJSONTyped;
exports.MultichainAccountBalanceResponseNetworksToJSON = MultichainAccountBalanceResponseNetworks.MultichainAccountBalanceResponseNetworksToJSON;
exports.MultichainAccountBalancesRequestFromJSON = MultichainAccountBalancesRequest.MultichainAccountBalancesRequestFromJSON;
exports.MultichainAccountBalancesRequestFromJSONTyped = MultichainAccountBalancesRequest.MultichainAccountBalancesRequestFromJSONTyped;
exports.MultichainAccountBalancesRequestToJSON = MultichainAccountBalancesRequest.MultichainAccountBalancesRequestToJSON;
exports.MultichainAccountBalancesRequestBalanceRequestsFromJSON = MultichainAccountBalancesRequestBalanceRequests.MultichainAccountBalancesRequestBalanceRequestsFromJSON;
exports.MultichainAccountBalancesRequestBalanceRequestsFromJSONTyped = MultichainAccountBalancesRequestBalanceRequests.MultichainAccountBalancesRequestBalanceRequestsFromJSONTyped;
exports.MultichainAccountBalancesRequestBalanceRequestsToJSON = MultichainAccountBalancesRequestBalanceRequests.MultichainAccountBalancesRequestBalanceRequestsToJSON;
exports.NameServiceFromJSON = NameService.NameServiceFromJSON;
exports.NameServiceFromJSONTyped = NameService.NameServiceFromJSONTyped;
exports.NameServiceToJSON = NameService.NameServiceToJSON;
exports.NameServiceDataFromJSON = NameServiceData.NameServiceDataFromJSON;
exports.NameServiceDataFromJSONTyped = NameServiceData.NameServiceDataFromJSONTyped;
exports.NameServiceDataToJSON = NameServiceData.NameServiceDataToJSON;
exports.NameServiceSdkSettingsFromJSON = NameServiceSdkSettings.NameServiceSdkSettingsFromJSON;
exports.NameServiceSdkSettingsFromJSONTyped = NameServiceSdkSettings.NameServiceSdkSettingsFromJSONTyped;
exports.NameServiceSdkSettingsToJSON = NameServiceSdkSettings.NameServiceSdkSettingsToJSON;
exports.NameServiceSdkSettingsEvmFromJSON = NameServiceSdkSettingsEvm.NameServiceSdkSettingsEvmFromJSON;
exports.NameServiceSdkSettingsEvmFromJSONTyped = NameServiceSdkSettingsEvm.NameServiceSdkSettingsEvmFromJSONTyped;
exports.NameServiceSdkSettingsEvmToJSON = NameServiceSdkSettingsEvm.NameServiceSdkSettingsEvmToJSON;
exports.NativeCurrencyFromJSON = NativeCurrency.NativeCurrencyFromJSON;
exports.NativeCurrencyFromJSONTyped = NativeCurrency.NativeCurrencyFromJSONTyped;
exports.NativeCurrencyToJSON = NativeCurrency.NativeCurrencyToJSON;
exports.NetworkFromJSON = Network.NetworkFromJSON;
exports.NetworkFromJSONTyped = Network.NetworkFromJSONTyped;
exports.NetworkToJSON = Network.NetworkToJSON;
Object.defineProperty(exports, 'NetworkTypeEnum', {
	enumerable: true,
	get: function () { return Network.NetworkTypeEnum; }
});
exports.NetworkConfigurationFromJSON = NetworkConfiguration.NetworkConfigurationFromJSON;
exports.NetworkConfigurationFromJSONTyped = NetworkConfiguration.NetworkConfigurationFromJSONTyped;
exports.NetworkConfigurationToJSON = NetworkConfiguration.NetworkConfigurationToJSON;
exports.NetworkConfigurationResponseFromJSON = NetworkConfigurationResponse.NetworkConfigurationResponseFromJSON;
exports.NetworkConfigurationResponseFromJSONTyped = NetworkConfigurationResponse.NetworkConfigurationResponseFromJSONTyped;
exports.NetworkConfigurationResponseToJSON = NetworkConfigurationResponse.NetworkConfigurationResponseToJSON;
exports.NextJsSettingsFromJSON = NextJsSettings.NextJsSettingsFromJSON;
exports.NextJsSettingsFromJSONTyped = NextJsSettings.NextJsSettingsFromJSONTyped;
exports.NextJsSettingsToJSON = NextJsSettings.NextJsSettingsToJSON;
Object.defineProperty(exports, 'NextViewEnum', {
	enumerable: true,
	get: function () { return NextViewEnum.NextViewEnum; }
});
exports.NextViewEnumFromJSON = NextViewEnum.NextViewEnumFromJSON;
exports.NextViewEnumFromJSONTyped = NextViewEnum.NextViewEnumFromJSONTyped;
exports.NextViewEnumToJSON = NextViewEnum.NextViewEnumToJSON;
exports.NonceResponseFromJSON = NonceResponse.NonceResponseFromJSON;
exports.NonceResponseFromJSONTyped = NonceResponse.NonceResponseFromJSONTyped;
exports.NonceResponseToJSON = NonceResponse.NonceResponseToJSON;
exports.NotFoundFromJSON = NotFound.NotFoundFromJSON;
exports.NotFoundFromJSONTyped = NotFound.NotFoundFromJSONTyped;
exports.NotFoundToJSON = NotFound.NotFoundToJSON;
exports.OAuthAccountFromJSON = OAuthAccount.OAuthAccountFromJSON;
exports.OAuthAccountFromJSONTyped = OAuthAccount.OAuthAccountFromJSONTyped;
exports.OAuthAccountToJSON = OAuthAccount.OAuthAccountToJSON;
exports.OauthInitAuthRequestFromJSON = OauthInitAuthRequest.OauthInitAuthRequestFromJSON;
exports.OauthInitAuthRequestFromJSONTyped = OauthInitAuthRequest.OauthInitAuthRequestFromJSONTyped;
exports.OauthInitAuthRequestToJSON = OauthInitAuthRequest.OauthInitAuthRequestToJSON;
exports.OauthProviderLoginUrlFromJSON = OauthProviderLoginUrl.OauthProviderLoginUrlFromJSON;
exports.OauthProviderLoginUrlFromJSONTyped = OauthProviderLoginUrl.OauthProviderLoginUrlFromJSONTyped;
exports.OauthProviderLoginUrlToJSON = OauthProviderLoginUrl.OauthProviderLoginUrlToJSON;
exports.OauthProviderRequestFromJSON = OauthProviderRequest.OauthProviderRequestFromJSON;
exports.OauthProviderRequestFromJSONTyped = OauthProviderRequest.OauthProviderRequestFromJSONTyped;
exports.OauthProviderRequestToJSON = OauthProviderRequest.OauthProviderRequestToJSON;
exports.OauthRequestFromJSON = OauthRequest.OauthRequestFromJSON;
exports.OauthRequestFromJSONTyped = OauthRequest.OauthRequestFromJSONTyped;
exports.OauthRequestToJSON = OauthRequest.OauthRequestToJSON;
exports.OauthResultRequestFromJSON = OauthResultRequest.OauthResultRequestFromJSON;
exports.OauthResultRequestFromJSONTyped = OauthResultRequest.OauthResultRequestFromJSONTyped;
exports.OauthResultRequestToJSON = OauthResultRequest.OauthResultRequestToJSON;
exports.OauthResultResponseFromJSON = OauthResultResponse.OauthResultResponseFromJSON;
exports.OauthResultResponseFromJSONTyped = OauthResultResponse.OauthResultResponseFromJSONTyped;
exports.OauthResultResponseToJSON = OauthResultResponse.OauthResultResponseToJSON;
Object.defineProperty(exports, 'OauthResultStatus', {
	enumerable: true,
	get: function () { return OauthResultStatus.OauthResultStatus; }
});
exports.OauthResultStatusFromJSON = OauthResultStatus.OauthResultStatusFromJSON;
exports.OauthResultStatusFromJSONTyped = OauthResultStatus.OauthResultStatusFromJSONTyped;
exports.OauthResultStatusToJSON = OauthResultStatus.OauthResultStatusToJSON;
Object.defineProperty(exports, 'OnrampProviders', {
	enumerable: true,
	get: function () { return OnrampProviders.OnrampProviders; }
});
exports.OnrampProvidersFromJSON = OnrampProviders.OnrampProvidersFromJSON;
exports.OnrampProvidersFromJSONTyped = OnrampProviders.OnrampProvidersFromJSONTyped;
exports.OnrampProvidersToJSON = OnrampProviders.OnrampProvidersToJSON;
exports.OpenRoomResponseFromJSON = OpenRoomResponse.OpenRoomResponseFromJSON;
exports.OpenRoomResponseFromJSONTyped = OpenRoomResponse.OpenRoomResponseFromJSONTyped;
exports.OpenRoomResponseToJSON = OpenRoomResponse.OpenRoomResponseToJSON;
exports.OpenRoomResponseForReshareFromJSON = OpenRoomResponseForReshare.OpenRoomResponseForReshareFromJSON;
exports.OpenRoomResponseForReshareFromJSONTyped = OpenRoomResponseForReshare.OpenRoomResponseForReshareFromJSONTyped;
exports.OpenRoomResponseForReshareToJSON = OpenRoomResponseForReshare.OpenRoomResponseForReshareToJSON;
exports.OpenRoomResponseWithServerKeygenIdsFromJSON = OpenRoomResponseWithServerKeygenIds.OpenRoomResponseWithServerKeygenIdsFromJSON;
exports.OpenRoomResponseWithServerKeygenIdsFromJSONTyped = OpenRoomResponseWithServerKeygenIds.OpenRoomResponseWithServerKeygenIdsFromJSONTyped;
exports.OpenRoomResponseWithServerKeygenIdsToJSON = OpenRoomResponseWithServerKeygenIds.OpenRoomResponseWithServerKeygenIdsToJSON;
exports.PasskeyAllowCredentialFromJSON = PasskeyAllowCredential.PasskeyAllowCredentialFromJSON;
exports.PasskeyAllowCredentialFromJSONTyped = PasskeyAllowCredential.PasskeyAllowCredentialFromJSONTyped;
exports.PasskeyAllowCredentialToJSON = PasskeyAllowCredential.PasskeyAllowCredentialToJSON;
exports.PasskeyAuthRequestFromJSON = PasskeyAuthRequest.PasskeyAuthRequestFromJSON;
exports.PasskeyAuthRequestFromJSONTyped = PasskeyAuthRequest.PasskeyAuthRequestFromJSONTyped;
exports.PasskeyAuthRequestToJSON = PasskeyAuthRequest.PasskeyAuthRequestToJSON;
Object.defineProperty(exports, 'PasskeyCredentialHint', {
	enumerable: true,
	get: function () { return PasskeyCredentialHint.PasskeyCredentialHint; }
});
exports.PasskeyCredentialHintFromJSON = PasskeyCredentialHint.PasskeyCredentialHintFromJSON;
exports.PasskeyCredentialHintFromJSONTyped = PasskeyCredentialHint.PasskeyCredentialHintFromJSONTyped;
exports.PasskeyCredentialHintToJSON = PasskeyCredentialHint.PasskeyCredentialHintToJSON;
Object.defineProperty(exports, 'PasskeyCredentialType', {
	enumerable: true,
	get: function () { return PasskeyCredentialType.PasskeyCredentialType; }
});
exports.PasskeyCredentialTypeFromJSON = PasskeyCredentialType.PasskeyCredentialTypeFromJSON;
exports.PasskeyCredentialTypeFromJSONTyped = PasskeyCredentialType.PasskeyCredentialTypeFromJSONTyped;
exports.PasskeyCredentialTypeToJSON = PasskeyCredentialType.PasskeyCredentialTypeToJSON;
exports.PasskeyExcludeCredentialFromJSON = PasskeyExcludeCredential.PasskeyExcludeCredentialFromJSON;
exports.PasskeyExcludeCredentialFromJSONTyped = PasskeyExcludeCredential.PasskeyExcludeCredentialFromJSONTyped;
exports.PasskeyExcludeCredentialToJSON = PasskeyExcludeCredential.PasskeyExcludeCredentialToJSON;
exports.PasskeyExtensionFromJSON = PasskeyExtension.PasskeyExtensionFromJSON;
exports.PasskeyExtensionFromJSONTyped = PasskeyExtension.PasskeyExtensionFromJSONTyped;
exports.PasskeyExtensionToJSON = PasskeyExtension.PasskeyExtensionToJSON;
exports.PasskeyRegisterRequestFromJSON = PasskeyRegisterRequest.PasskeyRegisterRequestFromJSON;
exports.PasskeyRegisterRequestFromJSONTyped = PasskeyRegisterRequest.PasskeyRegisterRequestFromJSONTyped;
exports.PasskeyRegisterRequestToJSON = PasskeyRegisterRequest.PasskeyRegisterRequestToJSON;
exports.PasskeyRegisterVerifyRequestFromJSON = PasskeyRegisterVerifyRequest.PasskeyRegisterVerifyRequestFromJSON;
exports.PasskeyRegisterVerifyRequestFromJSONTyped = PasskeyRegisterVerifyRequest.PasskeyRegisterVerifyRequestFromJSONTyped;
exports.PasskeyRegisterVerifyRequestToJSON = PasskeyRegisterVerifyRequest.PasskeyRegisterVerifyRequestToJSON;
exports.PasskeyRegistrationCredentialFromJSON = PasskeyRegistrationCredential.PasskeyRegistrationCredentialFromJSON;
exports.PasskeyRegistrationCredentialFromJSONTyped = PasskeyRegistrationCredential.PasskeyRegistrationCredentialFromJSONTyped;
exports.PasskeyRegistrationCredentialToJSON = PasskeyRegistrationCredential.PasskeyRegistrationCredentialToJSON;
exports.PasskeyRegistrationCredentialV2FromJSON = PasskeyRegistrationCredentialV2.PasskeyRegistrationCredentialV2FromJSON;
exports.PasskeyRegistrationCredentialV2FromJSONTyped = PasskeyRegistrationCredentialV2.PasskeyRegistrationCredentialV2FromJSONTyped;
exports.PasskeyRegistrationCredentialV2ToJSON = PasskeyRegistrationCredentialV2.PasskeyRegistrationCredentialV2ToJSON;
exports.PasskeyRelyingPartyFromJSON = PasskeyRelyingParty.PasskeyRelyingPartyFromJSON;
exports.PasskeyRelyingPartyFromJSONTyped = PasskeyRelyingParty.PasskeyRelyingPartyFromJSONTyped;
exports.PasskeyRelyingPartyToJSON = PasskeyRelyingParty.PasskeyRelyingPartyToJSON;
exports.PasskeyStorageFromJSON = PasskeyStorage.PasskeyStorageFromJSON;
exports.PasskeyStorageFromJSONTyped = PasskeyStorage.PasskeyStorageFromJSONTyped;
exports.PasskeyStorageToJSON = PasskeyStorage.PasskeyStorageToJSON;
Object.defineProperty(exports, 'PasskeyTransport', {
	enumerable: true,
	get: function () { return PasskeyTransport.PasskeyTransport; }
});
exports.PasskeyTransportFromJSON = PasskeyTransport.PasskeyTransportFromJSON;
exports.PasskeyTransportFromJSONTyped = PasskeyTransport.PasskeyTransportFromJSONTyped;
exports.PasskeyTransportToJSON = PasskeyTransport.PasskeyTransportToJSON;
exports.PasskeyUserFromJSON = PasskeyUser.PasskeyUserFromJSON;
exports.PasskeyUserFromJSONTyped = PasskeyUser.PasskeyUserFromJSONTyped;
exports.PasskeyUserToJSON = PasskeyUser.PasskeyUserToJSON;
Object.defineProperty(exports, 'PasswordSourceTypeEnum', {
	enumerable: true,
	get: function () { return PasswordSourceTypeEnum.PasswordSourceTypeEnum; }
});
exports.PasswordSourceTypeEnumFromJSON = PasswordSourceTypeEnum.PasswordSourceTypeEnumFromJSON;
exports.PasswordSourceTypeEnumFromJSONTyped = PasswordSourceTypeEnum.PasswordSourceTypeEnumFromJSONTyped;
exports.PasswordSourceTypeEnumToJSON = PasswordSourceTypeEnum.PasswordSourceTypeEnumToJSON;
exports.PrefetchRequestFromJSON = PrefetchRequest.PrefetchRequestFromJSON;
exports.PrefetchRequestFromJSONTyped = PrefetchRequest.PrefetchRequestFromJSONTyped;
exports.PrefetchRequestToJSON = PrefetchRequest.PrefetchRequestToJSON;
exports.PriceDataFromJSON = PriceData.PriceDataFromJSON;
exports.PriceDataFromJSONTyped = PriceData.PriceDataFromJSONTyped;
exports.PriceDataToJSON = PriceData.PriceDataToJSON;
exports.ProjectSettingsFromJSON = ProjectSettings.ProjectSettingsFromJSON;
exports.ProjectSettingsFromJSONTyped = ProjectSettings.ProjectSettingsFromJSONTyped;
exports.ProjectSettingsToJSON = ProjectSettings.ProjectSettingsToJSON;
exports.ProjectSettingsChainsFromJSON = ProjectSettingsChains.ProjectSettingsChainsFromJSON;
exports.ProjectSettingsChainsFromJSONTyped = ProjectSettingsChains.ProjectSettingsChainsFromJSONTyped;
exports.ProjectSettingsChainsToJSON = ProjectSettingsChains.ProjectSettingsChainsToJSON;
exports.ProjectSettingsDesignFromJSON = ProjectSettingsDesign.ProjectSettingsDesignFromJSON;
exports.ProjectSettingsDesignFromJSONTyped = ProjectSettingsDesign.ProjectSettingsDesignFromJSONTyped;
exports.ProjectSettingsDesignToJSON = ProjectSettingsDesign.ProjectSettingsDesignToJSON;
exports.ProjectSettingsDesignButtonFromJSON = ProjectSettingsDesignButton.ProjectSettingsDesignButtonFromJSON;
exports.ProjectSettingsDesignButtonFromJSONTyped = ProjectSettingsDesignButton.ProjectSettingsDesignButtonFromJSONTyped;
exports.ProjectSettingsDesignButtonToJSON = ProjectSettingsDesignButton.ProjectSettingsDesignButtonToJSON;
exports.ProjectSettingsDesignModalFromJSON = ProjectSettingsDesignModal.ProjectSettingsDesignModalFromJSON;
exports.ProjectSettingsDesignModalFromJSONTyped = ProjectSettingsDesignModal.ProjectSettingsDesignModalFromJSONTyped;
exports.ProjectSettingsDesignModalToJSON = ProjectSettingsDesignModal.ProjectSettingsDesignModalToJSON;
exports.ProjectSettingsDesignWidgetFromJSON = ProjectSettingsDesignWidget.ProjectSettingsDesignWidgetFromJSON;
exports.ProjectSettingsDesignWidgetFromJSONTyped = ProjectSettingsDesignWidget.ProjectSettingsDesignWidgetFromJSONTyped;
exports.ProjectSettingsDesignWidgetToJSON = ProjectSettingsDesignWidget.ProjectSettingsDesignWidgetToJSON;
exports.ProjectSettingsGeneralFromJSON = ProjectSettingsGeneral.ProjectSettingsGeneralFromJSON;
exports.ProjectSettingsGeneralFromJSONTyped = ProjectSettingsGeneral.ProjectSettingsGeneralFromJSONTyped;
exports.ProjectSettingsGeneralToJSON = ProjectSettingsGeneral.ProjectSettingsGeneralToJSON;
exports.ProjectSettingsKycFromJSON = ProjectSettingsKyc.ProjectSettingsKycFromJSON;
exports.ProjectSettingsKycFromJSONTyped = ProjectSettingsKyc.ProjectSettingsKycFromJSONTyped;
exports.ProjectSettingsKycToJSON = ProjectSettingsKyc.ProjectSettingsKycToJSON;
exports.ProjectSettingsPrivacyFromJSON = ProjectSettingsPrivacy.ProjectSettingsPrivacyFromJSON;
exports.ProjectSettingsPrivacyFromJSONTyped = ProjectSettingsPrivacy.ProjectSettingsPrivacyFromJSONTyped;
exports.ProjectSettingsPrivacyToJSON = ProjectSettingsPrivacy.ProjectSettingsPrivacyToJSON;
exports.ProjectSettingsSdkFromJSON = ProjectSettingsSdk.ProjectSettingsSdkFromJSON;
exports.ProjectSettingsSdkFromJSONTyped = ProjectSettingsSdk.ProjectSettingsSdkFromJSONTyped;
exports.ProjectSettingsSdkToJSON = ProjectSettingsSdk.ProjectSettingsSdkToJSON;
exports.ProjectSettingsSdkAccountAbstractionFromJSON = ProjectSettingsSdkAccountAbstraction.ProjectSettingsSdkAccountAbstractionFromJSON;
exports.ProjectSettingsSdkAccountAbstractionFromJSONTyped = ProjectSettingsSdkAccountAbstraction.ProjectSettingsSdkAccountAbstractionFromJSONTyped;
exports.ProjectSettingsSdkAccountAbstractionToJSON = ProjectSettingsSdkAccountAbstraction.ProjectSettingsSdkAccountAbstractionToJSON;
exports.ProjectSettingsSdkEmailSignInFromJSON = ProjectSettingsSdkEmailSignIn.ProjectSettingsSdkEmailSignInFromJSON;
exports.ProjectSettingsSdkEmailSignInFromJSONTyped = ProjectSettingsSdkEmailSignIn.ProjectSettingsSdkEmailSignInFromJSONTyped;
exports.ProjectSettingsSdkEmailSignInToJSON = ProjectSettingsSdkEmailSignIn.ProjectSettingsSdkEmailSignInToJSON;
exports.ProjectSettingsSdkEmbeddedWalletsFromJSON = ProjectSettingsSdkEmbeddedWallets.ProjectSettingsSdkEmbeddedWalletsFromJSON;
exports.ProjectSettingsSdkEmbeddedWalletsFromJSONTyped = ProjectSettingsSdkEmbeddedWallets.ProjectSettingsSdkEmbeddedWalletsFromJSONTyped;
exports.ProjectSettingsSdkEmbeddedWalletsToJSON = ProjectSettingsSdkEmbeddedWallets.ProjectSettingsSdkEmbeddedWalletsToJSON;
exports.ProjectSettingsSdkSocialSignInFromJSON = ProjectSettingsSdkSocialSignIn.ProjectSettingsSdkSocialSignInFromJSON;
exports.ProjectSettingsSdkSocialSignInFromJSONTyped = ProjectSettingsSdkSocialSignIn.ProjectSettingsSdkSocialSignInFromJSONTyped;
exports.ProjectSettingsSdkSocialSignInToJSON = ProjectSettingsSdkSocialSignIn.ProjectSettingsSdkSocialSignInToJSON;
exports.ProjectSettingsSdkWaasFromJSON = ProjectSettingsSdkWaas.ProjectSettingsSdkWaasFromJSON;
exports.ProjectSettingsSdkWaasFromJSONTyped = ProjectSettingsSdkWaas.ProjectSettingsSdkWaasFromJSONTyped;
exports.ProjectSettingsSdkWaasToJSON = ProjectSettingsSdkWaas.ProjectSettingsSdkWaasToJSON;
exports.ProjectSettingsSdkWaasDelegatedAccessFromJSON = ProjectSettingsSdkWaasDelegatedAccess.ProjectSettingsSdkWaasDelegatedAccessFromJSON;
exports.ProjectSettingsSdkWaasDelegatedAccessFromJSONTyped = ProjectSettingsSdkWaasDelegatedAccess.ProjectSettingsSdkWaasDelegatedAccessFromJSONTyped;
exports.ProjectSettingsSdkWaasDelegatedAccessToJSON = ProjectSettingsSdkWaasDelegatedAccess.ProjectSettingsSdkWaasDelegatedAccessToJSON;
exports.ProjectSettingsSdkWaasOnSignUpFromJSON = ProjectSettingsSdkWaasOnSignUp.ProjectSettingsSdkWaasOnSignUpFromJSON;
exports.ProjectSettingsSdkWaasOnSignUpFromJSONTyped = ProjectSettingsSdkWaasOnSignUp.ProjectSettingsSdkWaasOnSignUpFromJSONTyped;
exports.ProjectSettingsSdkWaasOnSignUpToJSON = ProjectSettingsSdkWaasOnSignUp.ProjectSettingsSdkWaasOnSignUpToJSON;
exports.ProjectSettingsSdkWalletConnectFromJSON = ProjectSettingsSdkWalletConnect.ProjectSettingsSdkWalletConnectFromJSON;
exports.ProjectSettingsSdkWalletConnectFromJSONTyped = ProjectSettingsSdkWalletConnect.ProjectSettingsSdkWalletConnectFromJSONTyped;
exports.ProjectSettingsSdkWalletConnectToJSON = ProjectSettingsSdkWalletConnect.ProjectSettingsSdkWalletConnectToJSON;
exports.ProjectSettingsSecurityFromJSON = ProjectSettingsSecurity.ProjectSettingsSecurityFromJSON;
exports.ProjectSettingsSecurityFromJSONTyped = ProjectSettingsSecurity.ProjectSettingsSecurityFromJSONTyped;
exports.ProjectSettingsSecurityToJSON = ProjectSettingsSecurity.ProjectSettingsSecurityToJSON;
exports.ProviderFromJSON = Provider.ProviderFromJSON;
exports.ProviderFromJSONTyped = Provider.ProviderFromJSONTyped;
exports.ProviderToJSON = Provider.ProviderToJSON;
exports.ProviderAgreementFromJSON = ProviderAgreement.ProviderAgreementFromJSON;
exports.ProviderAgreementFromJSONTyped = ProviderAgreement.ProviderAgreementFromJSONTyped;
exports.ProviderAgreementToJSON = ProviderAgreement.ProviderAgreementToJSON;
Object.defineProperty(exports, 'ProviderEntryPointVersionEnum', {
	enumerable: true,
	get: function () { return ProviderEntryPointVersionEnum.ProviderEntryPointVersionEnum; }
});
exports.ProviderEntryPointVersionEnumFromJSON = ProviderEntryPointVersionEnum.ProviderEntryPointVersionEnumFromJSON;
exports.ProviderEntryPointVersionEnumFromJSONTyped = ProviderEntryPointVersionEnum.ProviderEntryPointVersionEnumFromJSONTyped;
exports.ProviderEntryPointVersionEnumToJSON = ProviderEntryPointVersionEnum.ProviderEntryPointVersionEnumToJSON;
Object.defineProperty(exports, 'ProviderEnum', {
	enumerable: true,
	get: function () { return ProviderEnum.ProviderEnum; }
});
exports.ProviderEnumFromJSON = ProviderEnum.ProviderEnumFromJSON;
exports.ProviderEnumFromJSONTyped = ProviderEnum.ProviderEnumFromJSONTyped;
exports.ProviderEnumToJSON = ProviderEnum.ProviderEnumToJSON;
Object.defineProperty(exports, 'ProviderKernelVersionEnum', {
	enumerable: true,
	get: function () { return ProviderKernelVersionEnum.ProviderKernelVersionEnum; }
});
exports.ProviderKernelVersionEnumFromJSON = ProviderKernelVersionEnum.ProviderKernelVersionEnumFromJSON;
exports.ProviderKernelVersionEnumFromJSONTyped = ProviderKernelVersionEnum.ProviderKernelVersionEnumFromJSONTyped;
exports.ProviderKernelVersionEnumToJSON = ProviderKernelVersionEnum.ProviderKernelVersionEnumToJSON;
exports.ProviderMultichainAccountAbstractionProvidersFromJSON = ProviderMultichainAccountAbstractionProviders.ProviderMultichainAccountAbstractionProvidersFromJSON;
exports.ProviderMultichainAccountAbstractionProvidersFromJSONTyped = ProviderMultichainAccountAbstractionProviders.ProviderMultichainAccountAbstractionProvidersFromJSONTyped;
exports.ProviderMultichainAccountAbstractionProvidersToJSON = ProviderMultichainAccountAbstractionProviders.ProviderMultichainAccountAbstractionProvidersToJSON;
exports.PublicKeyCredentialDescriptorFromJSON = PublicKeyCredentialDescriptor.PublicKeyCredentialDescriptorFromJSON;
exports.PublicKeyCredentialDescriptorFromJSONTyped = PublicKeyCredentialDescriptor.PublicKeyCredentialDescriptorFromJSONTyped;
exports.PublicKeyCredentialDescriptorToJSON = PublicKeyCredentialDescriptor.PublicKeyCredentialDescriptorToJSON;
exports.PublicKeyCredentialParametersFromJSON = PublicKeyCredentialParameters.PublicKeyCredentialParametersFromJSON;
exports.PublicKeyCredentialParametersFromJSONTyped = PublicKeyCredentialParameters.PublicKeyCredentialParametersFromJSONTyped;
exports.PublicKeyCredentialParametersToJSON = PublicKeyCredentialParameters.PublicKeyCredentialParametersToJSON;
exports.PublicKeyCredentialRpEntityFromJSON = PublicKeyCredentialRpEntity.PublicKeyCredentialRpEntityFromJSON;
exports.PublicKeyCredentialRpEntityFromJSONTyped = PublicKeyCredentialRpEntity.PublicKeyCredentialRpEntityFromJSONTyped;
exports.PublicKeyCredentialRpEntityToJSON = PublicKeyCredentialRpEntity.PublicKeyCredentialRpEntityToJSON;
Object.defineProperty(exports, 'PublicKeyCredentialType', {
	enumerable: true,
	get: function () { return PublicKeyCredentialType.PublicKeyCredentialType; }
});
exports.PublicKeyCredentialTypeFromJSON = PublicKeyCredentialType.PublicKeyCredentialTypeFromJSON;
exports.PublicKeyCredentialTypeFromJSONTyped = PublicKeyCredentialType.PublicKeyCredentialTypeFromJSONTyped;
exports.PublicKeyCredentialTypeToJSON = PublicKeyCredentialType.PublicKeyCredentialTypeToJSON;
exports.PublicKeyCredentialUserEntityFromJSON = PublicKeyCredentialUserEntity.PublicKeyCredentialUserEntityFromJSON;
exports.PublicKeyCredentialUserEntityFromJSONTyped = PublicKeyCredentialUserEntity.PublicKeyCredentialUserEntityFromJSONTyped;
exports.PublicKeyCredentialUserEntityToJSON = PublicKeyCredentialUserEntity.PublicKeyCredentialUserEntityToJSON;
exports.PublishEventsFromJSON = PublishEvents.PublishEventsFromJSON;
exports.PublishEventsFromJSONTyped = PublishEvents.PublishEventsFromJSONTyped;
exports.PublishEventsToJSON = PublishEvents.PublishEventsToJSON;
exports.PublishEventsEventsFromJSON = PublishEventsEvents.PublishEventsEventsFromJSON;
exports.PublishEventsEventsFromJSONTyped = PublishEventsEvents.PublishEventsEventsFromJSONTyped;
exports.PublishEventsEventsToJSON = PublishEventsEvents.PublishEventsEventsToJSON;
Object.defineProperty(exports, 'PublishEventsEventsTypeEnum', {
	enumerable: true,
	get: function () { return PublishEventsEvents.PublishEventsEventsTypeEnum; }
});
exports.RampConfigurationFromJSON = RampConfiguration.RampConfigurationFromJSON;
exports.RampConfigurationFromJSONTyped = RampConfiguration.RampConfigurationFromJSONTyped;
exports.RampConfigurationToJSON = RampConfiguration.RampConfigurationToJSON;
exports.ReactSettingsFromJSON = ReactSettings.ReactSettingsFromJSON;
exports.ReactSettingsFromJSONTyped = ReactSettings.ReactSettingsFromJSONTyped;
exports.ReactSettingsToJSON = ReactSettings.ReactSettingsToJSON;
exports.RecoverMultipleClientKeySharesRequestFromJSON = RecoverMultipleClientKeySharesRequest.RecoverMultipleClientKeySharesRequestFromJSON;
exports.RecoverMultipleClientKeySharesRequestFromJSONTyped = RecoverMultipleClientKeySharesRequest.RecoverMultipleClientKeySharesRequestFromJSONTyped;
exports.RecoverMultipleClientKeySharesRequestToJSON = RecoverMultipleClientKeySharesRequest.RecoverMultipleClientKeySharesRequestToJSON;
exports.RecoverMultipleClientKeySharesResponseFromJSON = RecoverMultipleClientKeySharesResponse.RecoverMultipleClientKeySharesResponseFromJSON;
exports.RecoverMultipleClientKeySharesResponseFromJSONTyped = RecoverMultipleClientKeySharesResponse.RecoverMultipleClientKeySharesResponseFromJSONTyped;
exports.RecoverMultipleClientKeySharesResponseToJSON = RecoverMultipleClientKeySharesResponse.RecoverMultipleClientKeySharesResponseToJSON;
exports.RegisterEmbeddedWalletSessionKeyResponseFromJSON = RegisterEmbeddedWalletSessionKeyResponse.RegisterEmbeddedWalletSessionKeyResponseFromJSON;
exports.RegisterEmbeddedWalletSessionKeyResponseFromJSONTyped = RegisterEmbeddedWalletSessionKeyResponse.RegisterEmbeddedWalletSessionKeyResponseFromJSONTyped;
exports.RegisterEmbeddedWalletSessionKeyResponseToJSON = RegisterEmbeddedWalletSessionKeyResponse.RegisterEmbeddedWalletSessionKeyResponseToJSON;
exports.RegisterSessionKeyRequestFromJSON = RegisterSessionKeyRequest.RegisterSessionKeyRequestFromJSON;
exports.RegisterSessionKeyRequestFromJSONTyped = RegisterSessionKeyRequest.RegisterSessionKeyRequestFromJSONTyped;
exports.RegisterSessionKeyRequestToJSON = RegisterSessionKeyRequest.RegisterSessionKeyRequestToJSON;
exports.ReshareRequestFromJSON = ReshareRequest.ReshareRequestFromJSON;
exports.ReshareRequestFromJSONTyped = ReshareRequest.ReshareRequestFromJSONTyped;
exports.ReshareRequestToJSON = ReshareRequest.ReshareRequestToJSON;
Object.defineProperty(exports, 'ResidentKeyRequirement', {
	enumerable: true,
	get: function () { return ResidentKeyRequirement.ResidentKeyRequirement; }
});
exports.ResidentKeyRequirementFromJSON = ResidentKeyRequirement.ResidentKeyRequirementFromJSON;
exports.ResidentKeyRequirementFromJSONTyped = ResidentKeyRequirement.ResidentKeyRequirementFromJSONTyped;
exports.ResidentKeyRequirementToJSON = ResidentKeyRequirement.ResidentKeyRequirementToJSON;
exports.ScanWebsiteUrlRequestFromJSON = ScanWebsiteUrlRequest.ScanWebsiteUrlRequestFromJSON;
exports.ScanWebsiteUrlRequestFromJSONTyped = ScanWebsiteUrlRequest.ScanWebsiteUrlRequestFromJSONTyped;
exports.ScanWebsiteUrlRequestToJSON = ScanWebsiteUrlRequest.ScanWebsiteUrlRequestToJSON;
exports.ScanWebsiteUrlResponseFromJSON = ScanWebsiteUrlResponse.ScanWebsiteUrlResponseFromJSON;
exports.ScanWebsiteUrlResponseFromJSONTyped = ScanWebsiteUrlResponse.ScanWebsiteUrlResponseFromJSONTyped;
exports.ScanWebsiteUrlResponseToJSON = ScanWebsiteUrlResponse.ScanWebsiteUrlResponseToJSON;
exports.SdkSettingsRequestFromJSON = SdkSettingsRequest.SdkSettingsRequestFromJSON;
exports.SdkSettingsRequestFromJSONTyped = SdkSettingsRequest.SdkSettingsRequestFromJSONTyped;
exports.SdkSettingsRequestToJSON = SdkSettingsRequest.SdkSettingsRequestToJSON;
exports.SdkUserFromJSON = SdkUser.SdkUserFromJSON;
exports.SdkUserFromJSONTyped = SdkUser.SdkUserFromJSONTyped;
exports.SdkUserToJSON = SdkUser.SdkUserToJSON;
exports.SdkViewFromJSON = SdkView.SdkViewFromJSON;
exports.SdkViewFromJSONTyped = SdkView.SdkViewFromJSONTyped;
exports.SdkViewToJSON = SdkView.SdkViewToJSON;
exports.SdkViewSectionFromJSON = SdkViewSection.SdkViewSectionFromJSON;
exports.SdkViewSectionFromJSONTyped = SdkViewSection.SdkViewSectionFromJSONTyped;
exports.SdkViewSectionToJSON = SdkViewSection.SdkViewSectionToJSON;
Object.defineProperty(exports, 'SdkViewSectionAlignment', {
	enumerable: true,
	get: function () { return SdkViewSectionAlignment.SdkViewSectionAlignment; }
});
exports.SdkViewSectionAlignmentFromJSON = SdkViewSectionAlignment.SdkViewSectionAlignmentFromJSON;
exports.SdkViewSectionAlignmentFromJSONTyped = SdkViewSectionAlignment.SdkViewSectionAlignmentFromJSONTyped;
exports.SdkViewSectionAlignmentToJSON = SdkViewSectionAlignment.SdkViewSectionAlignmentToJSON;
Object.defineProperty(exports, 'SdkViewSectionType', {
	enumerable: true,
	get: function () { return SdkViewSectionType.SdkViewSectionType; }
});
exports.SdkViewSectionTypeFromJSON = SdkViewSectionType.SdkViewSectionTypeFromJSON;
exports.SdkViewSectionTypeFromJSONTyped = SdkViewSectionType.SdkViewSectionTypeFromJSONTyped;
exports.SdkViewSectionTypeToJSON = SdkViewSectionType.SdkViewSectionTypeToJSON;
Object.defineProperty(exports, 'SdkViewType', {
	enumerable: true,
	get: function () { return SdkViewType.SdkViewType; }
});
exports.SdkViewTypeFromJSON = SdkViewType.SdkViewTypeFromJSON;
exports.SdkViewTypeFromJSONTyped = SdkViewType.SdkViewTypeFromJSONTyped;
exports.SdkViewTypeToJSON = SdkViewType.SdkViewTypeToJSON;
exports.SdkViewUpdateRequestFromJSON = SdkViewUpdateRequest.SdkViewUpdateRequestFromJSON;
exports.SdkViewUpdateRequestFromJSONTyped = SdkViewUpdateRequest.SdkViewUpdateRequestFromJSONTyped;
exports.SdkViewUpdateRequestToJSON = SdkViewUpdateRequest.SdkViewUpdateRequestToJSON;
exports.SdkViewsResponseFromJSON = SdkViewsResponse.SdkViewsResponseFromJSON;
exports.SdkViewsResponseFromJSONTyped = SdkViewsResponse.SdkViewsResponseFromJSONTyped;
exports.SdkViewsResponseToJSON = SdkViewsResponse.SdkViewsResponseToJSON;
Object.defineProperty(exports, 'SignInProviderEnum', {
	enumerable: true,
	get: function () { return SignInProviderEnum.SignInProviderEnum; }
});
exports.SignInProviderEnumFromJSON = SignInProviderEnum.SignInProviderEnumFromJSON;
exports.SignInProviderEnumFromJSONTyped = SignInProviderEnum.SignInProviderEnumFromJSONTyped;
exports.SignInProviderEnumToJSON = SignInProviderEnum.SignInProviderEnumToJSON;
exports.SignMessageAuthorizationSignatureFromJSON = SignMessageAuthorizationSignature.SignMessageAuthorizationSignatureFromJSON;
exports.SignMessageAuthorizationSignatureFromJSONTyped = SignMessageAuthorizationSignature.SignMessageAuthorizationSignatureFromJSONTyped;
exports.SignMessageAuthorizationSignatureToJSON = SignMessageAuthorizationSignature.SignMessageAuthorizationSignatureToJSON;
exports.SignMessageContextFromJSON = SignMessageContext.SignMessageContextFromJSON;
exports.SignMessageContextFromJSONTyped = SignMessageContext.SignMessageContextFromJSONTyped;
exports.SignMessageContextToJSON = SignMessageContext.SignMessageContextToJSON;
exports.SignMessageEip7702AuthFromJSON = SignMessageEip7702Auth.SignMessageEip7702AuthFromJSON;
exports.SignMessageEip7702AuthFromJSONTyped = SignMessageEip7702Auth.SignMessageEip7702AuthFromJSONTyped;
exports.SignMessageEip7702AuthToJSON = SignMessageEip7702Auth.SignMessageEip7702AuthToJSON;
exports.SignMessageEvmMessageFromJSON = SignMessageEvmMessage.SignMessageEvmMessageFromJSON;
exports.SignMessageEvmMessageFromJSONTyped = SignMessageEvmMessage.SignMessageEvmMessageFromJSONTyped;
exports.SignMessageEvmMessageToJSON = SignMessageEvmMessage.SignMessageEvmMessageToJSON;
exports.SignMessageEvmMessageAnyOfFromJSON = SignMessageEvmMessageAnyOf.SignMessageEvmMessageAnyOfFromJSON;
exports.SignMessageEvmMessageAnyOfFromJSONTyped = SignMessageEvmMessageAnyOf.SignMessageEvmMessageAnyOfFromJSONTyped;
exports.SignMessageEvmMessageAnyOfToJSON = SignMessageEvmMessageAnyOf.SignMessageEvmMessageAnyOfToJSON;
exports.SignMessageEvmTransactionFromJSON = SignMessageEvmTransaction.SignMessageEvmTransactionFromJSON;
exports.SignMessageEvmTransactionFromJSONTyped = SignMessageEvmTransaction.SignMessageEvmTransactionFromJSONTyped;
exports.SignMessageEvmTransactionToJSON = SignMessageEvmTransaction.SignMessageEvmTransactionToJSON;
exports.SignMessageEvmTypedDataFromJSON = SignMessageEvmTypedData.SignMessageEvmTypedDataFromJSON;
exports.SignMessageEvmTypedDataFromJSONTyped = SignMessageEvmTypedData.SignMessageEvmTypedDataFromJSONTyped;
exports.SignMessageEvmTypedDataToJSON = SignMessageEvmTypedData.SignMessageEvmTypedDataToJSON;
exports.SignMessageEvmUserOperationFromJSON = SignMessageEvmUserOperation.SignMessageEvmUserOperationFromJSON;
exports.SignMessageEvmUserOperationFromJSONTyped = SignMessageEvmUserOperation.SignMessageEvmUserOperationFromJSONTyped;
exports.SignMessageEvmUserOperationToJSON = SignMessageEvmUserOperation.SignMessageEvmUserOperationToJSON;
exports.SignMessageSuiTransactionFromJSON = SignMessageSuiTransaction.SignMessageSuiTransactionFromJSON;
exports.SignMessageSuiTransactionFromJSONTyped = SignMessageSuiTransaction.SignMessageSuiTransactionFromJSONTyped;
exports.SignMessageSuiTransactionToJSON = SignMessageSuiTransaction.SignMessageSuiTransactionToJSON;
exports.SignMessageSvmTransactionFromJSON = SignMessageSvmTransaction.SignMessageSvmTransactionFromJSON;
exports.SignMessageSvmTransactionFromJSONTyped = SignMessageSvmTransaction.SignMessageSvmTransactionFromJSONTyped;
exports.SignMessageSvmTransactionToJSON = SignMessageSvmTransaction.SignMessageSvmTransactionToJSON;
exports.SignMessageUserOperationDataFromJSON = SignMessageUserOperationData.SignMessageUserOperationDataFromJSON;
exports.SignMessageUserOperationDataFromJSONTyped = SignMessageUserOperationData.SignMessageUserOperationDataFromJSONTyped;
exports.SignMessageUserOperationDataToJSON = SignMessageUserOperationData.SignMessageUserOperationDataToJSON;
exports.SignMessageWithWaasRequestFromJSON = SignMessageWithWaasRequest.SignMessageWithWaasRequestFromJSON;
exports.SignMessageWithWaasRequestFromJSONTyped = SignMessageWithWaasRequest.SignMessageWithWaasRequestFromJSONTyped;
exports.SignMessageWithWaasRequestToJSON = SignMessageWithWaasRequest.SignMessageWithWaasRequestToJSON;
exports.SignTransactionWithWaasRequestFromJSON = SignTransactionWithWaasRequest.SignTransactionWithWaasRequestFromJSON;
exports.SignTransactionWithWaasRequestFromJSONTyped = SignTransactionWithWaasRequest.SignTransactionWithWaasRequestFromJSONTyped;
exports.SignTransactionWithWaasRequestToJSON = SignTransactionWithWaasRequest.SignTransactionWithWaasRequestToJSON;
exports.SimulateEVMTransactionRequestFromJSON = SimulateEVMTransactionRequest.SimulateEVMTransactionRequestFromJSON;
exports.SimulateEVMTransactionRequestFromJSONTyped = SimulateEVMTransactionRequest.SimulateEVMTransactionRequestFromJSONTyped;
exports.SimulateEVMTransactionRequestToJSON = SimulateEVMTransactionRequest.SimulateEVMTransactionRequestToJSON;
exports.SimulateSVMTransactionRequestFromJSON = SimulateSVMTransactionRequest.SimulateSVMTransactionRequestFromJSON;
exports.SimulateSVMTransactionRequestFromJSONTyped = SimulateSVMTransactionRequest.SimulateSVMTransactionRequestFromJSONTyped;
exports.SimulateSVMTransactionRequestToJSON = SimulateSVMTransactionRequest.SimulateSVMTransactionRequestToJSON;
exports.SimulateTransactionResponseFromJSON = SimulateTransactionResponse.SimulateTransactionResponseFromJSON;
exports.SimulateTransactionResponseFromJSONTyped = SimulateTransactionResponse.SimulateTransactionResponseFromJSONTyped;
exports.SimulateTransactionResponseToJSON = SimulateTransactionResponse.SimulateTransactionResponseToJSON;
exports.SimulateUserOpRequestFromJSON = SimulateUserOpRequest.SimulateUserOpRequestFromJSON;
exports.SimulateUserOpRequestFromJSONTyped = SimulateUserOpRequest.SimulateUserOpRequestFromJSONTyped;
exports.SimulateUserOpRequestToJSON = SimulateUserOpRequest.SimulateUserOpRequestToJSON;
exports.SmartWalletPropertiesFromJSON = SmartWalletProperties.SmartWalletPropertiesFromJSON;
exports.SmartWalletPropertiesFromJSONTyped = SmartWalletProperties.SmartWalletPropertiesFromJSONTyped;
exports.SmartWalletPropertiesToJSON = SmartWalletProperties.SmartWalletPropertiesToJSON;
exports.SmsCountryCodeFromJSON = SmsCountryCode.SmsCountryCodeFromJSON;
exports.SmsCountryCodeFromJSONTyped = SmsCountryCode.SmsCountryCodeFromJSONTyped;
exports.SmsCountryCodeToJSON = SmsCountryCode.SmsCountryCodeToJSON;
exports.SmsVerificationCreateRequestFromJSON = SmsVerificationCreateRequest.SmsVerificationCreateRequestFromJSON;
exports.SmsVerificationCreateRequestFromJSONTyped = SmsVerificationCreateRequest.SmsVerificationCreateRequestFromJSONTyped;
exports.SmsVerificationCreateRequestToJSON = SmsVerificationCreateRequest.SmsVerificationCreateRequestToJSON;
exports.SmsVerificationCreateResponseFromJSON = SmsVerificationCreateResponse.SmsVerificationCreateResponseFromJSON;
exports.SmsVerificationCreateResponseFromJSONTyped = SmsVerificationCreateResponse.SmsVerificationCreateResponseFromJSONTyped;
exports.SmsVerificationCreateResponseToJSON = SmsVerificationCreateResponse.SmsVerificationCreateResponseToJSON;
exports.SmsVerificationRetryRequestFromJSON = SmsVerificationRetryRequest.SmsVerificationRetryRequestFromJSON;
exports.SmsVerificationRetryRequestFromJSONTyped = SmsVerificationRetryRequest.SmsVerificationRetryRequestFromJSONTyped;
exports.SmsVerificationRetryRequestToJSON = SmsVerificationRetryRequest.SmsVerificationRetryRequestToJSON;
exports.SmsVerificationVerifyRequestFromJSON = SmsVerificationVerifyRequest.SmsVerificationVerifyRequestFromJSON;
exports.SmsVerificationVerifyRequestFromJSONTyped = SmsVerificationVerifyRequest.SmsVerificationVerifyRequestFromJSONTyped;
exports.SmsVerificationVerifyRequestToJSON = SmsVerificationVerifyRequest.SmsVerificationVerifyRequestToJSON;
exports.SocialSignInProviderFromJSON = SocialSignInProvider.SocialSignInProviderFromJSON;
exports.SocialSignInProviderFromJSONTyped = SocialSignInProvider.SocialSignInProviderFromJSONTyped;
exports.SocialSignInProviderToJSON = SocialSignInProvider.SocialSignInProviderToJSON;
Object.defineProperty(exports, 'SocialSignInProviderEnum', {
	enumerable: true,
	get: function () { return SocialSignInProviderEnum.SocialSignInProviderEnum; }
});
exports.SocialSignInProviderEnumFromJSON = SocialSignInProviderEnum.SocialSignInProviderEnumFromJSON;
exports.SocialSignInProviderEnumFromJSONTyped = SocialSignInProviderEnum.SocialSignInProviderEnumFromJSONTyped;
exports.SocialSignInProviderEnumToJSON = SocialSignInProviderEnum.SocialSignInProviderEnumToJSON;
exports.SolanaTransactionOptimizationRequestFromJSON = SolanaTransactionOptimizationRequest.SolanaTransactionOptimizationRequestFromJSON;
exports.SolanaTransactionOptimizationRequestFromJSONTyped = SolanaTransactionOptimizationRequest.SolanaTransactionOptimizationRequestFromJSONTyped;
exports.SolanaTransactionOptimizationRequestToJSON = SolanaTransactionOptimizationRequest.SolanaTransactionOptimizationRequestToJSON;
exports.SolanaTransactionOptimizationResponseFromJSON = SolanaTransactionOptimizationResponse.SolanaTransactionOptimizationResponseFromJSON;
exports.SolanaTransactionOptimizationResponseFromJSONTyped = SolanaTransactionOptimizationResponse.SolanaTransactionOptimizationResponseFromJSONTyped;
exports.SolanaTransactionOptimizationResponseToJSON = SolanaTransactionOptimizationResponse.SolanaTransactionOptimizationResponseToJSON;
exports.SupportedOfframpsResponseFromJSON = SupportedOfframpsResponse.SupportedOfframpsResponseFromJSON;
exports.SupportedOfframpsResponseFromJSONTyped = SupportedOfframpsResponse.SupportedOfframpsResponseFromJSONTyped;
exports.SupportedOfframpsResponseToJSON = SupportedOfframpsResponse.SupportedOfframpsResponseToJSON;
exports.SupportedOnrampsResponseFromJSON = SupportedOnrampsResponse.SupportedOnrampsResponseFromJSON;
exports.SupportedOnrampsResponseFromJSONTyped = SupportedOnrampsResponse.SupportedOnrampsResponseFromJSONTyped;
exports.SupportedOnrampsResponseToJSON = SupportedOnrampsResponse.SupportedOnrampsResponseToJSON;
exports.SupportedSecurityMethodFromJSON = SupportedSecurityMethod.SupportedSecurityMethodFromJSON;
exports.SupportedSecurityMethodFromJSONTyped = SupportedSecurityMethod.SupportedSecurityMethodFromJSONTyped;
exports.SupportedSecurityMethodToJSON = SupportedSecurityMethod.SupportedSecurityMethodToJSON;
exports.SupportedSecurityMethodsFromJSON = SupportedSecurityMethods.SupportedSecurityMethodsFromJSON;
exports.SupportedSecurityMethodsFromJSONTyped = SupportedSecurityMethods.SupportedSecurityMethodsFromJSONTyped;
exports.SupportedSecurityMethodsToJSON = SupportedSecurityMethods.SupportedSecurityMethodsToJSON;
exports.TelegramPostRequestFromJSON = TelegramPostRequest.TelegramPostRequestFromJSON;
exports.TelegramPostRequestFromJSONTyped = TelegramPostRequest.TelegramPostRequestFromJSONTyped;
exports.TelegramPostRequestToJSON = TelegramPostRequest.TelegramPostRequestToJSON;
exports.TelegramUserFromJSON = TelegramUser.TelegramUserFromJSON;
exports.TelegramUserFromJSONTyped = TelegramUser.TelegramUserFromJSONTyped;
exports.TelegramUserToJSON = TelegramUser.TelegramUserToJSON;
Object.defineProperty(exports, 'ThresholdSignatureScheme', {
	enumerable: true,
	get: function () { return ThresholdSignatureScheme.ThresholdSignatureScheme; }
});
exports.ThresholdSignatureSchemeFromJSON = ThresholdSignatureScheme.ThresholdSignatureSchemeFromJSON;
exports.ThresholdSignatureSchemeFromJSONTyped = ThresholdSignatureScheme.ThresholdSignatureSchemeFromJSONTyped;
exports.ThresholdSignatureSchemeToJSON = ThresholdSignatureScheme.ThresholdSignatureSchemeToJSON;
Object.defineProperty(exports, 'TimeUnitEnum', {
	enumerable: true,
	get: function () { return TimeUnitEnum.TimeUnitEnum; }
});
exports.TimeUnitEnumFromJSON = TimeUnitEnum.TimeUnitEnumFromJSON;
exports.TimeUnitEnumFromJSONTyped = TimeUnitEnum.TimeUnitEnumFromJSONTyped;
exports.TimeUnitEnumToJSON = TimeUnitEnum.TimeUnitEnumToJSON;
exports.TokenBalanceFromJSON = TokenBalance.TokenBalanceFromJSON;
exports.TokenBalanceFromJSONTyped = TokenBalance.TokenBalanceFromJSONTyped;
exports.TokenBalanceToJSON = TokenBalance.TokenBalanceToJSON;
exports.TooManyRequestsFromJSON = TooManyRequests.TooManyRequestsFromJSON;
exports.TooManyRequestsFromJSONTyped = TooManyRequests.TooManyRequestsFromJSONTyped;
exports.TooManyRequestsToJSON = TooManyRequests.TooManyRequestsToJSON;
exports.TransferDestinationFromJSON = TransferDestination.TransferDestinationFromJSON;
exports.TransferDestinationFromJSONTyped = TransferDestination.TransferDestinationFromJSONTyped;
exports.TransferDestinationToJSON = TransferDestination.TransferDestinationToJSON;
exports.TransferDestinationResponseFromJSON = TransferDestinationResponse.TransferDestinationResponseFromJSON;
exports.TransferDestinationResponseFromJSONTyped = TransferDestinationResponse.TransferDestinationResponseFromJSONTyped;
exports.TransferDestinationResponseToJSON = TransferDestinationResponse.TransferDestinationResponseToJSON;
exports.TurnkeyCreateWalletAccountsRequestBodyFromJSON = TurnkeyCreateWalletAccountsRequestBody.TurnkeyCreateWalletAccountsRequestBodyFromJSON;
exports.TurnkeyCreateWalletAccountsRequestBodyFromJSONTyped = TurnkeyCreateWalletAccountsRequestBody.TurnkeyCreateWalletAccountsRequestBodyFromJSONTyped;
exports.TurnkeyCreateWalletAccountsRequestBodyToJSON = TurnkeyCreateWalletAccountsRequestBody.TurnkeyCreateWalletAccountsRequestBodyToJSON;
exports.TurnkeyCreateWalletAccountsRequestBodyParametersFromJSON = TurnkeyCreateWalletAccountsRequestBodyParameters.TurnkeyCreateWalletAccountsRequestBodyParametersFromJSON;
exports.TurnkeyCreateWalletAccountsRequestBodyParametersFromJSONTyped = TurnkeyCreateWalletAccountsRequestBodyParameters.TurnkeyCreateWalletAccountsRequestBodyParametersFromJSONTyped;
exports.TurnkeyCreateWalletAccountsRequestBodyParametersToJSON = TurnkeyCreateWalletAccountsRequestBodyParameters.TurnkeyCreateWalletAccountsRequestBodyParametersToJSON;
exports.TurnkeyDeleteEmbeddedWalletsRequestBodyFromJSON = TurnkeyDeleteEmbeddedWalletsRequestBody.TurnkeyDeleteEmbeddedWalletsRequestBodyFromJSON;
exports.TurnkeyDeleteEmbeddedWalletsRequestBodyFromJSONTyped = TurnkeyDeleteEmbeddedWalletsRequestBody.TurnkeyDeleteEmbeddedWalletsRequestBodyFromJSONTyped;
exports.TurnkeyDeleteEmbeddedWalletsRequestBodyToJSON = TurnkeyDeleteEmbeddedWalletsRequestBody.TurnkeyDeleteEmbeddedWalletsRequestBodyToJSON;
exports.TurnkeyDeleteEmbeddedWalletsRequestBodyParametersFromJSON = TurnkeyDeleteEmbeddedWalletsRequestBodyParameters.TurnkeyDeleteEmbeddedWalletsRequestBodyParametersFromJSON;
exports.TurnkeyDeleteEmbeddedWalletsRequestBodyParametersFromJSONTyped = TurnkeyDeleteEmbeddedWalletsRequestBodyParameters.TurnkeyDeleteEmbeddedWalletsRequestBodyParametersFromJSONTyped;
exports.TurnkeyDeleteEmbeddedWalletsRequestBodyParametersToJSON = TurnkeyDeleteEmbeddedWalletsRequestBodyParameters.TurnkeyDeleteEmbeddedWalletsRequestBodyParametersToJSON;
exports.TurnkeySignedRequestFromJSON = TurnkeySignedRequest.TurnkeySignedRequestFromJSON;
exports.TurnkeySignedRequestFromJSONTyped = TurnkeySignedRequest.TurnkeySignedRequestFromJSONTyped;
exports.TurnkeySignedRequestToJSON = TurnkeySignedRequest.TurnkeySignedRequestToJSON;
exports.TurnkeyStampFromJSON = TurnkeyStamp.TurnkeyStampFromJSON;
exports.TurnkeyStampFromJSONTyped = TurnkeyStamp.TurnkeyStampFromJSONTyped;
exports.TurnkeyStampToJSON = TurnkeyStamp.TurnkeyStampToJSON;
exports.TurnkeyWalletAccountFromJSON = TurnkeyWalletAccount.TurnkeyWalletAccountFromJSON;
exports.TurnkeyWalletAccountFromJSONTyped = TurnkeyWalletAccount.TurnkeyWalletAccountFromJSONTyped;
exports.TurnkeyWalletAccountToJSON = TurnkeyWalletAccount.TurnkeyWalletAccountToJSON;
exports.TurnkeyWalletPropertiesFromJSON = TurnkeyWalletProperties.TurnkeyWalletPropertiesFromJSON;
exports.TurnkeyWalletPropertiesFromJSONTyped = TurnkeyWalletProperties.TurnkeyWalletPropertiesFromJSONTyped;
exports.TurnkeyWalletPropertiesToJSON = TurnkeyWalletProperties.TurnkeyWalletPropertiesToJSON;
exports.UnauthorizedFromJSON = Unauthorized.UnauthorizedFromJSON;
exports.UnauthorizedFromJSONTyped = Unauthorized.UnauthorizedFromJSONTyped;
exports.UnauthorizedToJSON = Unauthorized.UnauthorizedToJSON;
exports.UnprocessableEntityFromJSON = UnprocessableEntity.UnprocessableEntityFromJSON;
exports.UnprocessableEntityFromJSONTyped = UnprocessableEntity.UnprocessableEntityFromJSONTyped;
exports.UnprocessableEntityToJSON = UnprocessableEntity.UnprocessableEntityToJSON;
Object.defineProperty(exports, 'UnprocessableEntityErrorCode', {
	enumerable: true,
	get: function () { return UnprocessableEntityErrorCode.UnprocessableEntityErrorCode; }
});
exports.UnprocessableEntityErrorCodeFromJSON = UnprocessableEntityErrorCode.UnprocessableEntityErrorCodeFromJSON;
exports.UnprocessableEntityErrorCodeFromJSONTyped = UnprocessableEntityErrorCode.UnprocessableEntityErrorCodeFromJSONTyped;
exports.UnprocessableEntityErrorCodeToJSON = UnprocessableEntityErrorCode.UnprocessableEntityErrorCodeToJSON;
exports.UnprocessableEntityErrorPayloadFromJSON = UnprocessableEntityErrorPayload.UnprocessableEntityErrorPayloadFromJSON;
exports.UnprocessableEntityErrorPayloadFromJSONTyped = UnprocessableEntityErrorPayload.UnprocessableEntityErrorPayloadFromJSONTyped;
exports.UnprocessableEntityErrorPayloadToJSON = UnprocessableEntityErrorPayload.UnprocessableEntityErrorPayloadToJSON;
exports.UpdateRecoveryEmailRequestFromJSON = UpdateRecoveryEmailRequest.UpdateRecoveryEmailRequestFromJSON;
exports.UpdateRecoveryEmailRequestFromJSONTyped = UpdateRecoveryEmailRequest.UpdateRecoveryEmailRequestFromJSONTyped;
exports.UpdateRecoveryEmailRequestToJSON = UpdateRecoveryEmailRequest.UpdateRecoveryEmailRequestToJSON;
exports.UpdateSelfResponseFromJSON = UpdateSelfResponse.UpdateSelfResponseFromJSON;
exports.UpdateSelfResponseFromJSONTyped = UpdateSelfResponse.UpdateSelfResponseFromJSONTyped;
exports.UpdateSelfResponseToJSON = UpdateSelfResponse.UpdateSelfResponseToJSON;
exports.UpdateSelfResponseAllOfFromJSON = UpdateSelfResponseAllOf.UpdateSelfResponseAllOfFromJSON;
exports.UpdateSelfResponseAllOfFromJSONTyped = UpdateSelfResponseAllOf.UpdateSelfResponseAllOfFromJSONTyped;
exports.UpdateSelfResponseAllOfToJSON = UpdateSelfResponseAllOf.UpdateSelfResponseAllOfToJSON;
exports.UpdateUserPasskeyRequestFromJSON = UpdateUserPasskeyRequest.UpdateUserPasskeyRequestFromJSON;
exports.UpdateUserPasskeyRequestFromJSONTyped = UpdateUserPasskeyRequest.UpdateUserPasskeyRequestFromJSONTyped;
exports.UpdateUserPasskeyRequestToJSON = UpdateUserPasskeyRequest.UpdateUserPasskeyRequestToJSON;
exports.UpgradeEmbeddedWalletToV2RequestFromJSON = UpgradeEmbeddedWalletToV2Request.UpgradeEmbeddedWalletToV2RequestFromJSON;
exports.UpgradeEmbeddedWalletToV2RequestFromJSONTyped = UpgradeEmbeddedWalletToV2Request.UpgradeEmbeddedWalletToV2RequestFromJSONTyped;
exports.UpgradeEmbeddedWalletToV2RequestToJSON = UpgradeEmbeddedWalletToV2Request.UpgradeEmbeddedWalletToV2RequestToJSON;
exports.UserFieldsFromJSON = UserFields.UserFieldsFromJSON;
exports.UserFieldsFromJSONTyped = UserFields.UserFieldsFromJSONTyped;
exports.UserFieldsToJSON = UserFields.UserFieldsToJSON;
Object.defineProperty(exports, 'UserFieldsCheckEnum', {
	enumerable: true,
	get: function () { return UserFieldsCheckEnum.UserFieldsCheckEnum; }
});
exports.UserFieldsCheckEnumFromJSON = UserFieldsCheckEnum.UserFieldsCheckEnumFromJSON;
exports.UserFieldsCheckEnumFromJSONTyped = UserFieldsCheckEnum.UserFieldsCheckEnumFromJSONTyped;
exports.UserFieldsCheckEnumToJSON = UserFieldsCheckEnum.UserFieldsCheckEnumToJSON;
exports.UserFieldsCheckParamsFromJSON = UserFieldsCheckParams.UserFieldsCheckParamsFromJSON;
exports.UserFieldsCheckParamsFromJSONTyped = UserFieldsCheckParams.UserFieldsCheckParamsFromJSONTyped;
exports.UserFieldsCheckParamsToJSON = UserFieldsCheckParams.UserFieldsCheckParamsToJSON;
exports.UserFieldsCheckResponseFromJSON = UserFieldsCheckResponse.UserFieldsCheckResponseFromJSON;
exports.UserFieldsCheckResponseFromJSONTyped = UserFieldsCheckResponse.UserFieldsCheckResponseFromJSONTyped;
exports.UserFieldsCheckResponseToJSON = UserFieldsCheckResponse.UserFieldsCheckResponseToJSON;
Object.defineProperty(exports, 'UserIdentifierTypeEnum', {
	enumerable: true,
	get: function () { return UserIdentifierTypeEnum.UserIdentifierTypeEnum; }
});
exports.UserIdentifierTypeEnumFromJSON = UserIdentifierTypeEnum.UserIdentifierTypeEnumFromJSON;
exports.UserIdentifierTypeEnumFromJSONTyped = UserIdentifierTypeEnum.UserIdentifierTypeEnumFromJSONTyped;
exports.UserIdentifierTypeEnumToJSON = UserIdentifierTypeEnum.UserIdentifierTypeEnumToJSON;
exports.UserOauthAccessTokenResponseFromJSON = UserOauthAccessTokenResponse.UserOauthAccessTokenResponseFromJSON;
exports.UserOauthAccessTokenResponseFromJSONTyped = UserOauthAccessTokenResponse.UserOauthAccessTokenResponseFromJSONTyped;
exports.UserOauthAccessTokenResponseToJSON = UserOauthAccessTokenResponse.UserOauthAccessTokenResponseToJSON;
exports.UserPasskeyFromJSON = UserPasskey.UserPasskeyFromJSON;
exports.UserPasskeyFromJSONTyped = UserPasskey.UserPasskeyFromJSONTyped;
exports.UserPasskeyToJSON = UserPasskey.UserPasskeyToJSON;
Object.defineProperty(exports, 'UserVerificationRequirement', {
	enumerable: true,
	get: function () { return UserVerificationRequirement.UserVerificationRequirement; }
});
exports.UserVerificationRequirementFromJSON = UserVerificationRequirement.UserVerificationRequirementFromJSON;
exports.UserVerificationRequirementFromJSONTyped = UserVerificationRequirement.UserVerificationRequirementFromJSONTyped;
exports.UserVerificationRequirementToJSON = UserVerificationRequirement.UserVerificationRequirementToJSON;
exports.UserWalletSelectionRequestFromJSON = UserWalletSelectionRequest.UserWalletSelectionRequestFromJSON;
exports.UserWalletSelectionRequestFromJSONTyped = UserWalletSelectionRequest.UserWalletSelectionRequestFromJSONTyped;
exports.UserWalletSelectionRequestToJSON = UserWalletSelectionRequest.UserWalletSelectionRequestToJSON;
exports.V6UserOpFromJSON = V6UserOp.V6UserOpFromJSON;
exports.V6UserOpFromJSONTyped = V6UserOp.V6UserOpFromJSONTyped;
exports.V6UserOpToJSON = V6UserOp.V6UserOpToJSON;
exports.V7UserOpFromJSON = V7UserOp.V7UserOpFromJSON;
exports.V7UserOpFromJSONTyped = V7UserOp.V7UserOpFromJSONTyped;
exports.V7UserOpToJSON = V7UserOp.V7UserOpToJSON;
exports.VerifyRequestFromJSON = VerifyRequest.VerifyRequestFromJSON;
exports.VerifyRequestFromJSONTyped = VerifyRequest.VerifyRequestFromJSONTyped;
exports.VerifyRequestToJSON = VerifyRequest.VerifyRequestToJSON;
exports.VerifyResponseFromJSON = VerifyResponse.VerifyResponseFromJSON;
exports.VerifyResponseFromJSONTyped = VerifyResponse.VerifyResponseFromJSONTyped;
exports.VerifyResponseToJSON = VerifyResponse.VerifyResponseToJSON;
exports.VerifyUnlinkRequestFromJSON = VerifyUnlinkRequest.VerifyUnlinkRequestFromJSON;
exports.VerifyUnlinkRequestFromJSONTyped = VerifyUnlinkRequest.VerifyUnlinkRequestFromJSONTyped;
exports.VerifyUnlinkRequestToJSON = VerifyUnlinkRequest.VerifyUnlinkRequestToJSON;
Object.defineProperty(exports, 'WaasBackupOptionsEnum', {
	enumerable: true,
	get: function () { return WaasBackupOptionsEnum.WaasBackupOptionsEnum; }
});
exports.WaasBackupOptionsEnumFromJSON = WaasBackupOptionsEnum.WaasBackupOptionsEnumFromJSON;
exports.WaasBackupOptionsEnumFromJSONTyped = WaasBackupOptionsEnum.WaasBackupOptionsEnumFromJSONTyped;
exports.WaasBackupOptionsEnumToJSON = WaasBackupOptionsEnum.WaasBackupOptionsEnumToJSON;
Object.defineProperty(exports, 'WaasChainEnum', {
	enumerable: true,
	get: function () { return WaasChainEnum.WaasChainEnum; }
});
exports.WaasChainEnumFromJSON = WaasChainEnum.WaasChainEnumFromJSON;
exports.WaasChainEnumFromJSONTyped = WaasChainEnum.WaasChainEnumFromJSONTyped;
exports.WaasChainEnumToJSON = WaasChainEnum.WaasChainEnumToJSON;
exports.WaasWalletPropertiesFromJSON = WaasWalletProperties.WaasWalletPropertiesFromJSON;
exports.WaasWalletPropertiesFromJSONTyped = WaasWalletProperties.WaasWalletPropertiesFromJSONTyped;
exports.WaasWalletPropertiesToJSON = WaasWalletProperties.WaasWalletPropertiesToJSON;
exports.WalletAdditionalAddressFromJSON = WalletAdditionalAddress.WalletAdditionalAddressFromJSON;
exports.WalletAdditionalAddressFromJSONTyped = WalletAdditionalAddress.WalletAdditionalAddressFromJSONTyped;
exports.WalletAdditionalAddressToJSON = WalletAdditionalAddress.WalletAdditionalAddressToJSON;
Object.defineProperty(exports, 'WalletAddressType', {
	enumerable: true,
	get: function () { return WalletAddressType.WalletAddressType; }
});
exports.WalletAddressTypeFromJSON = WalletAddressType.WalletAddressTypeFromJSON;
exports.WalletAddressTypeFromJSONTyped = WalletAddressType.WalletAddressTypeFromJSONTyped;
exports.WalletAddressTypeToJSON = WalletAddressType.WalletAddressTypeToJSON;
exports.WalletKeyShareInfoFromJSON = WalletKeyShareInfo.WalletKeyShareInfoFromJSON;
exports.WalletKeyShareInfoFromJSONTyped = WalletKeyShareInfo.WalletKeyShareInfoFromJSONTyped;
exports.WalletKeyShareInfoToJSON = WalletKeyShareInfo.WalletKeyShareInfoToJSON;
exports.WalletKeyShareInfoWithEncryptedAccountCredentialFromJSON = WalletKeyShareInfoWithEncryptedAccountCredential.WalletKeyShareInfoWithEncryptedAccountCredentialFromJSON;
exports.WalletKeyShareInfoWithEncryptedAccountCredentialFromJSONTyped = WalletKeyShareInfoWithEncryptedAccountCredential.WalletKeyShareInfoWithEncryptedAccountCredentialFromJSONTyped;
exports.WalletKeyShareInfoWithEncryptedAccountCredentialToJSON = WalletKeyShareInfoWithEncryptedAccountCredential.WalletKeyShareInfoWithEncryptedAccountCredentialToJSON;
exports.WalletKeyShareInfoWithEncryptedAccountCredentialAllOfFromJSON = WalletKeyShareInfoWithEncryptedAccountCredentialAllOf.WalletKeyShareInfoWithEncryptedAccountCredentialAllOfFromJSON;
exports.WalletKeyShareInfoWithEncryptedAccountCredentialAllOfFromJSONTyped = WalletKeyShareInfoWithEncryptedAccountCredentialAllOf.WalletKeyShareInfoWithEncryptedAccountCredentialAllOfFromJSONTyped;
exports.WalletKeyShareInfoWithEncryptedAccountCredentialAllOfToJSON = WalletKeyShareInfoWithEncryptedAccountCredentialAllOf.WalletKeyShareInfoWithEncryptedAccountCredentialAllOfToJSON;
exports.WalletPropertiesFromJSON = WalletProperties.WalletPropertiesFromJSON;
exports.WalletPropertiesFromJSONTyped = WalletProperties.WalletPropertiesFromJSONTyped;
exports.WalletPropertiesToJSON = WalletProperties.WalletPropertiesToJSON;
Object.defineProperty(exports, 'WalletProviderEnum', {
	enumerable: true,
	get: function () { return WalletProviderEnum.WalletProviderEnum; }
});
exports.WalletProviderEnumFromJSON = WalletProviderEnum.WalletProviderEnumFromJSON;
exports.WalletProviderEnumFromJSONTyped = WalletProviderEnum.WalletProviderEnumFromJSONTyped;
exports.WalletProviderEnumToJSON = WalletProviderEnum.WalletProviderEnumToJSON;
Object.defineProperty(exports, 'ZerodevBundlerProvider', {
	enumerable: true,
	get: function () { return ZerodevBundlerProvider.ZerodevBundlerProvider; }
});
exports.ZerodevBundlerProviderFromJSON = ZerodevBundlerProvider.ZerodevBundlerProviderFromJSON;
exports.ZerodevBundlerProviderFromJSONTyped = ZerodevBundlerProvider.ZerodevBundlerProviderFromJSONTyped;
exports.ZerodevBundlerProviderToJSON = ZerodevBundlerProvider.ZerodevBundlerProviderToJSON;
