// @dynamic-labs/sdk-react-core/src/lib/utils/hooks/useSocialAuth/farcaster/farcaster.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../../../../../../_virtual/_tslib.cjs');
var utils = require('@dynamic-labs/utils');
var dynamicEvents = require('../../../../events/dynamicEvents.cjs');
require('@dynamic-labs/iconic');
require('@dynamic-labs/wallet-connector-core');
require('react');
require('react/jsx-runtime');
require('../../../../context/ViewContext/ViewContext.cjs');
var logger = require('../../../../shared/logger.cjs');
require('@dynamic-labs/wallet-book');
require('../../../constants/colors.cjs');
require('../../../constants/values.cjs');
require('@dynamic-labs/sdk-api-core');
require('../../../../shared/consts/index.cjs');

const CONNECT_URL = 'https://relay.farcaster.xyz/v1/channel';
const STATUS_URL = 'https://relay.farcaster.xyz/v1/channel/status';
const STATUS_TIMEOUT = 180000;
const STATUS_INTERVAL = 1000;
const connect = (connectRequest) => _tslib.__awaiter(void 0, void 0, void 0, function* () {
    try {
        const response = yield utils.FetchService.fetch(CONNECT_URL, {
            body: JSON.stringify(connectRequest),
            headers: {
                'Content-Type': 'application/json',
            },
            method: 'POST',
        });
        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }
        return yield response.json();
    }
    catch (error) {
        logger.logger.error('Error connecting with farcaster:', error);
        return;
    }
});
const status = (connectStatusRequest_1, ...args_1) => _tslib.__awaiter(void 0, [connectStatusRequest_1, ...args_1], void 0, function* (connectStatusRequest, count = 1) {
    try {
        const response = yield utils.FetchService.fetch(STATUS_URL, {
            headers: {
                Authorization: `Bearer ${connectStatusRequest.channelToken}`,
                'Content-Type': 'application/json',
            },
            method: 'GET',
        });
        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }
        return yield response.json();
    }
    catch (error) {
        logger.logger.warn(`Error fetching farcaster connect status, trying again. Attempt: ${count}`, error);
        return { error: JSON.stringify(error), state: 'retry' };
    }
});
const watchStatus = (connectStatusRequest_2, ...args_2) => _tslib.__awaiter(void 0, [connectStatusRequest_2, ...args_2], void 0, function* (connectStatusRequest, { timeout = STATUS_TIMEOUT, interval = STATUS_INTERVAL } = {}) {
    const deadline = Date.now() + timeout;
    let shouldPoll = true;
    let storedResolve = undefined;
    // When we cancel, immediately resolve the current timeout promise and abort status watch
    dynamicEvents.dynamicEvents.on('farcasterConnectCancelled', () => {
        shouldPoll = false;
        storedResolve === null || storedResolve === void 0 ? void 0 : storedResolve();
        storedResolve = undefined;
    });
    // Declaring the promise callback outside of loop to avoid unnecessary garbage generation
    const timeoutCallback = (resolve) => {
        storedResolve = resolve;
        setTimeout(resolve, interval);
    };
    let count = 0;
    let timeoutError = 'Unknown error';
    while (shouldPoll && Date.now() < deadline) {
        count++;
        const response = yield status(connectStatusRequest, count);
        if (response.state === 'completed') {
            return { data: response, type: 'success' };
        }
        else if (response.state === 'retry') {
            timeoutError = response.error;
        }
        // response.state === 'pending'
        else
            timeoutError = 'Timed out: Farcaster connection status still pending';
        yield new Promise(timeoutCallback);
    }
    if (!shouldPoll) {
        return { type: 'cancelled' };
    }
    logger.logger.error(`Polling farcaster connect status timed out after ${timeout}ms`);
    return { error: timeoutError, type: 'timeout' };
});

exports.connect = connect;
exports.status = status;
exports.watchStatus = watchStatus;
