// @dynamic-labs/sdk-api-core/src/models/AuthStorageEnum.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Location in the browser where Dynamic will attempt to store the JWT token.
 * @export
 * @enum {string}
 */
exports.AuthStorageEnum = void 0;
(function (AuthStorageEnum) {
    AuthStorageEnum["Localstorage"] = "localstorage";
    AuthStorageEnum["Cookie"] = "cookie";
})(exports.AuthStorageEnum || (exports.AuthStorageEnum = {}));
function AuthStorageEnumFromJSON(json) {
    return AuthStorageEnumFromJSONTyped(json);
}
function AuthStorageEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function AuthStorageEnumToJSON(value) {
    return value;
}

exports.AuthStorageEnumFromJSON = AuthStorageEnumFromJSON;
exports.AuthStorageEnumFromJSONTyped = AuthStorageEnumFromJSONTyped;
exports.AuthStorageEnumToJSON = AuthStorageEnumToJSON;
