// @dynamic-labs/sdk-api-core/src/models/ExternalWalletFundingTokenRule.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
var ExternalWalletFundingTokenRule;
(function (ExternalWalletFundingTokenRule) {
    ExternalWalletFundingTokenRule["Recommended"] = "recommended";
    ExternalWalletFundingTokenRule["Exact"] = "exact";
})(ExternalWalletFundingTokenRule || (ExternalWalletFundingTokenRule = {}));
function ExternalWalletFundingTokenRuleFromJSON(json) {
    return ExternalWalletFundingTokenRuleFromJSONTyped(json);
}
function ExternalWalletFundingTokenRuleFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function ExternalWalletFundingTokenRuleToJSON(value) {
    return value;
}

export { ExternalWalletFundingTokenRule, ExternalWalletFundingTokenRuleFromJSON, ExternalWalletFundingTokenRuleFromJSONTyped, ExternalWalletFundingTokenRuleToJSON };
