// @dynamic-labs/sdk-api-core/src/models/GlobalWalletConnectionStatusEnum.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.GlobalWalletConnectionStatusEnum = void 0;
(function (GlobalWalletConnectionStatusEnum) {
    GlobalWalletConnectionStatusEnum["Connected"] = "connected";
    GlobalWalletConnectionStatusEnum["Disconnected"] = "disconnected";
})(exports.GlobalWalletConnectionStatusEnum || (exports.GlobalWalletConnectionStatusEnum = {}));
function GlobalWalletConnectionStatusEnumFromJSON(json) {
    return GlobalWalletConnectionStatusEnumFromJSONTyped(json);
}
function GlobalWalletConnectionStatusEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function GlobalWalletConnectionStatusEnumToJSON(value) {
    return value;
}

exports.GlobalWalletConnectionStatusEnumFromJSON = GlobalWalletConnectionStatusEnumFromJSON;
exports.GlobalWalletConnectionStatusEnumFromJSONTyped = GlobalWalletConnectionStatusEnumFromJSONTyped;
exports.GlobalWalletConnectionStatusEnumToJSON = GlobalWalletConnectionStatusEnumToJSON;
