// @dynamic-labs/sdk-api-core/src/models/EmbeddedWalletVersionEnum.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
var EmbeddedWalletVersionEnum;
(function (EmbeddedWalletVersionEnum) {
    EmbeddedWalletVersionEnum["V1"] = "V1";
    EmbeddedWalletVersionEnum["V2"] = "V2";
    EmbeddedWalletVersionEnum["V3"] = "V3";
})(EmbeddedWalletVersionEnum || (EmbeddedWalletVersionEnum = {}));
function EmbeddedWalletVersionEnumFromJSON(json) {
    return EmbeddedWalletVersionEnumFromJSONTyped(json);
}
function EmbeddedWalletVersionEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function EmbeddedWalletVersionEnumToJSON(value) {
    return value;
}

export { EmbeddedWalletVersionEnum, EmbeddedWalletVersionEnumFromJSON, EmbeddedWalletVersionEnumFromJSONTyped, EmbeddedWalletVersionEnumToJSON };
