// @dynamic-labs/sdk-api-core/src/models/JwtVerifiedCredentialFormatEnum.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.JwtVerifiedCredentialFormatEnum = void 0;
(function (JwtVerifiedCredentialFormatEnum) {
    JwtVerifiedCredentialFormatEnum["Blockchain"] = "blockchain";
    JwtVerifiedCredentialFormatEnum["Email"] = "email";
    JwtVerifiedCredentialFormatEnum["Oauth"] = "oauth";
    JwtVerifiedCredentialFormatEnum["Passkey"] = "passkey";
    JwtVerifiedCredentialFormatEnum["PhoneNumber"] = "phoneNumber";
    JwtVerifiedCredentialFormatEnum["ExternalUser"] = "externalUser";
})(exports.JwtVerifiedCredentialFormatEnum || (exports.JwtVerifiedCredentialFormatEnum = {}));
function JwtVerifiedCredentialFormatEnumFromJSON(json) {
    return JwtVerifiedCredentialFormatEnumFromJSONTyped(json);
}
function JwtVerifiedCredentialFormatEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function JwtVerifiedCredentialFormatEnumToJSON(value) {
    return value;
}

exports.JwtVerifiedCredentialFormatEnumFromJSON = JwtVerifiedCredentialFormatEnumFromJSON;
exports.JwtVerifiedCredentialFormatEnumFromJSONTyped = JwtVerifiedCredentialFormatEnumFromJSONTyped;
exports.JwtVerifiedCredentialFormatEnumToJSON = JwtVerifiedCredentialFormatEnumToJSON;
