// @dynamic-labs/sdk-api-core/src/models/MFAAction.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.MFAAction = void 0;
(function (MFAAction) {
    MFAAction["WalletWaasSign"] = "wallet.waas.sign";
    MFAAction["WalletWaasExport"] = "wallet.waas.export";
    MFAAction["WalletWaasReshare"] = "wallet.waas.reshare";
    MFAAction["WalletWaasRefresh"] = "wallet.waas.refresh";
    MFAAction["UserUpdate"] = "user.update";
})(exports.MFAAction || (exports.MFAAction = {}));
function MFAActionFromJSON(json) {
    return MFAActionFromJSONTyped(json);
}
function MFAActionFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function MFAActionToJSON(value) {
    return value;
}

exports.MFAActionFromJSON = MFAActionFromJSON;
exports.MFAActionFromJSONTyped = MFAActionFromJSONTyped;
exports.MFAActionToJSON = MFAActionToJSON;
