// @dynamic-labs/sdk-api-core/src/models/MfaBackupCodeAcknowledgement.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.MfaBackupCodeAcknowledgement = void 0;
(function (MfaBackupCodeAcknowledgement) {
    MfaBackupCodeAcknowledgement["Pending"] = "pending";
    MfaBackupCodeAcknowledgement["Complete"] = "complete";
})(exports.MfaBackupCodeAcknowledgement || (exports.MfaBackupCodeAcknowledgement = {}));
function MfaBackupCodeAcknowledgementFromJSON(json) {
    return MfaBackupCodeAcknowledgementFromJSONTyped(json);
}
function MfaBackupCodeAcknowledgementFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function MfaBackupCodeAcknowledgementToJSON(value) {
    return value;
}

exports.MfaBackupCodeAcknowledgementFromJSON = MfaBackupCodeAcknowledgementFromJSON;
exports.MfaBackupCodeAcknowledgementFromJSONTyped = MfaBackupCodeAcknowledgementFromJSONTyped;
exports.MfaBackupCodeAcknowledgementToJSON = MfaBackupCodeAcknowledgementToJSON;
