// @dynamic-labs/sdk-api-core/src/models/PasskeyTransport.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
var PasskeyTransport;
(function (PasskeyTransport) {
    PasskeyTransport["Usb"] = "usb";
    PasskeyTransport["Nfc"] = "nfc";
    PasskeyTransport["Ble"] = "ble";
    PasskeyTransport["Internal"] = "internal";
    PasskeyTransport["Hybrid"] = "hybrid";
})(PasskeyTransport || (PasskeyTransport = {}));
function PasskeyTransportFromJSON(json) {
    return PasskeyTransportFromJSONTyped(json);
}
function PasskeyTransportFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function PasskeyTransportToJSON(value) {
    return value;
}

export { PasskeyTransport, PasskeyTransportFromJSON, PasskeyTransportFromJSONTyped, PasskeyTransportToJSON };
