// @dynamic-labs/sdk-api-core/src/models/SsoProviderEnum.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.SsoProviderEnum = void 0;
(function (SsoProviderEnum) {
    SsoProviderEnum["Okta"] = "okta";
})(exports.SsoProviderEnum || (exports.SsoProviderEnum = {}));
function SsoProviderEnumFromJSON(json) {
    return SsoProviderEnumFromJSONTyped(json);
}
function SsoProviderEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function SsoProviderEnumToJSON(value) {
    return value;
}

exports.SsoProviderEnumFromJSON = SsoProviderEnumFromJSON;
exports.SsoProviderEnumFromJSONTyped = SsoProviderEnumFromJSONTyped;
exports.SsoProviderEnumToJSON = SsoProviderEnumToJSON;
