// @dynamic-labs/sdk-api-core/src/models/ZerodevBundlerProvider.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
var ZerodevBundlerProvider;
(function (ZerodevBundlerProvider) {
    ZerodevBundlerProvider["Stackup"] = "STACKUP";
    ZerodevBundlerProvider["Pimlico"] = "PIMLICO";
    ZerodevBundlerProvider["Alchemy"] = "ALCHEMY";
    ZerodevBundlerProvider["Gelato"] = "GELATO";
})(ZerodevBundlerProvider || (ZerodevBundlerProvider = {}));
function ZerodevBundlerProviderFromJSON(json) {
    return ZerodevBundlerProviderFromJSONTyped(json);
}
function ZerodevBundlerProviderFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function ZerodevBundlerProviderToJSON(value) {
    return value;
}

export { ZerodevBundlerProvider, ZerodevBundlerProviderFromJSON, ZerodevBundlerProviderFromJSONTyped, ZerodevBundlerProviderToJSON };
