// @dynamic-labs/sui-core/src/wallet/SuiWallet.js
'use client'
import { __awaiter } from '../../_virtual/_tslib.js';
import { Transaction } from '@mysten/sui/transactions';
import { Wallet } from '@dynamic-labs/wallet-connector-core';
import { MIST_PER_SUI } from '../utils/constants/constants.js';

class SuiWallet extends Wallet {
    /**
     * Sends the native balance of the wallet to the given address.
     * @param amount - The amount of balance
     * @param toAddress - The address to send the balance to.
     * @returns The transaction as base64 encoded bcs.
     */
    sendBalance(_a) {
        return __awaiter(this, arguments, void 0, function* ({ amount, toAddress, token, }) {
            var _b, _c;
            yield this._connector.connect();
            const transaction = new Transaction();
            if (token === null || token === void 0 ? void 0 : token.address) {
                const suiClient = yield this.getSuiClient();
                if (!suiClient) {
                    throw new Error('SUI client not found');
                }
                const { data: coins } = yield suiClient.getCoins({
                    coinType: token.address,
                    owner: this.address,
                });
                if (coins.length === 0) {
                    throw new Error('Token not found');
                }
                const decimals = (_b = token.decimals) !== null && _b !== void 0 ? _b : (_c = (yield suiClient.getCoinMetadata({
                    coinType: token.address,
                }))) === null || _c === void 0 ? void 0 : _c.decimals;
                const [coin] = transaction.splitCoins(coins[0].coinObjectId, [
                    BigInt(parseFloat(amount) * Math.pow(10, (decimals !== null && decimals !== void 0 ? decimals : 0))),
                ]);
                transaction.transferObjects([coin], toAddress);
                const result = yield this.signAndExecuteTransaction({
                    transaction,
                });
                return result.digest;
            }
            const mistAmount = Number(amount) * MIST_PER_SUI;
            // Let the transaction calculate gas automatically
            // https://sdk.mystenlabs.com/typescript/transaction-building/gas#gas-price
            const [coin] = transaction.splitCoins(transaction.gas, [mistAmount]);
            transaction.transferObjects([coin], toAddress);
            const result = yield this.signAndExecuteTransaction({
                transaction,
            });
            return result.digest;
        });
    }
    /**
     * Returns the Sui Client object initialized for the wallet's current network.
     * @returns The [SuiClient] object.
     */
    getSuiClient() {
        return __awaiter(this, void 0, void 0, function* () {
            return this._connector.getSuiClient();
        });
    }
    /**
     * @deprecated Use `getSuiClient` instead
     */
    getWalletClient() {
        return __awaiter(this, void 0, void 0, function* () {
            return this._connector.getSuiClient();
        });
    }
    /**
     * Returns the wallet's current active account.
     * @returns a readonly [WalletAccount] object.
     */
    getWalletAccount() {
        return __awaiter(this, void 0, void 0, function* () {
            return this._connector.getWalletAccount();
        });
    }
    /**
     * Returns the wallet's current active network.
     * @returns a string representing the active network, i.e. 'sui:devnet'.
     */
    getActiveNetwork() {
        return __awaiter(this, void 0, void 0, function* () {
            const account = yield this._connector.getWalletAccount();
            return account === null || account === void 0 ? void 0 : account.chains[0];
        });
    }
    /**
     * Signs a Sui [Transaction].
     * @param tx - The [Transaction] to sign.
     * @returns The signature of the signed transaction.
     */
    signTransaction(transaction) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this._connector.connect();
            const signedTransaction = yield this._connector.signTransactionFeature({
                transaction,
            });
            // is SignedTransaction
            if ('bytes' in signedTransaction) {
                return {
                    bytes: signedTransaction.bytes,
                    signature: signedTransaction.signature,
                };
            }
            // is SuiSignTransactionBlockOutput
            return {
                bytes: signedTransaction.transactionBlockBytes,
                signature: signedTransaction.signature,
            };
        });
    }
    signAndExecuteTransaction(_a) {
        return __awaiter(this, arguments, void 0, function* ({ transaction, legacyOptions, }) {
            return this._connector.signAndExecuteTransactionFeature({
                legacyOptions,
                transaction,
            });
        });
    }
}

export { SuiWallet };
