// @dynamic-labs/message-transport/src/index.js
'use client'
import { assertPackageVersion } from '@dynamic-labs/assert-package-version';
import { version } from '../package.js';
export { createMessageTransport } from './messageTransport/messageTransport.js';
export { applyDefaultMessageOrigin } from './messageTransport/decorators/applyDefaultMessageOrigin/applyDefaultMessageOrigin.js';
export { applyRecoveryManager, hasRecoveryManager } from './messageTransport/decorators/applyRecoveryManager/applyRecoveryManager.js';
export { applyTransportStamps } from './messageTransport/decorators/applyTransportStamps/applyTransportStamps.js';
export { makeWaitForInitEvent } from './messageTransport/decorators/makeWaitForInitEvent/makeWaitForInitEvent.js';
export { makeWaitForUnblock } from './messageTransport/decorators/makeWaitForUnblock/makeWaitForUnblock.js';
export { createRequestChannel } from './requestChannel/requestChannel.js';
export { getAckMessageType, getRejectMessageType, getResolveMessageType, isAckMessageType, isRejectMessageType, isRequestMessageType, isResolveMessageType } from './requestChannel/utils/utils.js';
export { isSerializedError } from './utils/isSerializedError/isSerializedError.js';
export { parseErrorFromTransport } from './utils/parseErrorFromTransport/parseErrorFromTransport.js';
export { parseMessageTransportData } from './utils/parseMessageTransportData/parseMessageTransportData.js';
export { serializeErrorForTransport } from './utils/serializeErrorForTransport/serializeErrorForTransport.js';
export { createEventEmitterForMessages } from './store/createEventEmitterForMessages/createEventEmitterForMessages.js';
export { createStore } from './store/store.js';
export { createStoreSetter } from './store/storeSetter/storeSetter.js';
export { isGlobalDebugEnabled, setGlobalDebugEnabled } from './debug/debug.js';
export { messageTransportDataJsonReplacer } from './messageTransportDataJson/messageTransportDataJsonReplacer.js';
export { messageTransportDataJsonReviver } from './messageTransportDataJson/messageTransportDataJsonReviver.js';

assertPackageVersion('@dynamic-labs/message-transport', version);
