// @dynamic-labs/sdk-react-core/src/lib/components/TypographyButton/TypographyButton.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../../../../_virtual/_tslib.cjs');
var jsxRuntime = require('react/jsx-runtime');
var classNames = require('../../utils/functions/classNames/classNames.cjs');
var Button = require('../Button/Button.cjs');
var Typography = require('../Typography/Typography.cjs');

const defaultTypographyVariantMap = {
    'brand-primary': 'button_primary',
    card: 'body_normal',
    danger: 'button_primary',
    primary: 'button_primary',
    secondary: 'button_secondary',
    tertiary: 'button_tertiary',
};
const defaultTypographyColorMap = {
    'brand-primary': 'primary',
    card: 'primary',
    danger: 'white',
    primary: 'primary',
    secondary: 'primary',
    tertiary: 'secondary',
};
const TypographyButton = (_a) => {
    var _b;
    var { children, typographyProps, copykey, disabled, leading, trailing, buttonVariant = Button.defaultButtonVariant, description } = _a, props = _tslib.__rest(_a, ["children", "typographyProps", "copykey", "disabled", "leading", "trailing", "buttonVariant", "description"]);
    const defaultTypographyVariant = defaultTypographyVariantMap[buttonVariant];
    const defaultTypographyColor = defaultTypographyColorMap[buttonVariant];
    return (jsxRuntime.jsxs(Button.Button, Object.assign({}, props, { disabled: disabled, buttonVariant: buttonVariant, children: [leading, jsxRuntime.jsxs("div", { className: classNames.classNames('typography-button__content', {
                    'typography-button__content--with-description': Boolean(description),
                }), children: [jsxRuntime.jsx(Typography.Typography, Object.assign({ variant: defaultTypographyVariant, as: 'span' }, typographyProps, { color: disabled
                            ? 'tertiary'
                            : (_b = typographyProps === null || typographyProps === void 0 ? void 0 : typographyProps.color) !== null && _b !== void 0 ? _b : defaultTypographyColor, copykey: copykey, children: children })), description && (jsxRuntime.jsx(Typography.Typography, { variant: 'body_small', color: 'secondary', weight: 'medium', children: description }))] }), trailing] })));
};

exports.TypographyButton = TypographyButton;
