// @dynamic-labs/sdk-api-core/src/models/AttestationConveyancePreference.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.AttestationConveyancePreference = void 0;
(function (AttestationConveyancePreference) {
    AttestationConveyancePreference["Direct"] = "direct";
    AttestationConveyancePreference["Enterprise"] = "enterprise";
    AttestationConveyancePreference["Indirect"] = "indirect";
    AttestationConveyancePreference["None"] = "none";
})(exports.AttestationConveyancePreference || (exports.AttestationConveyancePreference = {}));
function AttestationConveyancePreferenceFromJSON(json) {
    return AttestationConveyancePreferenceFromJSONTyped(json);
}
function AttestationConveyancePreferenceFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function AttestationConveyancePreferenceToJSON(value) {
    return value;
}

exports.AttestationConveyancePreferenceFromJSON = AttestationConveyancePreferenceFromJSON;
exports.AttestationConveyancePreferenceFromJSONTyped = AttestationConveyancePreferenceFromJSONTyped;
exports.AttestationConveyancePreferenceToJSON = AttestationConveyancePreferenceToJSON;
