// @dynamic-labs/sdk-api-core/src/models/ExternalWalletFundingDefaultChain.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.ExternalWalletFundingDefaultChain = void 0;
(function (ExternalWalletFundingDefaultChain) {
    ExternalWalletFundingDefaultChain["Ethereum"] = "ethereum";
    ExternalWalletFundingDefaultChain["Solana"] = "solana";
})(exports.ExternalWalletFundingDefaultChain || (exports.ExternalWalletFundingDefaultChain = {}));
function ExternalWalletFundingDefaultChainFromJSON(json) {
    return ExternalWalletFundingDefaultChainFromJSONTyped(json);
}
function ExternalWalletFundingDefaultChainFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function ExternalWalletFundingDefaultChainToJSON(value) {
    return value;
}

exports.ExternalWalletFundingDefaultChainFromJSON = ExternalWalletFundingDefaultChainFromJSON;
exports.ExternalWalletFundingDefaultChainFromJSONTyped = ExternalWalletFundingDefaultChainFromJSONTyped;
exports.ExternalWalletFundingDefaultChainToJSON = ExternalWalletFundingDefaultChainToJSON;
