// @dynamic-labs/sdk-api-core/src/models/GlobalWalletAccessControlTypeEnum.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
var GlobalWalletAccessControlTypeEnum;
(function (GlobalWalletAccessControlTypeEnum) {
    GlobalWalletAccessControlTypeEnum["Allow"] = "allow";
    GlobalWalletAccessControlTypeEnum["Block"] = "block";
})(GlobalWalletAccessControlTypeEnum || (GlobalWalletAccessControlTypeEnum = {}));
function GlobalWalletAccessControlTypeEnumFromJSON(json) {
    return GlobalWalletAccessControlTypeEnumFromJSONTyped(json);
}
function GlobalWalletAccessControlTypeEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function GlobalWalletAccessControlTypeEnumToJSON(value) {
    return value;
}

export { GlobalWalletAccessControlTypeEnum, GlobalWalletAccessControlTypeEnumFromJSON, GlobalWalletAccessControlTypeEnumFromJSONTyped, GlobalWalletAccessControlTypeEnumToJSON };
