// @dynamic-labs/sdk-api-core/src/models/ResidentKeyRequirement.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
var ResidentKeyRequirement;
(function (ResidentKeyRequirement) {
    ResidentKeyRequirement["Discouraged"] = "discouraged";
    ResidentKeyRequirement["Preferred"] = "preferred";
    ResidentKeyRequirement["Required"] = "required";
})(ResidentKeyRequirement || (ResidentKeyRequirement = {}));
function ResidentKeyRequirementFromJSON(json) {
    return ResidentKeyRequirementFromJSONTyped(json);
}
function ResidentKeyRequirementFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function ResidentKeyRequirementToJSON(value) {
    return value;
}

export { ResidentKeyRequirement, ResidentKeyRequirementFromJSON, ResidentKeyRequirementFromJSONTyped, ResidentKeyRequirementToJSON };
