// @dynamic-labs/sdk-api-core/src/models/ThresholdSignatureScheme.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
var ThresholdSignatureScheme;
(function (ThresholdSignatureScheme) {
    ThresholdSignatureScheme["TwoOfTwo"] = "TWO_OF_TWO";
    ThresholdSignatureScheme["TwoOfThree"] = "TWO_OF_THREE";
    ThresholdSignatureScheme["ThreeOfFive"] = "THREE_OF_FIVE";
})(ThresholdSignatureScheme || (ThresholdSignatureScheme = {}));
function ThresholdSignatureSchemeFromJSON(json) {
    return ThresholdSignatureSchemeFromJSONTyped(json);
}
function ThresholdSignatureSchemeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function ThresholdSignatureSchemeToJSON(value) {
    return value;
}

export { ThresholdSignatureScheme, ThresholdSignatureSchemeFromJSON, ThresholdSignatureSchemeFromJSONTyped, ThresholdSignatureSchemeToJSON };
