// @dynamic-labs/sdk-api-core/src/models/UserFieldsCheckEnum.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
var UserFieldsCheckEnum;
(function (UserFieldsCheckEnum) {
    UserFieldsCheckEnum["Email"] = "email";
    UserFieldsCheckEnum["Username"] = "username";
    UserFieldsCheckEnum["NameServiceSubdomainHandle"] = "name-service-subdomain-handle";
})(UserFieldsCheckEnum || (UserFieldsCheckEnum = {}));
function UserFieldsCheckEnumFromJSON(json) {
    return UserFieldsCheckEnumFromJSONTyped(json);
}
function UserFieldsCheckEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function UserFieldsCheckEnumToJSON(value) {
    return value;
}

export { UserFieldsCheckEnum, UserFieldsCheckEnumFromJSON, UserFieldsCheckEnumFromJSONTyped, UserFieldsCheckEnumToJSON };
