// @dynamic-labs/sdk-react-core/src/lib/utils/hooks/usePayWithDynamic/usePayWithDynamic.js
'use client'
import { __awaiter } from '../../../../../_virtual/_tslib.js';
import { useCallback } from 'react';
import '@dynamic-labs-sdk/client/core';
import '@dynamic-labs/sdk-api-core';
import '../../../client/client.js';
import '@dynamic-labs-sdk/client';
import '../../../config/ApiEndpoint.js';
import '../../constants/values.js';
import '@dynamic-labs/utils';
import '@dynamic-labs/multi-wallet';
import 'react-international-phone';
import '../../constants/colors.js';
import '@dynamic-labs/iconic';
import '@dynamic-labs/wallet-connector-core';
import 'react/jsx-runtime';
import '../../../context/ViewContext/ViewContext.js';
import { logger } from '../../../shared/logger.js';
import '@dynamic-labs/wallet-book';
import '../../../shared/consts/index.js';
import '../../../store/state/nonce/nonce.js';
import '@dynamic-labs/locale';
import '../../../store/state/dynamicContextProps/dynamicContextProps.js';
import '../../../store/state/primaryWalletId/primaryWalletId.js';
import '../../../store/state/connectedWalletsInfo/connectedWalletsInfo.js';
import '../../../events/dynamicEvents.js';
import { useOnboardingCompleteUser } from '../../../client/extension/user/useOnboardingCompleteUser/useOnboardingCompleteUser.js';
import { useProjectSettings } from '../../../client/extension/projectSettings/useProjectSettings/useProjectSettings.js';
import { useOnrampContext } from '../../../context/OnrampContext/OnrampContext.js';
import { useConnectUnifiedForFunding } from '../useConnectUnifiedForFunding/useConnectUnifiedForFunding.js';
import { useSocialAccounts } from '../useSocialAccounts/useSocialAccounts.js';
import { usePromptAmountAndFundWithWallet } from '../useFundWithWallet/usePromptAmountAndFundWithWallet/usePromptAmountAndFundWithWallet.js';
import { usePromptAmountAndFundWithExchange } from '../usePromptAndFundWithExchange/usePromptAndFundWithExchange.js';
import { useWalletItemActions } from '../useWalletItemActions/useWalletItemActions.js';
import { getWalletOptions } from '../../../store/state/walletOptions/walletOptions.js';
import '../../../context/DynamicContext/DynamicContext.js';
import '../../../store/state/loadingAndLifecycle/loadingAndLifecycle.js';
import '../../../context/CaptchaContext/CaptchaContext.js';
import '../../../context/ErrorContext/ErrorContext.js';
import '../../../context/AccessDeniedContext/AccessDeniedContext.js';
import '../../../context/AccountExistsContext/AccountExistsContext.js';
import '../../../context/UserWalletsContext/UserWalletsContext.js';
import '../../../store/state/authMode/authMode.js';
import '../../../context/VerificationContext/VerificationContext.js';
import 'react-dom';
import '../../functions/compareChains/compareChains.js';
import '../../../views/Passkey/utils/findPrimaryEmbeddedChain/findPrimaryEmbeddedChain.js';
import '../../../context/ThemeContext/ThemeContext.js';
import '../useUserUpdateRequest/useUpdateUser/userFieldsSchema.js';
import 'bs58';
import 'yup';
import '../../../context/MockContext/MockContext.js';
import '../../../views/CollectUserDataView/useFields.js';
import '../../../context/FieldsStateContext/FieldsStateContext.js';
import '../../../context/UserFieldEditorContext/UserFieldEditorContext.js';
import '@dynamic-labs/rpc-providers';
import 'react-i18next';
import '../../../components/Accordion/components/AccordionItem/AccordionItem.js';
import '../../../components/Alert/Alert.js';
import '../../../context/WalletContext/WalletContext.js';
import '../../../components/ShadowDOM/ShadowDOM.js';
import '../../../components/IconButton/IconButton.js';
import '../../../components/InlineWidget/InlineWidget.js';
import '../../../components/Input/Input.js';
import '../../../components/IsBrowser/IsBrowser.js';
import '../../../components/MenuList/Dropdown/Dropdown.js';
import '../../../components/OverlayCard/OverlayCard.js';
import '../../../components/Transition/ZoomTransition/ZoomTransition.js';
import '../../../components/Transition/SlideInUpTransition/SlideInUpTransition.js';
import '../../../components/Transition/OpacityTransition/OpacityTransition.js';
import '../../../components/PasskeyCreatedSuccessBanner/PasskeyCreatedSuccessBanner.js';
import '../../../components/Popper/Popper/Popper.js';
import '../../../components/Popper/PopperContext/PopperContext.js';
import 'react-focus-lock';
import 'qrcode';
import 'formik';
import '../useSubdomainCheck/useSubdomainCheck.js';
import '../../../context/WalletGroupContext/WalletGroupContext.js';
import '../../../context/IpConfigurationContext/IpConfigurationContext.js';
import '../../../context/SocialRedirectContext/SocialRedirectContext.js';
import '../../../context/ConnectWithOtpContext/ConnectWithOtpContext.js';
import '../../../widgets/DynamicBridgeWidget/views/WalletsView/components/SecondaryWallets/SecondaryWallets.js';
import '@hcaptcha/react-hcaptcha';
import '../../../context/LoadingContext/LoadingContext.js';
import '../../../widgets/DynamicWidget/context/DynamicWidgetContext.js';
import '../../../widgets/DynamicWidget/helpers/convertExchangeKeyAndProviderEnum.js';
import '../../../views/ExchangeWhitelistWarning/ExchangeWhitelistWarning.js';
import '../../../context/ErrorContext/hooks/useErrorText/useErrorText.js';
import '../../../context/FooterAnimationContext/index.js';
import '@dynamic-labs/types';
import '../useEmbeddedWallet/useSecureEnclaveEmbeddedWallet/constants.js';
import '../../../views/MfaChooseDeviceView/useGetMfaOptions/useGetMfaOptions.js';
import '../../../context/PasskeyContext/PasskeyContext.js';
import '../../../store/state/sendBalances.js';
import '../../../store/state/connectorsInitializing/connectorsInitializing.js';
import '../../../components/OverlayCardBase/OverlayCardTarget/OverlayCardTarget.js';
import '../../../widgets/DynamicWidget/components/DynamicWidgetHeader/DynamicWidgetHeader.js';
import '../../../views/TransactionConfirmationView/TransactionConfirmationView.js';
import '../../../widgets/DynamicWidget/components/PasskeyCard/PasskeyCard.js';
import '../../../../index.js';
import '../../../widgets/DynamicWidget/views/CryptoComOnramp/CryptoComOnramp.js';
import '../../../widgets/DynamicWidget/views/ReceiveWalletFunds/ReceiveWalletFunds.js';
import '../../../store/state/tokenBalances.js';
import '../../../store/state/multichainBalances.js';
import '../../../shared/utils/functions/getInitialUrl/getInitialUrl.js';
import { useInternalDynamicContext } from '../../../context/DynamicContext/useDynamicContext/useInternalDynamicContext/useInternalDynamicContext.js';

const usePayWithDynamic = () => {
    const connectUnifiedForFunding = useConnectUnifiedForFunding();
    const projectSettings = useProjectSettings();
    const { linkSocialAccount, signInWithSocialAccount } = useSocialAccounts();
    const user = useOnboardingCompleteUser();
    const { openOnramp, setEnabledOnrampProvidersTargetValue } = useOnrampContext();
    const promptWallet = usePromptAmountAndFundWithWallet();
    const promptExchange = usePromptAmountAndFundWithExchange();
    const { handleWalletItemClick } = useWalletItemActions();
    const { walletConnectorOptions } = getWalletOptions();
    const { setShowAuthFlow, setShowDynamicUserProfile } = useInternalDynamicContext();
    // Flow for funding via wallet
    const handleWalletFlow = useCallback(({ wallet, props, onBack }) => {
        // If user is not onboarded, trigger wallet connector UI
        if (!user) {
            const option = walletConnectorOptions.find(({ key }) => key === wallet.connector.key);
            if (option)
                handleWalletItemClick(option);
        }
        // Prompt for amount and execute wallet funding (shows deposit screen)
        promptWallet({
            amount: { rule: 'exact', value: props.tokenAmount },
            destinationAddress: props.destinationAddress,
            network: props.network,
            onBack,
            token: { rule: 'exact-with-amount', value: props.tokenSymbol },
            wallet: wallet,
        });
        setShowAuthFlow(false);
    }, [
        handleWalletItemClick,
        promptWallet,
        setShowAuthFlow,
        user,
        walletConnectorOptions,
    ]);
    // Flow for linking or signing in with social account before exchange
    const handleSocialConnect = useCallback((_a) => __awaiter(void 0, [_a], void 0, function* ({ provider, exchange, props, isSocialProviderAlreadyVerified, }) {
        // if connected social return since already signed in
        if (isSocialProviderAlreadyVerified) {
            return false;
        }
        const payload = {
            payingWithDynamic: {
                chainName: props.chainName,
                destinationAddress: props.destinationAddress,
                network: props.network,
                tokenAmount: props.tokenAmount,
                tokenSymbol: props.tokenSymbol,
            },
            triggerFundFromExchangeOnSuccess: exchange,
        };
        if (user) {
            // link if user exists
            yield linkSocialAccount(provider, payload);
        }
        else {
            // sign in if no user
            yield signInWithSocialAccount(provider, payload);
        }
        return true;
    }), [linkSocialAccount, signInWithSocialAccount, user]);
    // exchange flow
    const handleExchangeFlow = useCallback((_b) => __awaiter(void 0, [_b], void 0, function* ({ exchange, props, onBack }) {
        var _c, _d, _e, _f, _g;
        const fullExchange = (_c = projectSettings === null || projectSettings === void 0 ? void 0 : projectSettings.exchanges) === null || _c === void 0 ? void 0 : _c.find((e) => e.exchange === exchange);
        // Social provider for this exchange, if any
        const provider = fullExchange === null || fullExchange === void 0 ? void 0 : fullExchange.socialProvider;
        const isProviderSocialConnectEnabled = provider &&
            Boolean((_e = (_d = projectSettings === null || projectSettings === void 0 ? void 0 : projectSettings.providers) === null || _d === void 0 ? void 0 : _d.find((p) => p.provider === provider)) === null || _e === void 0 ? void 0 : _e.enabledAt);
        // Check if user already verified this social provider
        const isSocialProviderAlreadyVerified = (_g = (_f = user === null || user === void 0 ? void 0 : user.verifiedCredentials) === null || _f === void 0 ? void 0 : _f.some((c) => c.oauthProvider === provider)) !== null && _g !== void 0 ? _g : false;
        // If exchange requires social connect, handle that first
        if (isProviderSocialConnectEnabled && provider) {
            const handled = yield handleSocialConnect({
                exchange,
                isSocialProviderAlreadyVerified,
                props,
                provider,
            });
            // Social connect flow opened, stop here
            if (handled)
                return;
            // If already connected, prompt amount and fund with exchange
            promptExchange({
                exchange,
                onBack,
                payingWithDynamic: props,
            });
            return;
        }
        // No social provider = use on-ramp provider if configured
        if (fullExchange === null || fullExchange === void 0 ? void 0 : fullExchange.onRampProvider) {
            yield openOnramp({
                address: props.destinationAddress,
                chainName: props.chainName,
                network: String(props.network),
                onrampProvider: fullExchange.onRampProvider,
                overrideOnRamp: true,
                payingWithDynamic: true,
                token: props.tokenSymbol,
                tokenAmount: props.tokenAmount,
            });
        }
    }), [
        projectSettings === null || projectSettings === void 0 ? void 0 : projectSettings.exchanges,
        projectSettings === null || projectSettings === void 0 ? void 0 : projectSettings.providers,
        user === null || user === void 0 ? void 0 : user.verifiedCredentials,
        handleSocialConnect,
        promptExchange,
        openOnramp,
    ]);
    const payWithDynamic = useCallback((props) => __awaiter(void 0, void 0, void 0, function* () {
        try {
            setEnabledOnrampProvidersTargetValue({
                address: props.destinationAddress,
                chainName: props.chainName,
                network: String(props.network),
                token: props.tokenSymbol,
                tokenAmount: props.tokenAmount,
            });
            const result = yield connectUnifiedForFunding({
                initiatedByWidget: false,
            });
            if (result.kind === 'wallet') {
                handleWalletFlow({
                    onBack: () => {
                        setShowAuthFlow(true);
                        setShowDynamicUserProfile(false);
                        payWithDynamic(props);
                    },
                    props,
                    wallet: result.wallet,
                });
            }
            else if (result.kind === 'exchange') {
                yield handleExchangeFlow({
                    exchange: result.exchange,
                    onBack: () => {
                        setShowAuthFlow(true);
                        setShowDynamicUserProfile(false);
                        payWithDynamic(props);
                    },
                    props,
                });
            }
            else {
                yield openOnramp({
                    address: props.destinationAddress,
                    chainName: props.chainName,
                    network: String(props.network),
                    onrampProvider: result.onramp.id,
                    payingWithDynamic: true,
                    token: props.tokenSymbol,
                    tokenAmount: props.tokenAmount,
                });
            }
        }
        catch (error) {
            logger.error('usePayWithDynamic failed', error);
        }
    }), [
        setEnabledOnrampProvidersTargetValue,
        connectUnifiedForFunding,
        handleWalletFlow,
        setShowAuthFlow,
        setShowDynamicUserProfile,
        handleExchangeFlow,
        openOnramp,
    ]);
    return payWithDynamic;
};

export { usePayWithDynamic };
