// @dynamic-labs/sdk-api-core/src/models/AuthenticatorAttachment.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
var AuthenticatorAttachment;
(function (AuthenticatorAttachment) {
    AuthenticatorAttachment["CrossPlatform"] = "cross-platform";
    AuthenticatorAttachment["Platform"] = "platform";
})(AuthenticatorAttachment || (AuthenticatorAttachment = {}));
function AuthenticatorAttachmentFromJSON(json) {
    return AuthenticatorAttachmentFromJSONTyped(json);
}
function AuthenticatorAttachmentFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function AuthenticatorAttachmentToJSON(value) {
    return value;
}

export { AuthenticatorAttachment, AuthenticatorAttachmentFromJSON, AuthenticatorAttachmentFromJSONTyped, AuthenticatorAttachmentToJSON };
