// @dynamic-labs/sdk-api-core/src/models/PasskeyTransport.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.PasskeyTransport = void 0;
(function (PasskeyTransport) {
    PasskeyTransport["Usb"] = "usb";
    PasskeyTransport["Nfc"] = "nfc";
    PasskeyTransport["Ble"] = "ble";
    PasskeyTransport["Internal"] = "internal";
    PasskeyTransport["Hybrid"] = "hybrid";
})(exports.PasskeyTransport || (exports.PasskeyTransport = {}));
function PasskeyTransportFromJSON(json) {
    return PasskeyTransportFromJSONTyped(json);
}
function PasskeyTransportFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function PasskeyTransportToJSON(value) {
    return value;
}

exports.PasskeyTransportFromJSON = PasskeyTransportFromJSON;
exports.PasskeyTransportFromJSONTyped = PasskeyTransportFromJSONTyped;
exports.PasskeyTransportToJSON = PasskeyTransportToJSON;
