// @dynamic-labs/sdk-api-core/src/models/PasskeyCredentialHint.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
var PasskeyCredentialHint;
(function (PasskeyCredentialHint) {
    PasskeyCredentialHint["Hybrid"] = "hybrid";
    PasskeyCredentialHint["SecurityKey"] = "security-key";
    PasskeyCredentialHint["ClientDevice"] = "client-device";
})(PasskeyCredentialHint || (PasskeyCredentialHint = {}));
function PasskeyCredentialHintFromJSON(json) {
    return PasskeyCredentialHintFromJSONTyped(json);
}
function PasskeyCredentialHintFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function PasskeyCredentialHintToJSON(value) {
    return value;
}

export { PasskeyCredentialHint, PasskeyCredentialHintFromJSON, PasskeyCredentialHintFromJSONTyped, PasskeyCredentialHintToJSON };
