// @dynamic-labs/sdk-api-core/src/models/PasswordSourceTypeEnum.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.PasswordSourceTypeEnum = void 0;
(function (PasswordSourceTypeEnum) {
    PasswordSourceTypeEnum["Dynamic"] = "dynamic";
    PasswordSourceTypeEnum["User"] = "user";
})(exports.PasswordSourceTypeEnum || (exports.PasswordSourceTypeEnum = {}));
function PasswordSourceTypeEnumFromJSON(json) {
    return PasswordSourceTypeEnumFromJSONTyped(json);
}
function PasswordSourceTypeEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function PasswordSourceTypeEnumToJSON(value) {
    return value;
}

exports.PasswordSourceTypeEnumFromJSON = PasswordSourceTypeEnumFromJSON;
exports.PasswordSourceTypeEnumFromJSONTyped = PasswordSourceTypeEnumFromJSONTyped;
exports.PasswordSourceTypeEnumToJSON = PasswordSourceTypeEnumToJSON;
