// @dynamic-labs/sdk-api-core/src/models/PasskeyCredentialType.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.PasskeyCredentialType = void 0;
(function (PasskeyCredentialType) {
    PasskeyCredentialType["PublicKey"] = "public-key";
})(exports.PasskeyCredentialType || (exports.PasskeyCredentialType = {}));
function PasskeyCredentialTypeFromJSON(json) {
    return PasskeyCredentialTypeFromJSONTyped(json);
}
function PasskeyCredentialTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function PasskeyCredentialTypeToJSON(value) {
    return value;
}

exports.PasskeyCredentialTypeFromJSON = PasskeyCredentialTypeFromJSON;
exports.PasskeyCredentialTypeFromJSONTyped = PasskeyCredentialTypeFromJSONTyped;
exports.PasskeyCredentialTypeToJSON = PasskeyCredentialTypeToJSON;
