// @dynamic-labs/sdk-api-core/src/models/SdkViewSectionType.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
var SdkViewSectionType;
(function (SdkViewSectionType) {
    SdkViewSectionType["Email"] = "email";
    SdkViewSectionType["EmailAndPhone"] = "emailAndPhone";
    SdkViewSectionType["Phone"] = "phone";
    SdkViewSectionType["Separator"] = "separator";
    SdkViewSectionType["Social"] = "social";
    SdkViewSectionType["Text"] = "text";
    SdkViewSectionType["Wallet"] = "wallet";
    SdkViewSectionType["Passkey"] = "passkey";
})(SdkViewSectionType || (SdkViewSectionType = {}));
function SdkViewSectionTypeFromJSON(json) {
    return SdkViewSectionTypeFromJSONTyped(json);
}
function SdkViewSectionTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function SdkViewSectionTypeToJSON(value) {
    return value;
}

export { SdkViewSectionType, SdkViewSectionTypeFromJSON, SdkViewSectionTypeFromJSONTyped, SdkViewSectionTypeToJSON };
