// @dynamic-labs/sdk-react-core/src/lib/utils/functions/getTransactionLink/blockExplorerPatterns.js
'use client'
/**
 * Block explorer URL patterns for different chains
 * This mapping allows for easy configuration and extension of transaction link formats
 */
/**
 * Predefined patterns for known block explorers
 * Add new patterns here as needed for different chains
 */
const BLOCK_EXPLORER_PATTERNS = [
    {
        description: 'TRON - Uses hash fragment format',
        domain: 'tronscan.org',
        generateUrl: (baseUrl, hash) => {
            const url = new URL(baseUrl);
            return `${url.origin}/#/transaction/${hash}`;
        },
    },
];
/**
 * Default pattern for standard block explorers
 * Used when no specific pattern matches
 */
const DEFAULT_PATTERN = {
    description: 'Standard format used by most block explorers',
    domain: '*',
    generateUrl: (baseUrl, hash) => {
        const url = new URL(baseUrl);
        url.pathname = `/tx/${hash}`;
        return url.toString();
    },
};
/**
 * Find the appropriate pattern for a given block explorer URL
 */
const findBlockExplorerPattern = (blockExplorerUrl) => {
    const pattern = BLOCK_EXPLORER_PATTERNS.find((pattern) => blockExplorerUrl.includes(pattern.domain));
    return pattern || DEFAULT_PATTERN;
};

export { BLOCK_EXPLORER_PATTERNS, DEFAULT_PATTERN, findBlockExplorerPattern };
