// @dynamic-labs/embedded-wallet/src/lib/ExportHandler/TurnkeyExportHandler.js
'use client'
import { __awaiter } from '../../../_virtual/_tslib.js';
import { TurnkeyClient } from '@turnkey/http';
import { IframeStamper, KeyFormat } from '@turnkey/iframe-stamper';
import { getTLD, PlatformService, DynamicError } from '@dynamic-labs/utils';
import { turnkeyAuthenticatorRecoveryHandler } from '../AuthenticatorHandler/TurnkeyAuthenticatorRecoveryHandler.js';
import { BaseTurnkeyHandler } from '../BaseTurnkeyHandler.js';
import { TURNKEY_API_BASE_URL } from '../constants.js';
import '../utils/convertAttestationTransports/convertAttestationTransports.js';
import { logger } from '../utils/logger/logger.js';
import { TurnkeyPasskeyService } from '../utils/TurnkeyPasskeyService/TurnkeyPasskeyService.js';

const turnkeyExportUrl = 'https://export.turnkey.com';
class TurnkeyExportHandler extends BaseTurnkeyHandler {
    initExport(iframeContainer, iframeElementId) {
        return __awaiter(this, void 0, void 0, function* () {
            this.__iframeStamper = new IframeStamper({
                iframeContainer,
                iframeElementId,
                iframeUrl: turnkeyExportUrl,
            });
            yield this.__iframeStamper.init();
            this.__publicKey = this.__iframeStamper.publicKey();
            if (turnkeyAuthenticatorRecoveryHandler.isSessionActive()) {
                this.__turnkeyClient = turnkeyAuthenticatorRecoveryHandler.client;
            }
            else {
                let rpId = getTLD();
                if (!rpId) {
                    rpId = PlatformService.getHostname();
                }
                const passkeyStamper = TurnkeyPasskeyService.createWebauthnStamper({
                    rpId,
                });
                const apiKeyStamper = TurnkeyExportHandler === null || TurnkeyExportHandler === void 0 ? void 0 : TurnkeyExportHandler.apiKeyStamper;
                const stamper = apiKeyStamper !== null && apiKeyStamper !== void 0 ? apiKeyStamper : passkeyStamper;
                this.__turnkeyClient = new TurnkeyClient({
                    baseUrl: TURNKEY_API_BASE_URL,
                }, stamper);
            }
            return this.__publicKey;
        });
    }
    verifyExportWallet(_a) {
        return __awaiter(this, arguments, void 0, function* ({ exportBundle, organizationId, }) {
            if (!this.__iframeStamper) {
                throw new DynamicError('Cannot proceed with your request');
            }
            try {
                return yield this.__iframeStamper.injectWalletExportBundle(exportBundle, organizationId);
            }
            catch (err) {
                logger.error('Error while verifying export wallet', err);
                throw err;
            }
        });
    }
    verifyExportPrivateKey(_a) {
        return __awaiter(this, arguments, void 0, function* ({ exportBundle, organizationId, chain, }) {
            if (!this.__iframeStamper) {
                throw new DynamicError('Cannot proceed with your request');
            }
            const keyFormat = chain === 'solana' || chain === 'SOL'
                ? KeyFormat.Solana
                : KeyFormat.Hexadecimal;
            try {
                return yield this.__iframeStamper.injectKeyExportBundle(exportBundle, organizationId, keyFormat);
            }
            catch (err) {
                logger.error('Error while verifying export private key', err);
                throw err;
            }
        });
    }
    exportPrivateKey(_a) {
        return __awaiter(this, arguments, void 0, function* ({ privateKeyId, organizationId, }) {
            const apiKeyStamper = TurnkeyExportHandler === null || TurnkeyExportHandler === void 0 ? void 0 : TurnkeyExportHandler.apiKeyStamper;
            if (apiKeyStamper) {
                this.__turnkeyClient = new TurnkeyClient({
                    baseUrl: TURNKEY_API_BASE_URL,
                }, apiKeyStamper);
            }
            if (!this.__iframeStamper ||
                !this.__publicKey ||
                !this.__turnkeyClient ||
                !privateKeyId) {
                throw new DynamicError('Cannot proceed with your request');
            }
            try {
                const newActivity = yield this.__turnkeyClient.exportPrivateKey({
                    organizationId,
                    parameters: { privateKeyId, targetPublicKey: this.__publicKey },
                    timestampMs: String(Date.now()),
                    type: 'ACTIVITY_TYPE_EXPORT_PRIVATE_KEY',
                });
                return newActivity.activity;
            }
            catch (err) {
                logger.error('[TK] Error while completing export private key process', err);
                throw err;
            }
        });
    }
    exportWallet(_a) {
        return __awaiter(this, arguments, void 0, function* ({ walletId, organizationId, address, }) {
            const apiKeyStamper = TurnkeyExportHandler === null || TurnkeyExportHandler === void 0 ? void 0 : TurnkeyExportHandler.apiKeyStamper;
            if (apiKeyStamper) {
                this.__turnkeyClient = new TurnkeyClient({
                    baseUrl: TURNKEY_API_BASE_URL,
                }, apiKeyStamper);
            }
            if (!this.__iframeStamper ||
                !this.__publicKey ||
                !this.__turnkeyClient ||
                !walletId) {
                throw new DynamicError('Cannot proceed with your request');
            }
            try {
                if (address) {
                    const newActivity = yield this.__turnkeyClient.exportWalletAccount({
                        organizationId,
                        parameters: { address, targetPublicKey: this.__publicKey },
                        timestampMs: String(Date.now()),
                        type: 'ACTIVITY_TYPE_EXPORT_WALLET_ACCOUNT',
                    });
                    return newActivity.activity;
                }
                const newActivity = yield this.__turnkeyClient.exportWallet({
                    organizationId,
                    parameters: { targetPublicKey: this.__publicKey, walletId },
                    timestampMs: String(Date.now()),
                    type: 'ACTIVITY_TYPE_EXPORT_WALLET',
                });
                return newActivity.activity;
            }
            catch (err) {
                logger.error('[TK] Error while completing export wallet process', err);
                throw err;
            }
        });
    }
}

export { TurnkeyExportHandler };
