// @dynamic-labs/sdk-react-core/src/lib/widgets/DynamicWidget/views/CryptoComOnramp/CryptoComOnramp.js
'use client'
import { __awaiter } from '../../../../../../_virtual/_tslib.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import { useState, useRef, useEffect, useCallback } from 'react';
import { useTranslation } from 'react-i18next';
import { WalletIcon } from '@dynamic-labs/wallet-book';
import { ProviderEnum } from '@dynamic-labs/sdk-api-core';
import { Icon } from '../../../../components/Icon/Icon.js';
import { ModalHeader } from '../../../../components/ModalHeader/ModalHeader.js';
import { Typography } from '../../../../components/Typography/Typography.js';
import { TypographyButton } from '../../../../components/TypographyButton/TypographyButton.js';
import '../../../../context/DynamicContext/DynamicContext.js';
import '../../../../store/state/loadingAndLifecycle/loadingAndLifecycle.js';
import '@dynamic-labs/iconic';
import '@dynamic-labs/wallet-connector-core';
import { ReactComponent as SvgChevronDown } from '../../../../shared/assets/chevron-down.js';
import '../../../../context/ViewContext/ViewContext.js';
import '../../../../shared/logger.js';
import '@dynamic-labs/utils';
import '../../../../utils/constants/colors.js';
import '../../../../utils/constants/values.js';
import '../../../../shared/consts/index.js';
import '../../../../events/dynamicEvents.js';
import '../../../../context/CaptchaContext/CaptchaContext.js';
import '../../../../context/ErrorContext/ErrorContext.js';
import '@dynamic-labs/multi-wallet';
import 'react-international-phone';
import '../../../../store/state/nonce/nonce.js';
import '@dynamic-labs-sdk/client/core';
import '../../../../client/client.js';
import '@dynamic-labs-sdk/client';
import '../../../../config/ApiEndpoint.js';
import { getOnrampProviders } from '../../../../data/api/onramp/onramp.js';
import '@dynamic-labs/locale';
import { useEnvironmentId } from '../../../../store/state/dynamicContextProps/dynamicContextProps.js';
import '../../../../store/state/primaryWalletId/primaryWalletId.js';
import '../../../../store/state/connectedWalletsInfo/connectedWalletsInfo.js';
import '../../../../context/AccessDeniedContext/AccessDeniedContext.js';
import '../../../../context/AccountExistsContext/AccountExistsContext.js';
import '../../../../context/UserWalletsContext/UserWalletsContext.js';
import '../../../../store/state/authMode/authMode.js';
import '../../../../context/VerificationContext/VerificationContext.js';
import 'react-dom';
import '../../../../utils/functions/compareChains/compareChains.js';
import '../../../../views/Passkey/utils/findPrimaryEmbeddedChain/findPrimaryEmbeddedChain.js';
import '../../../../context/ThemeContext/ThemeContext.js';
import '../../../../utils/hooks/useUserUpdateRequest/useUpdateUser/userFieldsSchema.js';
import 'bs58';
import '@dynamic-labs/types';
import '../../../../context/SocialRedirectContext/SocialRedirectContext.js';
import '../../../../context/LoadingContext/LoadingContext.js';
import '../../../../context/WalletContext/WalletContext.js';
import '../../../../utils/hooks/useEmbeddedWallet/useSecureEnclaveEmbeddedWallet/constants.js';
import 'yup';
import '../../../../context/MockContext/MockContext.js';
import '../../../../views/CollectUserDataView/useFields.js';
import '../../../../context/FieldsStateContext/FieldsStateContext.js';
import '../../../../context/UserFieldEditorContext/UserFieldEditorContext.js';
import '@dynamic-labs/rpc-providers';
import '../../../../store/state/walletOptions/walletOptions.js';
import '../../../../components/Accordion/components/AccordionItem/AccordionItem.js';
import '../../../../components/Alert/Alert.js';
import '../../../../components/ShadowDOM/ShadowDOM.js';
import '../../../../components/IconButton/IconButton.js';
import '../../../../components/InlineWidget/InlineWidget.js';
import '../../../../components/Input/Input.js';
import '../../../../components/IsBrowser/IsBrowser.js';
import '../../../../components/MenuList/Dropdown/Dropdown.js';
import '../../../../components/OverlayCard/OverlayCard.js';
import '../../../../components/Transition/ZoomTransition/ZoomTransition.js';
import '../../../../components/Transition/SlideInUpTransition/SlideInUpTransition.js';
import '../../../../components/Transition/OpacityTransition/OpacityTransition.js';
import '../../../../components/PasskeyCreatedSuccessBanner/PasskeyCreatedSuccessBanner.js';
import '../../../../components/Popper/Popper/Popper.js';
import '../../../../components/Popper/PopperContext/PopperContext.js';
import 'react-focus-lock';
import 'qrcode';
import { shortenWalletAddress } from '../../../../shared/utils/functions/shortenWalletAddress/shortenWalletAddress.js';
import 'formik';
import '../../../../utils/hooks/useSubdomainCheck/useSubdomainCheck.js';
import '../../../../context/WalletGroupContext/WalletGroupContext.js';
import '../../../../context/IpConfigurationContext/IpConfigurationContext.js';
import '../../../../context/ConnectWithOtpContext/ConnectWithOtpContext.js';
import '../../../DynamicBridgeWidget/views/WalletsView/components/SecondaryWallets/SecondaryWallets.js';
import '@hcaptcha/react-hcaptcha';
import { useWidgetContext } from '../../context/DynamicWidgetContext.js';
import '../../helpers/convertExchangeKeyAndProviderEnum.js';
import '../../../../views/ExchangeWhitelistWarning/ExchangeWhitelistWarning.js';
import '../../../../context/ErrorContext/hooks/useErrorText/useErrorText.js';
import '../../../../context/FooterAnimationContext/index.js';
import '../../../../views/MfaChooseDeviceView/useGetMfaOptions/useGetMfaOptions.js';
import '../../../../context/PasskeyContext/PasskeyContext.js';
import '../../../../context/OnrampContext/OnrampContext.js';
import '../../../../store/state/sendBalances.js';
import '../../../../store/state/connectorsInitializing/connectorsInitializing.js';
import '../../../../components/OverlayCardBase/OverlayCardTarget/OverlayCardTarget.js';
import '../../components/DynamicWidgetHeader/DynamicWidgetHeader.js';
import '../../../../views/TransactionConfirmationView/TransactionConfirmationView.js';
import '../../components/PasskeyCard/PasskeyCard.js';
import '../../../../../index.js';
import '../ReceiveWalletFunds/ReceiveWalletFunds.js';
import '../../../../store/state/tokenBalances.js';
import '../../../../store/state/multichainBalances.js';
import '../../../../shared/utils/functions/getInitialUrl/getInitialUrl.js';
import { useInternalDynamicContext } from '../../../../context/DynamicContext/useDynamicContext/useInternalDynamicContext/useInternalDynamicContext.js';
import { FormattedInput } from '../ReceiveWalletFunds/FormattedInput/FormattedInput.js';
import { createFormattedInputEmitter } from '../ReceiveWalletFunds/FormattedInput/formattedInputEmitter.js';
import { formatValue } from '../ReceiveWalletFunds/utils/formatValue/formatValue.js';
import { isFiatOrStablecoin } from '../ReceiveWalletFunds/utils/isFiatToken/isFiatToken.js';
import { createFundingButtons } from '../ReceiveWalletFunds/utils/createFundingButtons/createFundingButtons.js';
import { CurrencySelectScreen } from './CurrencySelectScreen/CurrencySelectScreen.js';

const defaultQuickSuggestions = [25, 100, 500];
const availableCurrencies = [
    { code: 'USD', name: 'US Dollar', symbol: '$' },
    { code: 'EUR', name: 'Euro', symbol: '€' },
    { code: 'AUD', name: 'Australian Dollar', symbol: 'A$' },
    { code: 'CAD', name: 'Canadian Dollar', symbol: 'C$' },
    { code: 'GBP', name: 'British Pound', symbol: '£' },
    { code: 'BGN', name: 'Bulgarian Lev', symbol: 'лв' },
    { code: 'BRL', name: 'Brazilian Real', symbol: 'R$' },
    { code: 'CHF', name: 'Swiss Franc', symbol: 'CHF' },
    { code: 'CLP', name: 'Chilean Peso', symbol: '$' },
    { code: 'CNY', name: 'Chinese Yuan', symbol: '¥' },
    { code: 'COP', name: 'Colombian Peso', symbol: '$' },
    { code: 'DKK', name: 'Danish Krone', symbol: 'kr' },
    { code: 'HKD', name: 'Hong Kong Dollar', symbol: 'HK$' },
    { code: 'IDR', name: 'Indonesian Rupiah', symbol: 'Rp' },
    { code: 'INR', name: 'Indian Rupee', symbol: '₹' },
    { code: 'JPY', name: 'Japanese Yen', symbol: '¥' },
    { code: 'KRW', name: 'South Korean Won', symbol: '₩' },
    { code: 'MDL', name: 'Moldovan Leu', symbol: 'L' },
    { code: 'MOP', name: 'Macanese Pataca', symbol: 'MOP$' },
    { code: 'MXN', name: 'Mexican Peso', symbol: '$' },
    { code: 'MYR', name: 'Malaysian Ringgit', symbol: 'RM' },
    { code: 'NOK', name: 'Norwegian Krone', symbol: 'kr' },
    { code: 'PHP', name: 'Philippine Peso', symbol: '₱' },
    { code: 'RON', name: 'Romanian Leu', symbol: 'lei' },
    { code: 'RUB', name: 'Russian Ruble', symbol: '₽' },
    { code: 'SAR', name: 'Saudi Riyal', symbol: '﷼' },
    { code: 'SEK', name: 'Swedish Krona', symbol: 'kr' },
    { code: 'SGD', name: 'Singapore Dollar', symbol: 'S$' },
    { code: 'THB', name: 'Thai Baht', symbol: '฿' },
    { code: 'TRY', name: 'Turkish Lira', symbol: '₺' },
    { code: 'TWD', name: 'Taiwan Dollar', symbol: 'NT$' },
    { code: 'UAH', name: 'Ukrainian Hryvnia', symbol: '₴' },
    { code: 'ZAR', name: 'South African Rand', symbol: 'R' },
];
const inputEmitter = createFormattedInputEmitter();
const isCryptoComUrl = (url) => {
    try {
        const parsedUrl = new URL(url);
        return (parsedUrl.protocol === 'https:' &&
            (parsedUrl.hostname === 'crypto.com' ||
                parsedUrl.hostname.endsWith('.crypto.com')));
    }
    catch (_a) {
        return false;
    }
};
const CryptoComOnramp = ({ onBack, hideBackButton = false, onPaymentCreated, quickSuggestions = defaultQuickSuggestions, currency: initialCurrency, }) => {
    var _a, _b, _c, _d;
    const { t } = useTranslation();
    const { primaryWallet, setShowAuthFlow, setShowDynamicUserProfile, network } = useInternalDynamicContext();
    const { setDynamicWidgetView } = useWidgetContext();
    const environmentId = useEnvironmentId();
    const [hasUserInteracted, setHasUserInteracted] = useState(false);
    const [isCreatingPayment, setIsCreatingPayment] = useState(false);
    const [error, setError] = useState(null);
    const isCreatingPaymentRef = useRef(false);
    // Amount state management
    const [fiatAmount, setFiatAmount] = useState('');
    const [selectedCurrency, setSelectedCurrency] = useState(initialCurrency &&
        availableCurrencies.some((c) => c.code === initialCurrency)
        ? initialCurrency
        : 'USD');
    useEffect(() => {
        if (initialCurrency &&
            availableCurrencies.some((c) => c.code === initialCurrency)) {
            setSelectedCurrency(initialCurrency);
        }
    }, [initialCurrency]);
    const fiatSymbol = ((_a = availableCurrencies.find((c) => c.code === selectedCurrency)) === null || _a === void 0 ? void 0 : _a.code) || 'USD';
    const currencySymbol = ((_b = availableCurrencies.find((c) => c.code === selectedCurrency)) === null || _b === void 0 ? void 0 : _b.symbol) || '$';
    // Network selection state
    const [selectedNetworkId, setSelectedNetworkId] = useState(network);
    useEffect(() => {
        setSelectedNetworkId(network);
    }, [network]);
    const setAmountsByFiatValue = useCallback((amount) => {
        setFiatAmount(amount);
    }, []);
    const quickSuggestionsParsed = Array.isArray(quickSuggestions)
        ? quickSuggestions.map((value) => ({ display: `$${value}`, value }))
        : [];
    const showQuickSuggestions = quickSuggestionsParsed.length > 0 &&
        !hasUserInteracted &&
        selectedCurrency === 'USD';
    const [showCurrencySelect, setShowCurrencySelect] = useState(false);
    const handleQuickSuggestion = useCallback((value) => {
        setAmountsByFiatValue(value.toString());
        setHasUserInteracted(true);
    }, [setAmountsByFiatValue]);
    const handleBack = useCallback(() => {
        if (onBack) {
            onBack();
        }
        else {
            setDynamicWidgetView('choose-onramp-provider');
        }
    }, [onBack, setDynamicWidgetView]);
    // Primary data - only show the deposit amount and selected token
    const primaryData = {
        amount: fiatAmount,
        setAmount: setAmountsByFiatValue,
        symbol: fiatSymbol,
    };
    const inputSymbolProp = selectedCurrency === 'USD'
        ? { leading: currencySymbol }
        : { trailing: currencySymbol };
    const isFiatFiatOrStable = isFiatOrStablecoin(fiatSymbol);
    const createCryptoComPayment = useCallback(() => __awaiter(void 0, void 0, void 0, function* () {
        if (!primaryWallet ||
            !fiatAmount ||
            isNaN(parseFloat(fiatAmount)) ||
            !environmentId ||
            isCreatingPaymentRef.current)
            return;
        isCreatingPaymentRef.current = true;
        setIsCreatingPayment(true);
        setError(null);
        try {
            const onrampProviders = yield getOnrampProviders({
                chain: primaryWallet.chain,
                currency: selectedCurrency,
                environmentId,
                includeDisabled: true,
                networkId: selectedNetworkId ? String(selectedNetworkId) : undefined,
                tokenAmount: parseFloat(fiatAmount),
                walletAddress: primaryWallet.address,
            });
            const cryptoComProvider = onrampProviders.find((provider) => provider.provider === ProviderEnum.CryptoDotCom);
            if (!(cryptoComProvider === null || cryptoComProvider === void 0 ? void 0 : cryptoComProvider.url)) {
                throw new Error('Crypto.com provider not available');
            }
            if (!isCryptoComUrl(cryptoComProvider.url)) {
                throw new Error('Invalid Crypto.com URL');
            }
            if (onPaymentCreated) {
                onPaymentCreated(cryptoComProvider.url);
            }
            else {
                window.open(cryptoComProvider.url, '_blank', 'noopener,noreferrer');
            }
        }
        catch (error) {
            setError('Failed to create payment. Please try again.');
        }
        finally {
            isCreatingPaymentRef.current = false;
            setIsCreatingPayment(false);
        }
    }), [
        primaryWallet,
        fiatAmount,
        selectedCurrency,
        environmentId,
        selectedNetworkId,
        onPaymentCreated,
    ]);
    const disableSubmit = !fiatAmount || parseFloat(fiatAmount) <= 0 || isCreatingPayment;
    const { backButton, closeButton } = createFundingButtons({
        defaultBackAction: handleBack,
        hideBackButton,
        onBack,
        setShowAuthFlow,
        setShowDynamicUserProfile,
    });
    if (!primaryWallet)
        return null;
    if (showCurrencySelect) {
        return (jsx(CurrencySelectScreen, { onClose: () => setShowCurrencySelect(false), onSelectCurrency: (currencyCode) => {
                setSelectedCurrency(currencyCode);
                setShowCurrencySelect(false);
            }, currencies: availableCurrencies, currentCurrency: selectedCurrency }));
    }
    return (jsxs("div", { className: 'fund-from-wallet', children: [jsx(ModalHeader, { leading: backButton, trailing: closeButton, children: jsx(Typography, { variant: 'title', copykey: 'dyn_crypto_com_onramp.title', children: t('dyn_crypto_com_onramp.title', 'Deposit') }) }), jsxs("div", { className: 'fund-from-wallet__content', children: [jsxs("div", { className: 'fund-from-wallet__content__amount-container', children: [error && (jsx(Typography, { color: 'error-1', className: 'fund-from-wallet__content__amount-container__error', variant: 'body_normal', weight: 'medium', children: error })), primaryData.amount !== undefined && (jsx(FormattedInput, Object.assign({ className: 'fund-from-wallet__content__amount-container__amount', value: primaryData.amount, onChange: primaryData.setAmount, emitter: inputEmitter, onInteraction: () => setHasUserInteracted(true) }, inputSymbolProp))), showQuickSuggestions && (jsx("div", { className: 'fund-from-wallet__content__amount-container__quick-suggestions', children: quickSuggestionsParsed.map(({ display, value }) => (jsx(Typography, { variant: 'body_small', color: 'secondary', onClick: () => handleQuickSuggestion(value), weight: 'medium', children: display }, value))) }))] }), jsxs("div", { className: 'fund-from-wallet__content__currency-selector', children: [jsx(Typography, { variant: 'body_small', weight: 'medium', copykey: 'dyn_crypto_com_onramp.currency_label', children: t('dyn_crypto_com_onramp.currency_label', 'Payment Currency') }), jsxs("div", { className: 'fund-from-wallet__content__currency-selector__dropdown', onClick: () => setShowCurrencySelect(true), role: 'button', tabIndex: 0, onKeyDown: (e) => {
                                    if (e.key === 'Enter' || e.key === ' ') {
                                        e.preventDefault();
                                        setShowCurrencySelect(true);
                                    }
                                }, "aria-label": t('dyn_crypto_com_onramp.select_currency', 'Select Currency'), "aria-expanded": showCurrencySelect, "aria-haspopup": 'dialog', children: [jsxs("div", { className: 'fund-from-wallet__content__currency-selector__dropdown__content', children: [jsx(Typography, { variant: 'body_small', weight: 'medium', children: ((_c = availableCurrencies.find((c) => c.code === selectedCurrency)) === null || _c === void 0 ? void 0 : _c.name) || selectedCurrency }), jsx(Typography, { variant: 'body_small', color: 'secondary', weight: 'medium', children: ((_d = availableCurrencies.find((c) => c.code === selectedCurrency)) === null || _d === void 0 ? void 0 : _d.symbol) || selectedCurrency })] }), jsx("div", { className: 'fund-from-wallet__content__currency-selector__dropdown__icon', children: jsx(Icon, { color: 'text-tertiary', size: 'small', children: jsx(SvgChevronDown, { "data-testid": 'currency-select-dropdown' }) }) })] })] }), jsxs("div", { className: 'fund-from-wallet__content__wallet-card', children: [jsxs("div", { className: 'fund-from-wallet__content__wallet-card__wallet-details', children: [jsx(WalletIcon, { icon: primaryWallet.connector.metadata.icon, walletKey: primaryWallet.connector.key, className: 'fund-from-wallet__content__wallet-card__wallet-details__icon' }), jsxs("div", { className: 'fund-from-wallet__content__wallet-card__wallet-details__rows', children: [jsx(Typography, { variant: 'body_small', weight: 'medium', copykey: 'dyn_crypto_com_onramp.wallet_detail_to', children: t('dyn_crypto_com_onramp.wallet_detail_to', 'Tokens will be sent to') }), jsx(Typography, { variant: 'body_small', color: 'secondary', weight: 'medium', children: shortenWalletAddress(primaryWallet.address) })] })] }), jsx("div", { className: 'fund-from-wallet__content__wallet-card__balance', children: jsxs("div", { className: 'fund-from-wallet__content__wallet-card__balance__rows', children: [jsx(Typography, { variant: 'body_small', weight: 'medium', children: formatValue({
                                                maxDecimals: isFiatFiatOrStable ? 2 : 6,
                                                symbol: currencySymbol,
                                                value: fiatAmount || '0',
                                                withFixedZeros: isFiatFiatOrStable,
                                            }) }), jsx(Typography, { variant: 'body_small', color: 'secondary', weight: 'medium', children: selectedCurrency })] }) })] })] }), jsx(TypographyButton, { dataTestId: 'confirm-button', buttonVariant: 'brand-primary', buttonPadding: 'large', typographyProps: { color: 'white' }, className: 'fund-from-wallet__confirm-button', disabled: disableSubmit, onClick: createCryptoComPayment, children: isCreatingPayment
                    ? t('dyn_crypto_com_onramp.creating_payment', 'Creating Payment...')
                    : t('dyn_crypto_com_onramp.create_payment', 'Continue to Payment') })] }));
};

export { CryptoComOnramp };
