// @dynamic-labs/sdk-react-core/src/lib/widgets/DynamicWidget/views/CryptoComOnramp/CryptoComOnramp.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../../../../../../_virtual/_tslib.cjs');
var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var reactI18next = require('react-i18next');
var walletBook = require('@dynamic-labs/wallet-book');
var sdkApiCore = require('@dynamic-labs/sdk-api-core');
var Icon = require('../../../../components/Icon/Icon.cjs');
var ModalHeader = require('../../../../components/ModalHeader/ModalHeader.cjs');
var Typography = require('../../../../components/Typography/Typography.cjs');
var TypographyButton = require('../../../../components/TypographyButton/TypographyButton.cjs');
require('../../../../context/DynamicContext/DynamicContext.cjs');
require('../../../../store/state/loadingAndLifecycle/loadingAndLifecycle.cjs');
require('@dynamic-labs/iconic');
require('@dynamic-labs/wallet-connector-core');
var chevronDown = require('../../../../shared/assets/chevron-down.cjs');
require('../../../../context/ViewContext/ViewContext.cjs');
require('../../../../shared/logger.cjs');
require('@dynamic-labs/utils');
require('../../../../utils/constants/colors.cjs');
require('../../../../utils/constants/values.cjs');
require('../../../../shared/consts/index.cjs');
require('../../../../events/dynamicEvents.cjs');
require('../../../../context/CaptchaContext/CaptchaContext.cjs');
require('../../../../context/ErrorContext/ErrorContext.cjs');
require('@dynamic-labs/multi-wallet');
require('react-international-phone');
require('../../../../store/state/nonce/nonce.cjs');
require('@dynamic-labs-sdk/client/core');
require('../../../../client/client.cjs');
require('@dynamic-labs-sdk/client');
require('../../../../config/ApiEndpoint.cjs');
var onramp = require('../../../../data/api/onramp/onramp.cjs');
require('@dynamic-labs/locale');
var dynamicContextProps = require('../../../../store/state/dynamicContextProps/dynamicContextProps.cjs');
require('../../../../store/state/primaryWalletId/primaryWalletId.cjs');
require('../../../../store/state/connectedWalletsInfo/connectedWalletsInfo.cjs');
require('../../../../context/AccessDeniedContext/AccessDeniedContext.cjs');
require('../../../../context/AccountExistsContext/AccountExistsContext.cjs');
require('../../../../context/UserWalletsContext/UserWalletsContext.cjs');
require('../../../../store/state/authMode/authMode.cjs');
require('../../../../context/VerificationContext/VerificationContext.cjs');
require('react-dom');
require('../../../../utils/functions/compareChains/compareChains.cjs');
require('../../../../views/Passkey/utils/findPrimaryEmbeddedChain/findPrimaryEmbeddedChain.cjs');
require('../../../../context/ThemeContext/ThemeContext.cjs');
require('../../../../utils/hooks/useUserUpdateRequest/useUpdateUser/userFieldsSchema.cjs');
require('bs58');
require('@dynamic-labs/types');
require('../../../../context/SocialRedirectContext/SocialRedirectContext.cjs');
require('../../../../context/LoadingContext/LoadingContext.cjs');
require('../../../../context/WalletContext/WalletContext.cjs');
require('../../../../utils/hooks/useEmbeddedWallet/useSecureEnclaveEmbeddedWallet/constants.cjs');
require('yup');
require('../../../../context/MockContext/MockContext.cjs');
require('../../../../views/CollectUserDataView/useFields.cjs');
require('../../../../context/FieldsStateContext/FieldsStateContext.cjs');
require('../../../../context/UserFieldEditorContext/UserFieldEditorContext.cjs');
require('@dynamic-labs/rpc-providers');
require('../../../../store/state/walletOptions/walletOptions.cjs');
require('../../../../components/Accordion/components/AccordionItem/AccordionItem.cjs');
require('../../../../components/Alert/Alert.cjs');
require('../../../../components/ShadowDOM/ShadowDOM.cjs');
require('../../../../components/IconButton/IconButton.cjs');
require('../../../../components/InlineWidget/InlineWidget.cjs');
require('../../../../components/Input/Input.cjs');
require('../../../../components/IsBrowser/IsBrowser.cjs');
require('../../../../components/MenuList/Dropdown/Dropdown.cjs');
require('../../../../components/OverlayCard/OverlayCard.cjs');
require('../../../../components/Transition/ZoomTransition/ZoomTransition.cjs');
require('../../../../components/Transition/SlideInUpTransition/SlideInUpTransition.cjs');
require('../../../../components/Transition/OpacityTransition/OpacityTransition.cjs');
require('../../../../components/PasskeyCreatedSuccessBanner/PasskeyCreatedSuccessBanner.cjs');
require('../../../../components/Popper/Popper/Popper.cjs');
require('../../../../components/Popper/PopperContext/PopperContext.cjs');
require('react-focus-lock');
require('qrcode');
var shortenWalletAddress = require('../../../../shared/utils/functions/shortenWalletAddress/shortenWalletAddress.cjs');
require('formik');
require('../../../../utils/hooks/useSubdomainCheck/useSubdomainCheck.cjs');
require('../../../../context/WalletGroupContext/WalletGroupContext.cjs');
require('../../../../context/IpConfigurationContext/IpConfigurationContext.cjs');
require('../../../../context/ConnectWithOtpContext/ConnectWithOtpContext.cjs');
require('../../../DynamicBridgeWidget/views/WalletsView/components/SecondaryWallets/SecondaryWallets.cjs');
require('@hcaptcha/react-hcaptcha');
var DynamicWidgetContext = require('../../context/DynamicWidgetContext.cjs');
require('../../helpers/convertExchangeKeyAndProviderEnum.cjs');
require('../../../../views/ExchangeWhitelistWarning/ExchangeWhitelistWarning.cjs');
require('../../../../context/ErrorContext/hooks/useErrorText/useErrorText.cjs');
require('../../../../context/FooterAnimationContext/index.cjs');
require('../../../../views/MfaChooseDeviceView/useGetMfaOptions/useGetMfaOptions.cjs');
require('../../../../context/PasskeyContext/PasskeyContext.cjs');
require('../../../../context/OnrampContext/OnrampContext.cjs');
require('../../../../store/state/sendBalances.cjs');
require('../../../../store/state/connectorsInitializing/connectorsInitializing.cjs');
require('../../../../components/OverlayCardBase/OverlayCardTarget/OverlayCardTarget.cjs');
require('../../components/DynamicWidgetHeader/DynamicWidgetHeader.cjs');
require('../../../../views/TransactionConfirmationView/TransactionConfirmationView.cjs');
require('../../components/PasskeyCard/PasskeyCard.cjs');
require('../../../../../index.cjs');
require('../ReceiveWalletFunds/ReceiveWalletFunds.cjs');
require('../../../../store/state/tokenBalances.cjs');
require('../../../../store/state/multichainBalances.cjs');
require('../../../../shared/utils/functions/getInitialUrl/getInitialUrl.cjs');
var useInternalDynamicContext = require('../../../../context/DynamicContext/useDynamicContext/useInternalDynamicContext/useInternalDynamicContext.cjs');
var FormattedInput = require('../ReceiveWalletFunds/FormattedInput/FormattedInput.cjs');
var formattedInputEmitter = require('../ReceiveWalletFunds/FormattedInput/formattedInputEmitter.cjs');
var formatValue = require('../ReceiveWalletFunds/utils/formatValue/formatValue.cjs');
var isFiatToken = require('../ReceiveWalletFunds/utils/isFiatToken/isFiatToken.cjs');
var createFundingButtons = require('../ReceiveWalletFunds/utils/createFundingButtons/createFundingButtons.cjs');
var CurrencySelectScreen = require('./CurrencySelectScreen/CurrencySelectScreen.cjs');

const defaultQuickSuggestions = [25, 100, 500];
const availableCurrencies = [
    { code: 'USD', name: 'US Dollar', symbol: '$' },
    { code: 'EUR', name: 'Euro', symbol: '€' },
    { code: 'AUD', name: 'Australian Dollar', symbol: 'A$' },
    { code: 'CAD', name: 'Canadian Dollar', symbol: 'C$' },
    { code: 'GBP', name: 'British Pound', symbol: '£' },
    { code: 'BGN', name: 'Bulgarian Lev', symbol: 'лв' },
    { code: 'BRL', name: 'Brazilian Real', symbol: 'R$' },
    { code: 'CHF', name: 'Swiss Franc', symbol: 'CHF' },
    { code: 'CLP', name: 'Chilean Peso', symbol: '$' },
    { code: 'CNY', name: 'Chinese Yuan', symbol: '¥' },
    { code: 'COP', name: 'Colombian Peso', symbol: '$' },
    { code: 'DKK', name: 'Danish Krone', symbol: 'kr' },
    { code: 'HKD', name: 'Hong Kong Dollar', symbol: 'HK$' },
    { code: 'IDR', name: 'Indonesian Rupiah', symbol: 'Rp' },
    { code: 'INR', name: 'Indian Rupee', symbol: '₹' },
    { code: 'JPY', name: 'Japanese Yen', symbol: '¥' },
    { code: 'KRW', name: 'South Korean Won', symbol: '₩' },
    { code: 'MDL', name: 'Moldovan Leu', symbol: 'L' },
    { code: 'MOP', name: 'Macanese Pataca', symbol: 'MOP$' },
    { code: 'MXN', name: 'Mexican Peso', symbol: '$' },
    { code: 'MYR', name: 'Malaysian Ringgit', symbol: 'RM' },
    { code: 'NOK', name: 'Norwegian Krone', symbol: 'kr' },
    { code: 'PHP', name: 'Philippine Peso', symbol: '₱' },
    { code: 'RON', name: 'Romanian Leu', symbol: 'lei' },
    { code: 'RUB', name: 'Russian Ruble', symbol: '₽' },
    { code: 'SAR', name: 'Saudi Riyal', symbol: '﷼' },
    { code: 'SEK', name: 'Swedish Krona', symbol: 'kr' },
    { code: 'SGD', name: 'Singapore Dollar', symbol: 'S$' },
    { code: 'THB', name: 'Thai Baht', symbol: '฿' },
    { code: 'TRY', name: 'Turkish Lira', symbol: '₺' },
    { code: 'TWD', name: 'Taiwan Dollar', symbol: 'NT$' },
    { code: 'UAH', name: 'Ukrainian Hryvnia', symbol: '₴' },
    { code: 'ZAR', name: 'South African Rand', symbol: 'R' },
];
const inputEmitter = formattedInputEmitter.createFormattedInputEmitter();
const isCryptoComUrl = (url) => {
    try {
        const parsedUrl = new URL(url);
        return (parsedUrl.protocol === 'https:' &&
            (parsedUrl.hostname === 'crypto.com' ||
                parsedUrl.hostname.endsWith('.crypto.com')));
    }
    catch (_a) {
        return false;
    }
};
const CryptoComOnramp = ({ onBack, hideBackButton = false, onPaymentCreated, quickSuggestions = defaultQuickSuggestions, currency: initialCurrency, }) => {
    var _a, _b, _c, _d;
    const { t } = reactI18next.useTranslation();
    const { primaryWallet, setShowAuthFlow, setShowDynamicUserProfile, network } = useInternalDynamicContext.useInternalDynamicContext();
    const { setDynamicWidgetView } = DynamicWidgetContext.useWidgetContext();
    const environmentId = dynamicContextProps.useEnvironmentId();
    const [hasUserInteracted, setHasUserInteracted] = React.useState(false);
    const [isCreatingPayment, setIsCreatingPayment] = React.useState(false);
    const [error, setError] = React.useState(null);
    const isCreatingPaymentRef = React.useRef(false);
    // Amount state management
    const [fiatAmount, setFiatAmount] = React.useState('');
    const [selectedCurrency, setSelectedCurrency] = React.useState(initialCurrency &&
        availableCurrencies.some((c) => c.code === initialCurrency)
        ? initialCurrency
        : 'USD');
    React.useEffect(() => {
        if (initialCurrency &&
            availableCurrencies.some((c) => c.code === initialCurrency)) {
            setSelectedCurrency(initialCurrency);
        }
    }, [initialCurrency]);
    const fiatSymbol = ((_a = availableCurrencies.find((c) => c.code === selectedCurrency)) === null || _a === void 0 ? void 0 : _a.code) || 'USD';
    const currencySymbol = ((_b = availableCurrencies.find((c) => c.code === selectedCurrency)) === null || _b === void 0 ? void 0 : _b.symbol) || '$';
    // Network selection state
    const [selectedNetworkId, setSelectedNetworkId] = React.useState(network);
    React.useEffect(() => {
        setSelectedNetworkId(network);
    }, [network]);
    const setAmountsByFiatValue = React.useCallback((amount) => {
        setFiatAmount(amount);
    }, []);
    const quickSuggestionsParsed = Array.isArray(quickSuggestions)
        ? quickSuggestions.map((value) => ({ display: `$${value}`, value }))
        : [];
    const showQuickSuggestions = quickSuggestionsParsed.length > 0 &&
        !hasUserInteracted &&
        selectedCurrency === 'USD';
    const [showCurrencySelect, setShowCurrencySelect] = React.useState(false);
    const handleQuickSuggestion = React.useCallback((value) => {
        setAmountsByFiatValue(value.toString());
        setHasUserInteracted(true);
    }, [setAmountsByFiatValue]);
    const handleBack = React.useCallback(() => {
        if (onBack) {
            onBack();
        }
        else {
            setDynamicWidgetView('choose-onramp-provider');
        }
    }, [onBack, setDynamicWidgetView]);
    // Primary data - only show the deposit amount and selected token
    const primaryData = {
        amount: fiatAmount,
        setAmount: setAmountsByFiatValue,
        symbol: fiatSymbol,
    };
    const inputSymbolProp = selectedCurrency === 'USD'
        ? { leading: currencySymbol }
        : { trailing: currencySymbol };
    const isFiatFiatOrStable = isFiatToken.isFiatOrStablecoin(fiatSymbol);
    const createCryptoComPayment = React.useCallback(() => _tslib.__awaiter(void 0, void 0, void 0, function* () {
        if (!primaryWallet ||
            !fiatAmount ||
            isNaN(parseFloat(fiatAmount)) ||
            !environmentId ||
            isCreatingPaymentRef.current)
            return;
        isCreatingPaymentRef.current = true;
        setIsCreatingPayment(true);
        setError(null);
        try {
            const onrampProviders = yield onramp.getOnrampProviders({
                chain: primaryWallet.chain,
                currency: selectedCurrency,
                environmentId,
                includeDisabled: true,
                networkId: selectedNetworkId ? String(selectedNetworkId) : undefined,
                tokenAmount: parseFloat(fiatAmount),
                walletAddress: primaryWallet.address,
            });
            const cryptoComProvider = onrampProviders.find((provider) => provider.provider === sdkApiCore.ProviderEnum.CryptoDotCom);
            if (!(cryptoComProvider === null || cryptoComProvider === void 0 ? void 0 : cryptoComProvider.url)) {
                throw new Error('Crypto.com provider not available');
            }
            if (!isCryptoComUrl(cryptoComProvider.url)) {
                throw new Error('Invalid Crypto.com URL');
            }
            if (onPaymentCreated) {
                onPaymentCreated(cryptoComProvider.url);
            }
            else {
                window.open(cryptoComProvider.url, '_blank', 'noopener,noreferrer');
            }
        }
        catch (error) {
            setError('Failed to create payment. Please try again.');
        }
        finally {
            isCreatingPaymentRef.current = false;
            setIsCreatingPayment(false);
        }
    }), [
        primaryWallet,
        fiatAmount,
        selectedCurrency,
        environmentId,
        selectedNetworkId,
        onPaymentCreated,
    ]);
    const disableSubmit = !fiatAmount || parseFloat(fiatAmount) <= 0 || isCreatingPayment;
    const { backButton, closeButton } = createFundingButtons.createFundingButtons({
        defaultBackAction: handleBack,
        hideBackButton,
        onBack,
        setShowAuthFlow,
        setShowDynamicUserProfile,
    });
    if (!primaryWallet)
        return null;
    if (showCurrencySelect) {
        return (jsxRuntime.jsx(CurrencySelectScreen.CurrencySelectScreen, { onClose: () => setShowCurrencySelect(false), onSelectCurrency: (currencyCode) => {
                setSelectedCurrency(currencyCode);
                setShowCurrencySelect(false);
            }, currencies: availableCurrencies, currentCurrency: selectedCurrency }));
    }
    return (jsxRuntime.jsxs("div", { className: 'fund-from-wallet', children: [jsxRuntime.jsx(ModalHeader.ModalHeader, { leading: backButton, trailing: closeButton, children: jsxRuntime.jsx(Typography.Typography, { variant: 'title', copykey: 'dyn_crypto_com_onramp.title', children: t('dyn_crypto_com_onramp.title', 'Deposit') }) }), jsxRuntime.jsxs("div", { className: 'fund-from-wallet__content', children: [jsxRuntime.jsxs("div", { className: 'fund-from-wallet__content__amount-container', children: [error && (jsxRuntime.jsx(Typography.Typography, { color: 'error-1', className: 'fund-from-wallet__content__amount-container__error', variant: 'body_normal', weight: 'medium', children: error })), primaryData.amount !== undefined && (jsxRuntime.jsx(FormattedInput.FormattedInput, Object.assign({ className: 'fund-from-wallet__content__amount-container__amount', value: primaryData.amount, onChange: primaryData.setAmount, emitter: inputEmitter, onInteraction: () => setHasUserInteracted(true) }, inputSymbolProp))), showQuickSuggestions && (jsxRuntime.jsx("div", { className: 'fund-from-wallet__content__amount-container__quick-suggestions', children: quickSuggestionsParsed.map(({ display, value }) => (jsxRuntime.jsx(Typography.Typography, { variant: 'body_small', color: 'secondary', onClick: () => handleQuickSuggestion(value), weight: 'medium', children: display }, value))) }))] }), jsxRuntime.jsxs("div", { className: 'fund-from-wallet__content__currency-selector', children: [jsxRuntime.jsx(Typography.Typography, { variant: 'body_small', weight: 'medium', copykey: 'dyn_crypto_com_onramp.currency_label', children: t('dyn_crypto_com_onramp.currency_label', 'Payment Currency') }), jsxRuntime.jsxs("div", { className: 'fund-from-wallet__content__currency-selector__dropdown', onClick: () => setShowCurrencySelect(true), role: 'button', tabIndex: 0, onKeyDown: (e) => {
                                    if (e.key === 'Enter' || e.key === ' ') {
                                        e.preventDefault();
                                        setShowCurrencySelect(true);
                                    }
                                }, "aria-label": t('dyn_crypto_com_onramp.select_currency', 'Select Currency'), "aria-expanded": showCurrencySelect, "aria-haspopup": 'dialog', children: [jsxRuntime.jsxs("div", { className: 'fund-from-wallet__content__currency-selector__dropdown__content', children: [jsxRuntime.jsx(Typography.Typography, { variant: 'body_small', weight: 'medium', children: ((_c = availableCurrencies.find((c) => c.code === selectedCurrency)) === null || _c === void 0 ? void 0 : _c.name) || selectedCurrency }), jsxRuntime.jsx(Typography.Typography, { variant: 'body_small', color: 'secondary', weight: 'medium', children: ((_d = availableCurrencies.find((c) => c.code === selectedCurrency)) === null || _d === void 0 ? void 0 : _d.symbol) || selectedCurrency })] }), jsxRuntime.jsx("div", { className: 'fund-from-wallet__content__currency-selector__dropdown__icon', children: jsxRuntime.jsx(Icon.Icon, { color: 'text-tertiary', size: 'small', children: jsxRuntime.jsx(chevronDown.ReactComponent, { "data-testid": 'currency-select-dropdown' }) }) })] })] }), jsxRuntime.jsxs("div", { className: 'fund-from-wallet__content__wallet-card', children: [jsxRuntime.jsxs("div", { className: 'fund-from-wallet__content__wallet-card__wallet-details', children: [jsxRuntime.jsx(walletBook.WalletIcon, { icon: primaryWallet.connector.metadata.icon, walletKey: primaryWallet.connector.key, className: 'fund-from-wallet__content__wallet-card__wallet-details__icon' }), jsxRuntime.jsxs("div", { className: 'fund-from-wallet__content__wallet-card__wallet-details__rows', children: [jsxRuntime.jsx(Typography.Typography, { variant: 'body_small', weight: 'medium', copykey: 'dyn_crypto_com_onramp.wallet_detail_to', children: t('dyn_crypto_com_onramp.wallet_detail_to', 'Tokens will be sent to') }), jsxRuntime.jsx(Typography.Typography, { variant: 'body_small', color: 'secondary', weight: 'medium', children: shortenWalletAddress.shortenWalletAddress(primaryWallet.address) })] })] }), jsxRuntime.jsx("div", { className: 'fund-from-wallet__content__wallet-card__balance', children: jsxRuntime.jsxs("div", { className: 'fund-from-wallet__content__wallet-card__balance__rows', children: [jsxRuntime.jsx(Typography.Typography, { variant: 'body_small', weight: 'medium', children: formatValue.formatValue({
                                                maxDecimals: isFiatFiatOrStable ? 2 : 6,
                                                symbol: currencySymbol,
                                                value: fiatAmount || '0',
                                                withFixedZeros: isFiatFiatOrStable,
                                            }) }), jsxRuntime.jsx(Typography.Typography, { variant: 'body_small', color: 'secondary', weight: 'medium', children: selectedCurrency })] }) })] })] }), jsxRuntime.jsx(TypographyButton.TypographyButton, { dataTestId: 'confirm-button', buttonVariant: 'brand-primary', buttonPadding: 'large', typographyProps: { color: 'white' }, className: 'fund-from-wallet__confirm-button', disabled: disableSubmit, onClick: createCryptoComPayment, children: isCreatingPayment
                    ? t('dyn_crypto_com_onramp.creating_payment', 'Creating Payment...')
                    : t('dyn_crypto_com_onramp.create_payment', 'Continue to Payment') })] }));
};

exports.CryptoComOnramp = CryptoComOnramp;
