// @dynamic-labs/wallet-connector-core/src/index.js
'use client'
import { assertPackageVersion } from '@dynamic-labs/assert-package-version';
import { version } from '../package.js';
export { walletConnectorEvents } from './events/walletConnectorEvents.js';
export { logger } from './utils/logger.js';
export { ProviderLookup } from './utils/ProviderLookup/ProviderLookup.js';
export { getChainInfo, getChainInfoWithOverrides, setChainInfoOverrides } from './utils/getChainInfo/getChainInfo.js';
export { isEmailOTPWalletConnector } from './utils/isEmailOTPWalletConnector/isEmailOTPWalletConnector.js';
export { isEmailWalletConnector } from './utils/isEmailWalletConnector/isEmailWalletConnector.js';
export { isTurnkeyWalletConnector } from './utils/isTurnkeyWalletConnector/isTurnkeyWalletConnector.js';
export { isSameAddress } from './utils/isSameAddress/isSameAddress.js';
export { isSocialWalletConnector } from './utils/isSocialWalletConnector/isSocialWalletConnector.js';
export { addHexPrefix, bufferToHex, utf8ToHex } from './utils/encoding.js';
export { getWalletConnectorByKey } from './utils/getWalletConnectorByKey.js';
export { shouldLowercaseAddress } from './utils/shouldLowercaseAddress.js';
export { getDeepLink } from './utils/walletConnectDeepLinks/walletConnectDeepLinks.js';
export { performPlatformSpecificConnectionMethod } from './utils/walletConnectDeepLinks/performPlatformSpecificConnectionMethod.js';
export { isEmbeddedConnector } from './utils/isEmbeddedConnector/isEmbeddedConnector.js';
export { isMagicConnector } from './utils/isMagicConnector/isMagicConnector.js';
export { isBloctoConnector } from './utils/isBloctoConnector/isBloctoConnector.js';
export { isPasskeyWalletConnector } from './utils/isPasskeyWalletConnector/isPasskeyWalletConnector.js';
export { isAccountAbstractionConnector } from './utils/isAccountAbstractionConnector/isAccountAbstractionConnector.js';
export { isBitcoinConnector } from './utils/isBitcoinConnector/isBitcoinConnector.js';
export { isHardwareWalletConnector } from './utils/isHardwareWalletConnector/isHardwareWalletConnector.js';
export { eventListenerHandlers } from './utils/eventListenerHandlers.js';
export { isHex } from './utils/isHex/isHex.js';
export { isPhantomRedirectConnector } from './utils/isPhantomRedirectConnector/isPhantomRedirectConnector.js';
export { getRpcUrlForChain } from './utils/getRpcUrlForChain/getRpcUrlForChain.js';
export { isWalletConnectConnector } from './utils/isWalletConnectConnector/isWalletConnectConnector.js';
export { isSendBalanceWalletConnector } from './utils/isSendBalanceWalletConnector/isSendBalanceWalletConnector.js';
export { isSessionKeyCompatibleWallet } from './utils/isSessionKeyCompatible/isSessionKeyCompatibleWallet/isSessionKeyCompatibleWallet.js';
export { isSessionKeyCompatibleWalletConnector } from './utils/isSessionKeyCompatible/isSessionKeyCompatibleWalletConnector/isSessionKeyCompatibleWalletConnector.js';
export { getMobileExperience } from './utils/getMobileExperience/getMobileExperience.js';
export { getWalletMetadataFromWalletBook } from './utils/getWalletMetadataFromWalletBook/getWalletMetadataFromWalletBook.js';
export { isConnectorMethodSupported } from './utils/isConnectorMethodSupported/isConnectorMethodSupported.js';
export { isConnectorEventSupported } from './utils/isConnectorEventSupported/isConnectorEventSupported.js';
export { getWalletLinks } from './utils/getWalletLinks/getWalletLinks.js';
export { convertWalletToBaseWallet, convertWalletToBaseWalletPartial } from './utils/convertWalletToBaseWallet/convertWalletToBaseWallet.js';
export { isDynamicWaasConnector } from './utils/isDynamicWaasConnector/isDynamicWaasConnector.js';
export { getWalletProvider } from './utils/getWalletProvider/getWalletProvider.js';
export { Wallet } from './Wallet/Wallet.js';
export { WalletConnectorBase } from './WalletConnectorBase/WalletConnectorBase.js';
export { Chains, socialProviders } from './WalletConnectorBase/types.js';
export { WalletBookSingleton } from './WalletBookSingleton/WalletBookSingleton.js';

assertPackageVersion('@dynamic-labs/wallet-connector-core', version);
