// @dynamic-labs/sdk-api-core/src/models/CustomFieldType.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.CustomFieldType = void 0;
(function (CustomFieldType) {
    CustomFieldType["Text"] = "text";
    CustomFieldType["Checkbox"] = "checkbox";
    CustomFieldType["Select"] = "select";
})(exports.CustomFieldType || (exports.CustomFieldType = {}));
function CustomFieldTypeFromJSON(json) {
    return CustomFieldTypeFromJSONTyped(json);
}
function CustomFieldTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function CustomFieldTypeToJSON(value) {
    return value;
}

exports.CustomFieldTypeFromJSON = CustomFieldTypeFromJSON;
exports.CustomFieldTypeFromJSONTyped = CustomFieldTypeFromJSONTyped;
exports.CustomFieldTypeToJSON = CustomFieldTypeToJSON;
