// @dynamic-labs/sdk-api-core/src/models/EmbeddedWalletAuthType.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.EmbeddedWalletAuthType = void 0;
(function (EmbeddedWalletAuthType) {
    EmbeddedWalletAuthType["Mpc"] = "mpc";
})(exports.EmbeddedWalletAuthType || (exports.EmbeddedWalletAuthType = {}));
function EmbeddedWalletAuthTypeFromJSON(json) {
    return EmbeddedWalletAuthTypeFromJSONTyped(json);
}
function EmbeddedWalletAuthTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function EmbeddedWalletAuthTypeToJSON(value) {
    return value;
}

exports.EmbeddedWalletAuthTypeFromJSON = EmbeddedWalletAuthTypeFromJSON;
exports.EmbeddedWalletAuthTypeFromJSONTyped = EmbeddedWalletAuthTypeFromJSONTyped;
exports.EmbeddedWalletAuthTypeToJSON = EmbeddedWalletAuthTypeToJSON;
