// @dynamic-labs/sdk-api-core/src/models/ExchangeKeyEnum.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 * Source exchange identifier
 * @export
 * @enum {string}
 */
exports.ExchangeKeyEnum = void 0;
(function (ExchangeKeyEnum) {
    ExchangeKeyEnum["Coinbase"] = "coinbase";
    ExchangeKeyEnum["Kraken"] = "kraken";
})(exports.ExchangeKeyEnum || (exports.ExchangeKeyEnum = {}));
function ExchangeKeyEnumFromJSON(json) {
    return ExchangeKeyEnumFromJSONTyped(json);
}
function ExchangeKeyEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function ExchangeKeyEnumToJSON(value) {
    return value;
}

exports.ExchangeKeyEnumFromJSON = ExchangeKeyEnumFromJSON;
exports.ExchangeKeyEnumFromJSONTyped = ExchangeKeyEnumFromJSONTyped;
exports.ExchangeKeyEnumToJSON = ExchangeKeyEnumToJSON;
