// @dynamic-labs/sdk-api-core/src/models/ProviderEntryPointVersionEnum.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
var ProviderEntryPointVersionEnum;
(function (ProviderEntryPointVersionEnum) {
    ProviderEntryPointVersionEnum["V6"] = "v6";
    ProviderEntryPointVersionEnum["V7"] = "v7";
})(ProviderEntryPointVersionEnum || (ProviderEntryPointVersionEnum = {}));
function ProviderEntryPointVersionEnumFromJSON(json) {
    return ProviderEntryPointVersionEnumFromJSONTyped(json);
}
function ProviderEntryPointVersionEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function ProviderEntryPointVersionEnumToJSON(value) {
    return value;
}

export { ProviderEntryPointVersionEnum, ProviderEntryPointVersionEnumFromJSON, ProviderEntryPointVersionEnumFromJSONTyped, ProviderEntryPointVersionEnumToJSON };
