// @dynamic-labs/sdk-api-core/src/models/TimeUnitEnum.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.TimeUnitEnum = void 0;
(function (TimeUnitEnum) {
    TimeUnitEnum["Days"] = "days";
    TimeUnitEnum["Hours"] = "hours";
    TimeUnitEnum["Minutes"] = "minutes";
})(exports.TimeUnitEnum || (exports.TimeUnitEnum = {}));
function TimeUnitEnumFromJSON(json) {
    return TimeUnitEnumFromJSONTyped(json);
}
function TimeUnitEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function TimeUnitEnumToJSON(value) {
    return value;
}

exports.TimeUnitEnumFromJSON = TimeUnitEnumFromJSON;
exports.TimeUnitEnumFromJSONTyped = TimeUnitEnumFromJSONTyped;
exports.TimeUnitEnumToJSON = TimeUnitEnumToJSON;
