// @dynamic-labs/sdk-api-core/src/models/UserFieldsCheckEnum.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.UserFieldsCheckEnum = void 0;
(function (UserFieldsCheckEnum) {
    UserFieldsCheckEnum["Email"] = "email";
    UserFieldsCheckEnum["Username"] = "username";
    UserFieldsCheckEnum["NameServiceSubdomainHandle"] = "name-service-subdomain-handle";
})(exports.UserFieldsCheckEnum || (exports.UserFieldsCheckEnum = {}));
function UserFieldsCheckEnumFromJSON(json) {
    return UserFieldsCheckEnumFromJSONTyped(json);
}
function UserFieldsCheckEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function UserFieldsCheckEnumToJSON(value) {
    return value;
}

exports.UserFieldsCheckEnumFromJSON = UserFieldsCheckEnumFromJSON;
exports.UserFieldsCheckEnumFromJSONTyped = UserFieldsCheckEnumFromJSONTyped;
exports.UserFieldsCheckEnumToJSON = UserFieldsCheckEnumToJSON;
