// @dynamic-labs/sdk-api-core/src/models/WalletAddressType.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.WalletAddressType = void 0;
(function (WalletAddressType) {
    WalletAddressType["Ordinals"] = "ordinals";
    WalletAddressType["Payment"] = "payment";
    WalletAddressType["Cosmos"] = "cosmos";
    WalletAddressType["Evm"] = "evm";
})(exports.WalletAddressType || (exports.WalletAddressType = {}));
function WalletAddressTypeFromJSON(json) {
    return WalletAddressTypeFromJSONTyped(json);
}
function WalletAddressTypeFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function WalletAddressTypeToJSON(value) {
    return value;
}

exports.WalletAddressTypeFromJSON = WalletAddressTypeFromJSON;
exports.WalletAddressTypeFromJSONTyped = WalletAddressTypeFromJSONTyped;
exports.WalletAddressTypeToJSON = WalletAddressTypeToJSON;
