// @dynamic-labs/sdk-api-core/src/models/EcdsaValidatorOptions.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
exports.EcdsaValidatorOptions = void 0;
(function (EcdsaValidatorOptions) {
    EcdsaValidatorOptions["SignerToEcdsa"] = "zerodev_signer_to_ecdsa";
    EcdsaValidatorOptions["MultiChain"] = "zerodev_multi_chain";
})(exports.EcdsaValidatorOptions || (exports.EcdsaValidatorOptions = {}));
function EcdsaValidatorOptionsFromJSON(json) {
    return EcdsaValidatorOptionsFromJSONTyped(json);
}
function EcdsaValidatorOptionsFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function EcdsaValidatorOptionsToJSON(value) {
    return value;
}

exports.EcdsaValidatorOptionsFromJSON = EcdsaValidatorOptionsFromJSON;
exports.EcdsaValidatorOptionsFromJSONTyped = EcdsaValidatorOptionsFromJSONTyped;
exports.EcdsaValidatorOptionsToJSON = EcdsaValidatorOptionsToJSON;
