// @dynamic-labs/sdk-api-core/src/models/ProviderKernelVersionEnum.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
var ProviderKernelVersionEnum;
(function (ProviderKernelVersionEnum) {
    ProviderKernelVersionEnum["V24"] = "v2_4";
    ProviderKernelVersionEnum["V30"] = "v3_0";
    ProviderKernelVersionEnum["V31"] = "v3_1";
    ProviderKernelVersionEnum["V32"] = "v3_2";
    ProviderKernelVersionEnum["V33"] = "v3_3";
})(ProviderKernelVersionEnum || (ProviderKernelVersionEnum = {}));
function ProviderKernelVersionEnumFromJSON(json) {
    return ProviderKernelVersionEnumFromJSONTyped(json);
}
function ProviderKernelVersionEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function ProviderKernelVersionEnumToJSON(value) {
    return value;
}

export { ProviderKernelVersionEnum, ProviderKernelVersionEnumFromJSON, ProviderKernelVersionEnumFromJSONTyped, ProviderKernelVersionEnumToJSON };
