// @dynamic-labs/sdk-api-core/src/models/UserVerificationRequirement.js
/* tslint:disable */
/* eslint-disable */
/**
 * Dashboard API
 * Dashboard API documentation
 *
 * The version of the OpenAPI document: 1.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
/**
 *
 * @export
 * @enum {string}
 */
var UserVerificationRequirement;
(function (UserVerificationRequirement) {
    UserVerificationRequirement["Discouraged"] = "discouraged";
    UserVerificationRequirement["Preferred"] = "preferred";
    UserVerificationRequirement["Required"] = "required";
})(UserVerificationRequirement || (UserVerificationRequirement = {}));
function UserVerificationRequirementFromJSON(json) {
    return UserVerificationRequirementFromJSONTyped(json);
}
function UserVerificationRequirementFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function UserVerificationRequirementToJSON(value) {
    return value;
}

export { UserVerificationRequirement, UserVerificationRequirementFromJSON, UserVerificationRequirementFromJSONTyped, UserVerificationRequirementToJSON };
