// @dynamic-labs/sdk-react-core/src/lib/components/Input/Input.js
'use client'
import { __rest } from '../../../../_virtual/_tslib.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import { forwardRef } from 'react';
import { classNames } from '../../utils/functions/classNames/classNames.js';
import { Typography } from '../Typography/Typography.js';

const Input = forwardRef((_a, ref) => {
    var { className = '', id, label, type, onChange, onBlur, name, value, optional, error = false, success = false, disabled, message, placeholder, variant = 'dense', suffix, containerClassName } = _a, rest = __rest(_a, ["className", "id", "label", "type", "onChange", "onBlur", "name", "value", "optional", "error", "success", "disabled", "message", "placeholder", "variant", "suffix", "containerClassName"]);
    return (jsxs("div", { className: classNames('input__container', containerClassName, {
            'input__container--dense': variant === 'dense',
            'input__container--regular': variant === 'regular',
        }), children: [jsx("input", Object.assign({}, rest, { name: name, value: value, onChange: onChange, onBlur: onBlur, type: type, id: id, placeholder: variant === 'regular' || !label ? placeholder : label, disabled: disabled, className: classNames('input', className, {
                    'input--no-label': !label,
                    input__error: error,
                    input__has_trailing: optional,
                    input__success: success,
                }), ref: ref })), label && (jsx("label", { htmlFor: id, className: classNames('input__label'), children: label })), suffix && jsx("div", { className: classNames('input__suffix'), children: suffix }), optional && jsx("span", { className: 'input__trailing', children: "Optional" }), Boolean(message) && (jsx(Typography, { variant: 'body_mini', className: classNames('input__message', {
                    input__message__error: error,
                    input__message__success: success,
                }), children: message }))] }));
});
Input.displayName = 'Input';

export { Input };
